<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>The "MSnSet" Class for MS Proteomics Expression Data and Meta-Data — MSnSet-class • MSnbase</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="The " msnset class for ms proteomics expression data and meta-data msnset-class><meta property="og:description" content='The MSnSet holds quantified expression data for MS proteomics
  data and the experimental meta-data.
  The MSnSet class is derived from the
  "eSet" class and mimics the
  "ExpressionSet" class classically used for
  microarray data.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.29.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/v04-benchmarking.html">MSnbase benchmarking</a>
    </li>
    <li>
      <a href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lgatto/MSnbase/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The "MSnSet" Class for MS Proteomics Expression Data and Meta-Data</h1>
    
    <div class="hidden name"><code>MSnSet-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>MSnSet</code> holds quantified expression data for MS proteomics
  data and the experimental meta-data.
  The <code>MSnSet</code> class is derived from the
  <code>"eSet"</code> class and mimics the
  <code>"ExpressionSet"</code> class classically used for
  microarray data.</p>
    </div>


    <div id="objects-from-the-class">
    <h2>Objects from the Class</h2>
    

<p>The constructor <code>MSnSet(exprs, fData, pData)</code> can be used to
  create <code>MSnSet</code> instances. Argument <code>exprs</code> is a
  <code>matrix</code> and <code>fData</code> and <code>pData</code> must be of class
  <code>data.frame</code> or <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.AnnotatedDataFrame.html" class="external-link">AnnotatedDataFrame</a>"</code> and all
  must meet the dimensions and name validity constrains.</p>
<p>Objects can also be created by calls of the form <code>new("MSnSet",
  exprs, ...)</code>.  See also <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html" class="external-link">ExpressionSet</a>"</code> for
  helpful information.  Expression data produced from other softwares
  can thus make use of this standardized data container to benefit
  <code>R</code> and <code>Bioconductor</code> packages. Proteomics expression data
  available as spreadsheets, as produced by third-party software such as
  Proteome Discoverer, MaxQuant, ... can be imported using the
  <code><a href="readMSnSet.html">readMSnSet</a></code> and <code><a href="readMSnSet.html">readMSnSet2</a></code> functions.</p>
<p>Coercion methods are also available to transform <code>MSnSet</code> objects
  to <code>IBSpectra</code>, to <code>data.frame</code> and to/from
  <code>ExpressionSet</code> and <code>SummarizedExperiment</code> objects. In the
  latter case, the metadata available in the <code>protocolData</code>,
  <code>experimentData</code> are completely dropped, and only the logging
  information of the <code>processingData</code> slot is retained. All these
  metadata can be subsequently be added using the
  <code>addMSnSetMetadata</code> (see examples below). When converting a
  <code>SummarizedExperiment</code> to an <code>MSnSet</code>, the respective
  metadata slots will be populated if available in the
  <code>SummarizedExperiment</code> metadata.</p>
<p>In the frame of the <code>MSnbase</code> package, <code>MSnSet</code> instances
  can be generated from <code>"<a href="MSnExp-class.html">MSnExp</a>"</code> experiments using
  the <code><a href="quantify-methods.html">quantify</a></code> method).</p>
    </div>
    <div id="slots">
    <h2>Slots</h2>
    

<p></p><dl><dt><code>qual</code>:</dt>
<dd><p>Object of class <code>"data.frame"</code> that records
      peaks data for each of the reporter ions to be used as quality
      metrics.</p></dd>


    <dt><code>processingData</code>:</dt>
<dd><p>Object of class
      <code>"<a href="MSnProcess-class.html">MSnProcess</a>"</code> that records all processing.</p></dd>


    <dt><code>assayData</code>:</dt>
<dd><p>Object of class <code>"assayData"</code>
      containing a matrix with equal with column number equal to
      <code>nrow(phenoData)</code>. <code>assayData</code> must contain a matrix
      <code>exprs</code> with rows represening features (e.g., reporters ions)
      and columns representing samples. See the <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.AssayData.html" class="external-link">AssayData</a>"</code>
      class, <code><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></code> and <code><a href="https://rdrr.io/pkg/Biobase/man/assayData.html" class="external-link">assayData</a></code> accessor
      for more details. This slot in indirectly inherited from
      <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>.</p></dd>


    <dt><code>phenoData</code>:</dt>
<dd><p>Object of class <code>"AnnotatedDataFrame"</code>
      containing experimenter-supplied variables describing sample (i.e
      the individual tags for an labelled MS experiment) (indireclty
      inherited from <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>). See
      <code><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">phenoData</a></code> and the <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code> class
      for more details. This slot can be accessed as a <code>data.frame</code>
      with <code>pData</code> and be replaced by a new valid (i.e. of
      compatible dimensions and row names) <code>data.frame</code> with
      <code>pData()&lt;-</code>.</p></dd>


    <dt><code>featureData</code>:</dt>
<dd><p>Object of class
      <code>"AnnotatedDataFrame"</code> containing variables describing
      features (spectra in our case), e.g. identificaiton data, peptide
      sequence, identification score,... (inherited indirectly from
      <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>). See <code><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">featureData</a></code> and
      the <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code> class for more details. This slot
      can be accessed as a <code>data.frame</code> with <code>fData</code> and be
      replaced by a new valid (i.e. of compatible dimensions and row
      names) <code>data.frame</code> with <code>fData()&lt;-</code>.</p></dd>


    <dt><code>experimentData</code>:</dt>
<dd><p>Object of class
      <code>"<a href="MIAPE-class.html">MIAPE</a>"</code>, containing details of experimental
      methods (inherited from <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>). See
      <code><a href="https://rdrr.io/pkg/Biobase/man/abstract.html" class="external-link">experimentData</a></code> and the <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>
      class for more details.</p></dd>


    <dt><code>annotation</code>:</dt>
<dd><p>not used here.</p></dd>


    <dt><code>protocolData</code>:</dt>
<dd><p>Object of class
      <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.AnnotatedDataFrame.html" class="external-link">AnnotatedDataFrame</a>"</code> containing
      equipment-generated variables (inherited indirectly from
      <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>). See
      <code><a href="https://rdrr.io/pkg/Biobase/man/protocolData.html" class="external-link">protocolData</a></code> and the <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>
      class for more details.</p></dd>


    <dt><code>.__classVersion__</code>:</dt>
<dd><p>Object of class
      <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.Versions.html" class="external-link">Versions</a>"</code> describing the versions of R,
      the Biobase package, <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>,
      <code>"<a href="pSet-class.html">pSet</a>"</code> and <code>MSnSet</code> of the
      current instance. Intended for developer use and debugging (inherited
      indirectly from <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>).</p></dd>

  
</dl></div>
    <div id="extends">
    <h2>Extends</h2>
    <p>Class <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>, directly.
  Class <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.VersionedBiobase.html" class="external-link">VersionedBiobase</a>"</code>, by class "eSet", distance 2.
  Class <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.Versioned.html" class="external-link">Versioned</a>"</code>, by class "eSet", distance 3.</p>
    </div>
    <div id="methods">
    <h2>Methods</h2>
    <p>MSnSet specific methods or over-riding it's super-class are described
  below. See also more <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code> for
  inherited methods.</p>
<p></p><dl><dt>acquisitionNum</dt>
<dd><p><code>acquisitionNum(signature(object = "MSnSet"))</code>: Returns the
      a numeric vector with acquisition number of each spectrum. The vector
      names are the corresponding spectrum names.
      The information is extracted from the object's <code>featureData</code> slot.</p></dd>


    <dt>fromFile</dt>
<dd><p><code>fromFile(signature(object = "MSnSet"))</code>: get the index of
      the file (in <code>fileNames(object)</code>) from which the raw
      spectra from which the corresponding feature were originally
      read. The relevant information is extracted from the object's
      <code>featureData</code> slot.</p>
<p>Returns a numeric vector with names corresponding to the spectrum
      names.</p></dd>


    <dt>dim</dt>
<dd><p><code>signature(x = "MSnSet")</code>: Returns the dimensions of
      object's assay data, i.e the number of samples and the number of
      features.</p></dd>


    <dt>fileNames</dt>
<dd><p><code>signature(object = "MSnSet")</code>: Access file
      names in the <code>processingData</code> slot.</p></dd>


    <dt>msInfo</dt>
<dd><p><code>signature(object = "MSnSet")</code>: Prints the MIAPE-MS
      meta-data stored in the <code>experimentData</code> slot.</p></dd>


    <dt>processingData</dt>
<dd><p><code>signature(object = "MSnSet")</code>: Access the
      <code>processingData</code> slot.</p></dd>


    <dt>show</dt>
<dd><p><code>signature(object = "MSnSet")</code>: Displays object
      content as text.</p></dd>


    <dt>qual</dt>
<dd><p><code>signature(object = "MSnSet")</code>: Access the reporter
      ion peaks description.</p></dd>


    <dt>purityCorrect</dt>
<dd><p><code>signature(object = "MSnSet", impurities =
	"matrix")</code>: performs reporter ions purity correction. See
      <code><a href="purityCorrect-methods.html">purityCorrect</a></code> documentation for more details.</p></dd>


    <dt>normalise</dt>
<dd><p><code>signature(object = "MSnSet")</code>: Performs
      <code>MSnSet</code> normalisation. See <code><a href="normalise-methods.html">normalise</a></code> for more
      details.</p></dd>


    <dt>t</dt>
<dd><p><code>signature(x = "MSnSet")</code>: Returns a transposed
      <code>MSnSet</code> object where features are now aligned along columns
      and samples along rows and the <code>phenoData</code> and
      <code>featureData</code> slots have been swapped. The
      <code>protocolData</code> slot is always dropped.</p></dd>


    <dt>as(,"ExpressionSet")</dt>
<dd><p><code>signature(x = "MSnSet")</code>: Coerce
      object from <code>MSnSet</code> to
      <code><a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html" class="external-link">ExpressionSet-class</a></code>. The <code>experimentData</code> slot
      is converted to a <code>MIAME</code> instance. It is also possible to
      coerce an <code>ExpressionSet</code> to and <code>MSnSet</code>, in which case
      the <code>experimentData</code> slot is newly initialised.</p></dd>


    <dt>as(,"IBSpectra")</dt>
<dd><p><code>signature(x = "MSnSet")</code>: Coerce
      object from <code>MSnSet</code> to <code>IBSpectra</code> from the
      <code>isobar</code> package.</p></dd>


    <dt>as(,"data.frame")</dt>
<dd><p><code>signature(x = "MSnSet")</code>: Coerce
      object from <code>MSnSet</code> to <code>data.frame</code>. The <code>MSnSet</code>
      is transposed and the <code>PhenoData</code> slot is appended.</p></dd>


    <dt>as(,"SummarizedExperiment")</dt>
<dd><p><code>signature(x = "MSnSet")</code>:
      Coerce object from <code>MSnSet</code> to
      <code>SummarizedExperiment</code>. Only part of the metadata is
      retained. See <code>addMSnSetMetadata</code> and the example below for
      details.</p></dd>


    <dt>write.exprs</dt>
<dd><p><code>signature(x = "MSnSet")</code>: Writes expression values
      to a tab-separated file (default is <code>tmp.txt</code>). The
      <code>fDataCols</code> parameter can be used to specify which
      <code>featureData</code> columns (as column names, column number or
      <code>logical</code>) to append on the right of the expression matrix.
      The following arguments are the same as <code>write.table</code>.</p></dd>


    <dt>combine</dt>
<dd><p><code>signature(x = "MSnSet", y = "MSnSet", ...)</code>: Combines
      2 or more <code>MSnSet</code> instances according to their feature names.
      Note that the <code>qual</code> slot and the processing information are
      silently dropped.</p></dd>


    <dt>topN</dt>
<dd><p><code>signature(object = "MSnSet", groupBy, n = 3, fun,
	..., verbose = isMSnbaseVerbose())</code>:
      Selects the <code>n</code> most intense features (typically peptides or
      spectra) out of all available for each set defined by
      <code>groupBy</code> (typically proteins) and creates a new instance of
      class <code>MSnSet</code>. If less than <code>n</code> features are available,
      all are selected. The <code>ncol(object)</code> features are summerised
      using <code>fun</code> (default is <code>sum</code>) prior to be ordered in
      decreasing order. Additional parameters can be passed to
      <code>fun</code> through <code>...</code>, for instance to control the
      behaviour of <code>topN</code> in case of <code>NA</code> values.
      (Works also with <code>matrix</code> instances.)</p>
<p>See also the <code><a href="nQuants.html">nQuants</a></code> function to retrieve the
      actual number of retained peptides out of <code>n</code>.</p>
<p>A complete use case using <code>topN</code> and <code>nQuants</code> is
      detailed in the <code>synapter</code> package vignette.</p></dd>


    <dt>filterNA</dt>
<dd><p><code>signature(object = "MSnSet", pNA = "numeric",
      pattern = "character", droplevels = "logical")</code>: This method
      subsets <code>object</code> by removing features that have (strictly)
      more than <code>pNA</code> percent of NA values. Default <code>pNA</code> is
      0, which removes any feature that exhibits missing data.
      The method can also be used with a character pattern composed of
      <code>0</code> or <code>1</code> characters only. A <code>0</code> represent a
      column/sample that is allowed a missing values, while
      columns/samples with and <code>1</code> must not have <code>NA</code>s.</p>
<p>This method also accepts <code>matrix</code>
      instances. <code>droplevels</code> defines whether unused levels in the
      feature meta-data ought to be lost. Default is <code>TRUE</code>.
      See the <code>droplevels</code> method below.</p>
<p>See also the <code><a href="plotNA-methods.html">is.na.MSnSet</a></code> and <code><a href="plotNA-methods.html">plotNA</a></code>
      methods for missing data exploration.</p></dd>


    <dt>filterZero</dt>
<dd><p><code>signature(object = "MSnSet", pNA = "numeric",
      pattern = "character", droplevels = "logical")</code>: As
      <code>filterNA</code>, but for zeros.</p></dd>


    <dt>filterMsLevel</dt>
<dd><p><code>signature(object = "MSnSet", msLevel. =
      "numeric", fcol = "character")</code> Keeps only spectra with level
      <code>msLevel.</code>, as defined by the <code>fcol</code> feature variable
      (default is <code>"msLevel"</code>).</p></dd>


    <dt>log</dt>
<dd><p><code>signature(object = "MSnSet", base = "numeric")</code> Log
      transforms <code>exprs(object)</code> using
      <code><a href="https://rdrr.io/r/base/Log.html" class="external-link">base::log</a></code>. <code>base</code> (defaults is <code>e='exp(1)'</code>) must
      be a positive or complex number, the base with respect to which
      logarithms are computed.</p></dd>


    <dt>droplevels</dt>
<dd><p><code>signature(x = "MSnSet", ...)</code>Drops the unused
      factor levels in the <code>featureData</code> slot. See
      <code><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></code> for details.</p></dd>


    <dt>impute</dt>
<dd><p><code>signature(object = "MSnSet", ...)</code>
      Performs data imputation on the <code>MSnSet</code> object.
      See <code>impute</code> for more details.</p></dd>


    <dt>trimws</dt>
<dd><p><code>signature(object = "MSnSet", ...)</code>Trim leading and/or
      trailing white spaces in the feature data slot. Also available for
      <code>data.frame</code> objects. See <code>?base::<a href="https://rdrr.io/r/base/trimws.html" class="external-link">trimws</a></code>
      for details.</p></dd>


  
</dl><p>Additional accessors for the experimental metadata
  (<code>experimentData</code> slot) are defined. See
  <code>"<a href="MIAPE-class.html">MIAPE</a>"</code> for details.</p>
    </div>
    <div id="plotting">
    <h2>Plotting</h2>
    <p></p><dl><dt>meanSdPlot</dt>
<dd><p><code>signature(object = "MSnSet")</code> Plots row
      standard deviations versus row means. See
      <code>meanSdPlot</code> (<code>vsn</code> package) for more details.</p></dd>


    <dt>image</dt>
<dd><p><code>signature(x = "MSnSet", facetBy = "character",
	sOrderBy = "character", legend = "character", low = "character",
	high = "character", fnames = "logical", nmax =
	"numeric")</code> Produces an heatmap of expression values in the
	<code>x</code> object. Simple horizontal facetting is enabled by
	passing a single character as <code>facetBy</code>. Arbitrary
	facetting can be performed manually by saving the return value
	of the method (see example below). Re-ordering of the samples is
	possible by providing the name of a phenotypic variable to
	<code>sOrderBy</code>. The title of the legend can be set with
	<code>legend</code> and the colours with the <code>low</code> and
	<code>high</code> arguments. If any negative value is detected in the
	data, the values are considered as log fold-changes and a
	divergent colour scale is used. Otherwise, a gradient from low
	to high is used. To scale the quantitative data in <code>x</code>
	prior to plotting, please see the <code>scale</code> method.</p>
<p>When there are more than <code>nmax</code> (default is 50)
	features/rows, these are not printed. This behaviour can be
	controlled by setting <code>fnames</code> to <code>TRUE</code> (always
	print) or <code>FALSE</code> (never print). See examples below.</p>
<p>The code is based on Vlad Petyuk's
	<code>vp.misc::image_msnset</code>. The previous version of this
	method is still available through the <code>image2</code> function.</p></dd>


    <dt>plotNA</dt>
<dd><p><code>signature(object = "MSnSet", pNA =
	"numeric")</code>
	Plots missing data for an <code>MSnSet</code> instance. <code>pNA</code> is a
	<code>numeric</code> of length 1 that specifies the percentage
	of accepted missing data values per features. This value will be
	highlighted with a point on the figure, illustrating the overall
	percentage of NA values in the full data set and the number of
	proteins retained. Default is 1/2. See also
	<code><a href="plotNA-methods.html">plotNA</a></code>.</p></dd>


    <dt>MAplot</dt>
<dd><p><code>signature(object = "MSnSet", log.it = "logical",
	base = "numeric", ...)</code>
      Produces MA plots (Ratio as a function
      of average intensity) for the samples in <code>object</code>. If
      <code>ncol(object) == 2</code>, then one MA plot is produced using the
      <code>ma.plot</code> function from the <code>affy</code> package. If
      <code>object</code> has more than 2 columns, then
      <code>mva.pairs</code>. <code>log.it</code> specifies is the data
      should be log-transformed (default is <code>TRUE</code>) using
      <code>base</code>. Further <code>...</code> arguments will be passed to the
      respective functions.</p></dd>


    <dt>addIdentificationData</dt>
<dd><p><code>signature(object = "MSnSet", ...)</code>:
      Adds identification data to a <code>MSnSet</code> instance.
      See <code><a href="addIdentificationData-methods.html">addIdentificationData</a></code> documentation for
      more details and examples.</p></dd>


    <dt>removeNoId</dt>
<dd><p><code>signature(object = "MSnSet", fcol =
	"pepseq", keep = NULL)</code>: Removes non-identified features. See
      <code><a href="removeNoId-methods.html">removeNoId</a></code> documentation for more details and
      examples.</p></dd>


    <dt>removeMultipleAssignment</dt>
<dd><p><code>signature(object = "MSnSet",
	fcol = "nprot")</code>: Removes protein groups (or feature belong to
	protein groups) with more than one member. The latter is defined
	by extracting a feature variable (default is
	<code>"nprot"</code>). Also removes non-identified features.</p></dd>


    <dt>idSummary</dt>
<dd><p><code>signature(object = "MSnSet", ...)</code>: Prints a
      summary that lists the percentage of identified features per file
      (called <code>coverage</code>).</p></dd>


  
</dl></div>
    <div id="functions">
    <h2>Functions</h2>
    <p></p><dl><dt>updateFvarLabels</dt>
<dd><p><code>signature(object, label, sep)</code> This
      function updates <code>object</code>'s featureData variable labels by
      appending <code>label</code>. By default, <code>label</code> is the variable
      name and the separator <code>sep</code> is <code>.</code>.</p></dd>


    <dt>updateSampleNames</dt>
<dd><p><code>signature(object, label, sep)</code> This
      function updates <code>object</code>'s sample names by appending
      <code>label</code>. By default, <code>label</code> is the variable name and
      the separator <code>sep</code> is <code>.</code>.</p></dd>


    <dt>updateFeatureNames</dt>
<dd><p><code>signature(object, label, sep)</code> This
      function updates <code>object</code>'s feature names by appending
      <code>label</code>. By default, <code>label</code> is the variable name and
      the separator <code>sep</code> is <code>.</code>.</p></dd>


    <dt>ms2df</dt>
<dd><p><code>signature(x, fcols)</code> Coerces the <code>MSnSet</code> instance
      to a <code>data.frame</code>. The direction of the data is retained and
      the feature variable labels that match <code>fcol</code> are appended to
      the expression values. See also <code>as(x, "data.frame")</code> above.</p></dd>


    <dt>addMSnSetMetadata</dt>
<dd><p><code>signature(x, y)</code> When coercing an
      <code>MSnSet</code> <code>y</code> to a <code>SummarizedExperiment</code> <code>x</code>
      with <code>x &lt;- as(y, "SummarizedExperiment")</code>, most of <code>y</code>'s
      metadata is lost. Only the file names, the processing log and the
      MSnbase version from the <code>processingData</code> slots are passed
      along. The <code>addMSnSetMetadata</code> function can be used to add
      the complete <code>processingData</code>, <code>experimentData</code> and
      <code>protocolData</code> slots. The downside of this is that MSnbase is
      now required to use the <code>SummarizedExperiment</code> object.</p></dd>


  
</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>Laurent Gatto &lt;lg390@cam.ac.uk&gt;</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index">

<p><code>"<a href="https://rdrr.io/pkg/Biobase/man/class.eSet.html" class="external-link">eSet</a>"</code>, <code>"<a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html" class="external-link">ExpressionSet</a>"</code> and
  <code><a href="quantify-methods.html">quantify</a></code>. <code>MSnSet</code> quantitation values and
  annotation can be exported to a file with
  <code>write.exprs</code>. See <code><a href="readMSnSet.html">readMSnSet</a></code> to
  create and <code>MSnSet</code> using data available in a spreadsheet or
  <code>data.frame</code>.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">msnset</span> <span class="op">&lt;-</span> <span class="va">msnset</span><span class="op">[</span><span class="fl">10</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">exprs</span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span></span>
<span class="r-in"><span><span class="fu">exprs</span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X18       TRUE      FALSE      FALSE       TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X19       TRUE       TRUE      FALSE      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X2       FALSE      FALSE      FALSE      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X20      FALSE      FALSE      FALSE      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X21      FALSE      FALSE      FALSE      FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X22      FALSE      FALSE      FALSE      FALSE</span>
<span class="r-in"><span><span class="fu">featureNames</span><span class="op">(</span><span class="fu">filterNA</span><span class="op">(</span><span class="va">msnset</span>, pNA <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "X2"  "X20" "X21" "X22"</span>
<span class="r-in"><span><span class="fu">featureNames</span><span class="op">(</span><span class="fu">filterNA</span><span class="op">(</span><span class="va">msnset</span>, pattern <span class="op">=</span> <span class="st">"0110"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "X18" "X2"  "X20" "X21" "X22"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">12</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>otherpdata <span class="op">=</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>otherfdata <span class="op">=</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x0</span> <span class="op">&lt;-</span> <span class="fu">MSnSet</span><span class="op">(</span><span class="va">M</span>, <span class="va">fd</span>, <span class="va">pd</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">sampleNames</span><span class="op">(</span><span class="va">x0</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1" "2" "3"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">12</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>otherpdata <span class="op">=</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>otherfdata <span class="op">=</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">MSnSet</span><span class="op">(</span><span class="va">M</span>, <span class="va">fd</span>, <span class="va">pd</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">sampleNames</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "A" "B" "C"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Visualisation</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/pRolocdata" class="external-link">"pRolocdata"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dunkley2006</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">## Changing colours</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span>, high <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span>, high <span class="op">=</span> <span class="st">"darkgreen"</span>, low <span class="op">=</span> <span class="st">"yellow"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">## Forcing feature names</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span>, fnames <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">## Facetting</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span>, facetBy <span class="op">=</span> <span class="st">"replicate"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span> <span class="co">## for facet_grid</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">replicate</span> <span class="op">~</span> <span class="va">membrane.prep</span>, scales <span class="op">=</span> <span class="st">'free'</span>, space <span class="op">=</span> <span class="st">'free'</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-7.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">markers</span> <span class="op">~</span> <span class="va">replicate</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-8.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">## Fold-changes</span></span></span>
<span class="r-in"><span><span class="va">dd</span> <span class="op">&lt;-</span> <span class="va">dunkley2006</span></span></span>
<span class="r-in"><span><span class="fu">exprs</span><span class="op">(</span><span class="va">dd</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">exprs</span><span class="op">(</span><span class="va">dd</span><span class="op">)</span> <span class="op">-</span> <span class="fl">0.25</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dd</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-9.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dd</span>, low <span class="op">=</span> <span class="st">"green"</span>, high <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-10.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">## Feature names are displayed by default for smaller data</span></span></span>
<span class="r-in"><span><span class="va">dunkley2006</span> <span class="op">&lt;-</span> <span class="va">dunkley2006</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-11.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="va">dunkley2006</span>, legend <span class="op">=</span> <span class="st">"hello"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="MSnSet-class-12.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Coercion</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">"SummarizedExperiment"</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="mzRident2dfr.html">as</a></span><span class="op">(</span><span class="va">msnset</span>, <span class="st">"SummarizedExperiment"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu">metadata</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="co">## only logging</span></span></span>
<span class="r-in"><span>    <span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">addMSnSetMetadata</span><span class="op">(</span><span class="va">se</span>, <span class="va">msnset</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu">metadata</span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="co">## all metadata</span></span></span>
<span class="r-in"><span>    <span class="va">msnset2</span> <span class="op">&lt;-</span> <span class="fu"><a href="mzRident2dfr.html">as</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"MSnSet"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="pSet-class.html">processingData</a></span><span class="op">(</span><span class="va">msnset2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - - - Processing information - - -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data loaded: Wed May 11 18:54:39 2011 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iTRAQ4 quantification by trapezoidation: Wed Apr  1 21:41:53 2015 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Subset [55,4][6,4] Tue Apr 30 17:21:37 2024 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  MSnbase version: 1.1.22 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="mzRident2dfr.html">as</a></span><span class="op">(</span><span class="va">msnset</span>, <span class="st">"ExpressionSet"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ExpressionSet (storageMode: lockedEnvironment)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assayData: 6 features, 4 samples </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   element names: exprs </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> protocolData: none</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phenoData</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   sampleNames: iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   varLabels: mz reporters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   varMetadata: labelDescription</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> featureData</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   featureNames: X18 X19 ... X22 (6 total)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   fvarLabels: spectrum ProteinAccession ... collision.energy (15 total)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   fvarMetadata: labelDescription</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> experimentData: use 'experimentData(object)'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Annotation: No annotation </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

