<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Performs reporter ions purity correction — purityCorrect-methods • MSnbase</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Performs reporter ions purity correction — purityCorrect-methods" />
<meta property="og:description" content="Manufacturers sometimes provide purity correction values indicating
  the percentages of each reporter ion that have masses differing by
  +/- n Da from the nominal reporter ion mass due to isotopic
  variants. This correction is generally applied after reporter peaks
  quantitation.
Purity correction here is applied using solve from the
  base package using the purity correction values as coefficient of
  the linear system and the reporter quantities as the right-hand side
  of the linear system. 'NA' values are ignored and negative
  intensities after correction are also set to 'NA'.
A more elaborated purity correction method is described in
  Shadforth et al., i-Tracker: for quantitative proteomics using
  iTRAQ. BMC Genomics. 2005 Oct 20;6:145. (PMID 16242023).
Function makeImpuritiesMatrix(x, filename, edit = TRUE) helps
  the user to create such a matrix. The function can be used in two ways.
  If given an integer x, it is used as the dimension of the
  square matrix (i.e the number of reporter ions). For TMT6-plex and
  iTRAQ4-plex, default values taken from manufacturer's certification
  sheets are used as templates, but batch specific values should be used
  whenever possible. Alternatively, the filename of a csv
  spreadsheet can be provided. The sheet should define the correction
  factors as illustrated below (including reporter names in the first
  column and header row) and the corresponding correction matrix is
  calculated. Examples of such csv files are available in the
  package's extdata directory. Use
  dir(system.file(&quot;extdata&quot;, package = &quot;MSnbase&quot;), pattern =
    &quot;PurityCorrection&quot;, full.names = TRUE) to locate them.
  If edit = TRUE, the the matrix can be edited before
  it is returned." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.15.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/v04-benchmarking.html">MSnbase benchmarking</a>
    </li>
    <li>
      <a href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lgatto/MSnbase/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Performs reporter ions purity correction</h1>
    
    <div class="hidden name"><code>purityCorrect-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Manufacturers sometimes provide purity correction values indicating
  the percentages of each reporter ion that have masses differing by
  +/- n Da from the nominal reporter ion mass due to isotopic
  variants. This correction is generally applied after reporter peaks
  quantitation.</p>
<p>Purity correction here is applied using <code>solve</code> from the
  <code>base</code> package using the purity correction values as coefficient of
  the linear system and the reporter quantities as the right-hand side
  of the linear system. 'NA' values are ignored and negative
  intensities after correction are also set to 'NA'.</p>
<p>A more elaborated purity correction method is described in
  Shadforth <em>et al.</em>, i-Tracker: for quantitative proteomics using
  iTRAQ. BMC Genomics. 2005 Oct 20;6:145. (PMID 16242023).</p>
<p>Function <code>makeImpuritiesMatrix(x, filename, edit = TRUE)</code> helps
  the user to create such a matrix. The function can be used in two ways.
  If given an integer <code>x</code>, it is used as the dimension of the
  square matrix (i.e the number of reporter ions). For TMT6-plex and
  iTRAQ4-plex, default values taken from manufacturer's certification
  sheets are used as templates, but batch specific values should be used
  whenever possible. Alternatively, the <code>filename</code> of a <code>csv</code>
  spreadsheet can be provided. The sheet should define the correction
  factors as illustrated below (including reporter names in the first
  column and header row) and the corresponding correction matrix is
  calculated. Examples of such <code>csv</code> files are available in the
  package's <code>extdata</code> directory. Use
  <code><a href='https://rdrr.io/r/base/list.files.html'>dir(system.file("extdata", package = "MSnbase"), pattern =
    "PurityCorrection", full.names = TRUE)</a></code> to locate them.
  If <code>edit = TRUE</code>, the the matrix can be edited before
  it is returned.</p>
    </div>


    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>An object of class <code>"<a href='MSnSet-class.html'>MSnSet</a>"</code>.</p></td>
    </tr>
    <tr>
      <th>impurities</th>
      <td><p>A square 'matrix' of dim equal to ncol(object)
    defining the correction coefficients to be applied.
    The reporter ions should be ordered along the columns
    and the relative percentages along the rows.</p>
<p>As an example, below is the correction factors as provided in an ABI
    iTRAQ 4-plex certificate of analysis:</p><table class='table'>
<tr><td>reporter</td><td>% of -2</td><td>% of -1</td><td>% of +1</td><td>% of +2</td></tr>
<tr><td>114</td><td>0.0</td><td>1.0</td><td>5.9</td><td>0.2</td></tr>
<tr><td>115</td><td>0.0</td><td>2.0</td><td>5.6</td><td>0.1</td></tr>
<tr><td>116</td><td>0.0</td><td>3.0</td><td>4.5</td><td>0.1</td></tr>
<tr><td>117</td><td>0.1</td><td>4.0</td><td>3.5</td><td>0.1</td></tr>
</table>


<p>The impurity table will be</p><table class='table'>
<tr><td>0.929</td><td>0.059</td><td>0.002</td><td>0.000</td></tr>
<tr><td>0.020</td><td>0.923</td><td>0.056</td><td>0.001</td></tr>
<tr><td>0.000</td><td>0.030</td><td>0.924</td><td>0.045</td></tr>
<tr><td>0.000</td><td>0.001</td><td>0.040</td><td>0.923</td></tr>
</table>
<p>where, the diagonal is computed as 100 - sum of rows of the
    original table and subsequent cells are directly filled in.</p>
<p>Similarly, for TMT 6-plex tags, we observe</p><table class='table'>
<tr><td>reporter</td><td>% of -3</td><td>% of -2</td><td>% of -1</td><td>% of +1 %</td><td>% of +2</td><td>% of +3</td></tr>
<tr><td>126</td><td>0</td><td>0</td><td>0</td><td>6.1</td><td>0</td><td>0</td></tr>
<tr><td>127</td><td>0</td><td>0</td><td>0.5</td><td>6.7</td><td>0</td><td>0</td></tr>
<tr><td>128</td><td>0</td><td>0</td><td>1.1</td><td>4.2</td><td>0</td><td>0</td></tr>
<tr><td>129</td><td>0</td><td>0</td><td>1.7</td><td>4.1</td><td>0</td><td>0</td></tr>
<tr><td>130</td><td>0</td><td>0</td><td>1.6</td><td>2.1</td><td>0</td><td>0</td></tr>
<tr><td>131</td><td>0</td><td>0.2</td><td>3.2</td><td>2.8</td><td>0</td><td>0</td></tr>
</table>
<p>and obtain the following impurity correction matrix</p><table class='table'>
<tr><td>0.939</td><td>0.061</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>0.005</td><td>0.928</td><td>0.067</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>0.000</td><td>0.011</td><td>0.947</td><td>0.042</td><td>0.000</td><td>0.000</td></tr>
<tr><td>0.000</td><td>0.000</td><td>0.017</td><td>0.942</td><td>0.041</td><td>0.000</td></tr>
<tr><td>0.000</td><td>0.000</td><td>0.000</td><td>0.016</td><td>0.963</td><td>0.021</td></tr>
<tr><td>0.000</td><td>0.000</td><td>0.000</td><td>0.002</td><td>0.032</td><td>0.938</td></tr>
</table>


<p>For iTRAQ 8-plex, given the following correction factors (to make
    such a matrix square, if suffices to add -4, -3, +3 and +4 columns
    filled with zeros):</p>
<p></p><table class='table'>
<tr><td>TAG</td><td>-2</td><td>-1</td><td>+1</td><td>+2</td></tr>
<tr><td>113</td><td>0</td><td>2.5</td><td>3</td><td>0.1</td></tr>
<tr><td>114</td><td>0</td><td>1</td><td>5.9</td><td>0.2</td></tr>
<tr><td>115</td><td>0</td><td>2</td><td>5.6</td><td>0.1</td></tr>
<tr><td>116</td><td>0</td><td>3</td><td>4.5</td><td>0.1</td></tr>
<tr><td>117</td><td>0.1</td><td>4</td><td>3.5</td><td>0.1</td></tr>
<tr><td>118</td><td>0.1</td><td>2</td><td>3</td><td>0.1</td></tr>
<tr><td>119</td><td>0.1</td><td>2</td><td>4</td><td>0.1</td></tr>
<tr><td>121</td><td>0.1</td><td>2</td><td>3</td><td>0.1</td></tr>
</table>


<p>we calculate the impurity correction matrix shown below</p>
<p></p><table class='table'>
<tr><td></td><td>113</td><td>114</td><td>115</td><td>116</td><td>117</td><td>118</td><td>119</td><td>121</td></tr>
<tr><td>% reporter 113</td><td>0.944</td><td>0.030</td><td>0.001</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 114</td><td>0.010</td><td>0.929</td><td>0.059</td><td>0.002</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 115</td><td>0.000</td><td>0.020</td><td>0.923</td><td>0.056</td><td>0.001</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 116</td><td>0.000</td><td>0.000</td><td>0.030</td><td>0.924</td><td>0.045</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 117</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.040</td><td>0.923</td><td>0.035</td><td>0.001</td><td>0.000</td></tr>
<tr><td>% reporter 118</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.020</td><td>0.948</td><td>0.030</td><td>0.001</td></tr>
<tr><td>% reporter 119</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.020</td><td>0.938</td><td>0.040</td></tr>
<tr><td>% reporter 121</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.020</td><td>0.948</td></tr>
</table>



<p>Finally, for a TMT 10-plex impurity matrix (for example lot <a href='https://www.thermofisher.com/document-connect/document-connect.html?url=https://assets.thermofisher.com/TFS-Assets/LSG/certificate/Certificates%20of%20Analysis/RH239932\_90309.pdf'>RH239932</a>)</p>
<p></p><table class='table'>
<tr><td>.</td><td>-2</td><td>-1</td><td>1</td><td>2</td><td>126</td><td>0.0</td><td>0.0</td><td>5.0 (127C)</td><td>0.0 (128C)</td></tr>
<tr><td>127N</td><td>0.0</td><td>0.2</td><td>5.8 (128N)</td><td>0.0 (129N)</td><td>127C</td><td>0.0</td><td>0.3 (126)</td><td>4.8 (128C)</td><td>0.0 (129C)</td></tr>
<tr><td>128N</td><td>0.0</td><td>0.4 (127N)</td><td>4.1 (129N)</td><td>0.0 (130N)</td><td>128C</td><td>0.0 (126)</td><td>0.6 (127C)</td><td>3.0 (129C)</td><td>0.0 (130C)</td></tr>
<tr><td>129N</td><td>0.0 (127N)</td><td>0.8 (128N)</td><td>3.5 (130N)</td><td>0.0 (131)</td><td>129C</td><td>0.0 (127C)</td><td>1.4 (128C)</td><td>2.4 (130C)</td><td>0.0</td></tr>
<tr><td>130N</td><td>0.1 (128N)</td><td>1.5 (129N)</td><td>2.4 (131)</td><td>3.2</td><td>130C</td><td>0.0 (128C)</td><td>1.7 (129C)</td><td>1.8</td><td>0.0</td></tr>
</table>


<p>(Note that a previous example, taken from lot
    <a href='https://www.thermofisher.com/document-connect/document-connect.html?url=https://assets.thermofisher.com/TFS-Assets/LSG/certificate/Certificates%20of%20Analysis/PB199188A\_90110.pdf'>PB199188A</a>,
    contained a typo.)</p>
<p>the impurity correction matrix is</p>
<p></p><table class='table'>
<tr><td>.</td><td>126</td><td>127N</td><td>127C</td><td>128N</td><td>128C</td><td>129N</td><td>129C</td><td>130N</td><td>130C</td><td>131</td></tr>
<tr><td>% reporter 126</td><td>0.950</td><td>0.000</td><td>0.050</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 127N</td><td>0.000</td><td>0.940</td><td>0.000</td><td>0.058</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 127C</td><td>0.003</td><td>0.000</td><td>0.949</td><td>0.000</td><td>0.048</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 128N</td><td>0.000</td><td>0.004</td><td>0.000</td><td>0.955</td><td>0.000</td><td>0.041</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 128C</td><td>0.000</td><td>0.000</td><td>0.006</td><td>0.000</td><td>0.964</td><td>0.000</td><td>0.030</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 129N</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.008</td><td>0.000</td><td>0.957</td><td>0.000</td><td>0.035</td><td>0.000</td><td>0.000</td></tr>
<tr><td>% reporter 129C</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.014</td><td>0.000</td><td>0.962</td><td>0.000</td><td>0.024</td><td>0.000</td></tr>
<tr><td>% reporter 130N</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.000</td><td>0.015</td><td>0.000</td><td>0.928</td><td>0.000</td><td>0.024</td></tr>
<tr><td>% reporter 130C</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.017</td><td>0.000</td><td>0.965</td><td>0.000</td></tr>
<tr><td>% reporter 131</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.002</td><td>0.000</td><td>0.020</td><td>0.000</td><td>0.956</td></tr>
</table>



<p>These examples are provided as defaults impurity correction matrices
    in <code>makeImpuritiesMatrix</code>.</p>
<p></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    <p></p><dl'>
    <dt><code>signature(object = "MSnSet", impurities = "matrix")</code></dt><dd><p></p></dd>
  
</dl>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## quantifying full experiment</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>msnset</span>)
<span class='no'>impurities</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>0.929</span>,<span class='fl'>0.059</span>,<span class='fl'>0.002</span>,<span class='fl'>0.000</span>,
                       <span class='fl'>0.020</span>,<span class='fl'>0.923</span>,<span class='fl'>0.056</span>,<span class='fl'>0.001</span>,
                       <span class='fl'>0.000</span>,<span class='fl'>0.030</span>,<span class='fl'>0.924</span>,<span class='fl'>0.045</span>,
                       <span class='fl'>0.000</span>,<span class='fl'>0.001</span>,<span class='fl'>0.040</span>,<span class='fl'>0.923</span>),
                     <span class='kw'>nrow</span><span class='kw'>=</span><span class='fl'>4</span>, <span class='kw'>byrow</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'>## or, using makeImpuritiesMatrix()</span>
<span class='kw'>if</span> (<span class='fl'>FALSE</span>) <span class='no'>impurities</span> <span class='kw'>&lt;-</span> <span class='fu'>makeImpuritiesMatrix</span>(<span class='fl'>4</span>)
<span class='no'>msnset.crct</span> <span class='kw'>&lt;-</span> <span class='fu'>purityCorrect</span>(<span class='no'>msnset</span>, <span class='no'>impurities</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='fu'>exprs</span>(<span class='no'>msnset</span>))</div><div class='output co'>#&gt;     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117
#&gt; X1   1347.6158  2247.3097  3927.6931  7661.1463
#&gt; X10   739.9861   799.3501   712.5983   940.6793
#&gt; X11 27638.3582 33394.0252 32104.2879 26628.7278
#&gt; X12 31892.8928 33634.6980 37674.7272 37227.7119
#&gt; X13 26143.7542 29677.4781 29089.0593 27902.5608
#&gt; X14  6448.0829  6234.1957  6902.8903  6437.2303</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='fu'>exprs</span>(<span class='no'>msnset.crct</span>))</div><div class='output co'>#&gt;     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117
#&gt; X1   1402.9442  2214.0346  3762.2549  8114.4429
#&gt; X10   779.4666   793.0792   678.8083   985.2003
#&gt; X11 29034.3781 33271.0470 31484.7131 27279.1383
#&gt; X12 33618.9092 33046.3075 37031.6133 38492.1376
#&gt; X13 27508.0038 29440.9296 28390.4561 28814.2463
#&gt; X14  6809.7600  6090.7894  6799.5030  6636.1450</div><div class='input'><span class='fu'><a href='pSet-class.html'>processingData</a></span>(<span class='no'>msnset.crct</span>)</div><div class='output co'>#&gt; - - - Processing information - - -
#&gt; Data loaded: Wed May 11 18:54:39 2011 
#&gt; iTRAQ4 quantification by trapezoidation: Wed Apr  1 21:41:53 2015 
#&gt; Purity corrected: Fri May 15 14:10:29 2020 
#&gt;  MSnbase version: 1.1.22 </div><div class='input'>
<span class='co'>## default impurity matrix for iTRAQ 8-plex</span>
<span class='fu'>makeImpuritiesMatrix</span>(<span class='fl'>8</span>, <span class='kw'>edit</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt;                  113   114   115   116   117   118   119   121
#&gt; % reporter 113 0.944 0.030 0.001 0.000 0.000 0.000 0.000 0.000
#&gt; % reporter 114 0.010 0.929 0.059 0.002 0.000 0.000 0.000 0.000
#&gt; % reporter 115 0.000 0.020 0.923 0.056 0.001 0.000 0.000 0.000
#&gt; % reporter 116 0.000 0.000 0.030 0.924 0.045 0.001 0.000 0.000
#&gt; % reporter 117 0.000 0.000 0.001 0.040 0.923 0.035 0.001 0.000
#&gt; % reporter 118 0.000 0.000 0.000 0.001 0.020 0.948 0.030 0.001
#&gt; % reporter 119 0.000 0.000 0.000 0.000 0.001 0.020 0.938 0.040
#&gt; % reporter 121 0.000 0.000 0.000 0.000 0.000 0.001 0.020 0.948</div><div class='input'>
<span class='co'>## default impurity matrix for TMT 10-plex</span>
<span class='fu'>makeImpuritiesMatrix</span>(<span class='fl'>10</span>, <span class='kw'>edit</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt;                   126  127N  127C  128N  128C  129N  129C  130N  130C   131
#&gt; % reporter 126  0.950 0.000 0.050 0.000 0.000 0.000 0.000 0.000 0.000 0.000
#&gt; % reporter 127N 0.000 0.940 0.000 0.058 0.000 0.000 0.000 0.000 0.000 0.000
#&gt; % reporter 127C 0.003 0.000 0.949 0.000 0.048 0.000 0.000 0.000 0.000 0.000
#&gt; % reporter 128N 0.000 0.004 0.000 0.955 0.000 0.041 0.000 0.000 0.000 0.000
#&gt; % reporter 128C 0.000 0.000 0.006 0.000 0.964 0.000 0.030 0.000 0.000 0.000
#&gt; % reporter 129N 0.000 0.000 0.000 0.008 0.000 0.957 0.000 0.035 0.000 0.000
#&gt; % reporter 129C 0.000 0.000 0.000 0.000 0.014 0.000 0.962 0.000 0.024 0.000
#&gt; % reporter 130N 0.000 0.000 0.000 0.001 0.000 0.015 0.000 0.928 0.000 0.024
#&gt; % reporter 130C 0.000 0.000 0.000 0.000 0.000 0.000 0.017 0.000 0.965 0.000
#&gt; % reporter 131  0.000 0.000 0.000 0.000 0.000 0.002 0.000 0.020 0.000 0.956</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


