<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Performs reporter ions purity correction — purityCorrect-methods • MSnbase</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Performs reporter ions purity correction — purityCorrect-methods"><meta name="description" content="Manufacturers sometimes provide purity correction values indicating
  the percentages of each reporter ion that have masses differing by
  +/- n Da from the nominal reporter ion mass due to isotopic
  variants. This correction is generally applied after reporter peaks
  quantitation.
Purity correction here is applied using solve from the
  base package using the purity correction values as coefficient of
  the linear system and the reporter quantities as the right-hand side
  of the linear system. 'NA' values are ignored and negative
  intensities after correction are also set to 'NA'.
A more elaborated purity correction method is described in
  Shadforth et al., i-Tracker: for quantitative proteomics using
  iTRAQ. BMC Genomics. 2005 Oct 20;6:145. (PMID 16242023).
Function makeImpuritiesMatrix(x, filename, edit = TRUE) helps
  the user to create such a matrix. The function can be used in two ways.
  If given an integer x, it is used as the dimension of the
  square matrix (i.e the number of reporter ions). For TMT6-plex and
  iTRAQ4-plex, default values taken from manufacturer's certification
  sheets are used as templates, but batch specific values should be used
  whenever possible. Alternatively, the filename of a csv
  spreadsheet can be provided. The sheet should define the correction
  factors as illustrated below (including reporter names in the first
  column and header row) and the corresponding correction matrix is
  calculated. Examples of such csv files are available in the
  package's extdata directory. Use
  dir(system.file(&quot;extdata&quot;, package = &quot;MSnbase&quot;), pattern =
    &quot;PurityCorrection&quot;, full.names = TRUE) to locate them.
  If edit = TRUE, the the matrix can be edited before
  it is returned."><meta property="og:description" content="Manufacturers sometimes provide purity correction values indicating
  the percentages of each reporter ion that have masses differing by
  +/- n Da from the nominal reporter ion mass due to isotopic
  variants. This correction is generally applied after reporter peaks
  quantitation.
Purity correction here is applied using solve from the
  base package using the purity correction values as coefficient of
  the linear system and the reporter quantities as the right-hand side
  of the linear system. 'NA' values are ignored and negative
  intensities after correction are also set to 'NA'.
A more elaborated purity correction method is described in
  Shadforth et al., i-Tracker: for quantitative proteomics using
  iTRAQ. BMC Genomics. 2005 Oct 20;6:145. (PMID 16242023).
Function makeImpuritiesMatrix(x, filename, edit = TRUE) helps
  the user to create such a matrix. The function can be used in two ways.
  If given an integer x, it is used as the dimension of the
  square matrix (i.e the number of reporter ions). For TMT6-plex and
  iTRAQ4-plex, default values taken from manufacturer's certification
  sheets are used as templates, but batch specific values should be used
  whenever possible. Alternatively, the filename of a csv
  spreadsheet can be provided. The sheet should define the correction
  factors as illustrated below (including reporter names in the first
  column and header row) and the corresponding correction matrix is
  calculated. Examples of such csv files are available in the
  package's extdata directory. Use
  dir(system.file(&quot;extdata&quot;, package = &quot;MSnbase&quot;), pattern =
    &quot;PurityCorrection&quot;, full.names = TRUE) to locate them.
  If edit = TRUE, the the matrix can be edited before
  it is returned."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MSnbase</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.35.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a></li>
    <li><a class="dropdown-item" href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a></li>
    <li><a class="dropdown-item" href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a></li>
    <li><a class="dropdown-item" href="../articles/v04-benchmarking.html">MSnbase benchmarking</a></li>
    <li><a class="dropdown-item" href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/lgatto/MSnbase/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Performs reporter ions purity correction</h1>

      <div class="d-none name"><code>purityCorrect-methods.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Manufacturers sometimes provide purity correction values indicating
  the percentages of each reporter ion that have masses differing by
  +/- n Da from the nominal reporter ion mass due to isotopic
  variants. This correction is generally applied after reporter peaks
  quantitation.</p>
<p>Purity correction here is applied using <code>solve</code> from the
  <code>base</code> package using the purity correction values as coefficient of
  the linear system and the reporter quantities as the right-hand side
  of the linear system. 'NA' values are ignored and negative
  intensities after correction are also set to 'NA'.</p>
<p>A more elaborated purity correction method is described in
  Shadforth <em>et al.</em>, i-Tracker: for quantitative proteomics using
  iTRAQ. BMC Genomics. 2005 Oct 20;6:145. (PMID 16242023).</p>
<p>Function <code>makeImpuritiesMatrix(x, filename, edit = TRUE)</code> helps
  the user to create such a matrix. The function can be used in two ways.
  If given an integer <code>x</code>, it is used as the dimension of the
  square matrix (i.e the number of reporter ions). For TMT6-plex and
  iTRAQ4-plex, default values taken from manufacturer's certification
  sheets are used as templates, but batch specific values should be used
  whenever possible. Alternatively, the <code>filename</code> of a <code>csv</code>
  spreadsheet can be provided. The sheet should define the correction
  factors as illustrated below (including reporter names in the first
  column and header row) and the corresponding correction matrix is
  calculated. Examples of such <code>csv</code> files are available in the
  package's <code>extdata</code> directory. Use
  <code>dir(system.file("extdata", package = "MSnbase"), pattern =
    "PurityCorrection", full.names = TRUE)</code> to locate them.
  If <code>edit = TRUE</code>, the the matrix can be edited before
  it is returned.</p>
    </div>


    <div class="section level2">
    <h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a></h2>
    <p></p><dl><dt><code>signature(object = "MSnSet", impurities = "matrix")</code></dt>
<dd><p></p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <p></p>
<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>An object of class <code>"<a href="MSnSet-class.html">MSnSet</a>"</code>.</p></dd>

  <dt id="arg-impurities">impurities<a class="anchor" aria-label="anchor" href="#arg-impurities"></a></dt>
<dd><p>A square 'matrix' of dim equal to ncol(object)
    defining the correction coefficients to be applied.
    The reporter ions should be ordered along the columns
    and the relative percentages along the rows.</p>
<p>As an example, below is the correction factors as provided in an ABI
    iTRAQ 4-plex certificate of analysis:</p><table class="table table"><tr><td>reporter</td><td>% of -2</td><td>% of -1</td><td>% of +1</td><td>% of +2</td></tr><tr><td>114</td><td>0.0</td><td>1.0</td><td>5.9</td><td>0.2</td></tr><tr><td>115</td><td>0.0</td><td>2.0</td><td>5.6</td><td>0.1</td></tr><tr><td>116</td><td>0.0</td><td>3.0</td><td>4.5</td><td>0.1</td></tr><tr><td>117</td><td>0.1</td><td>4.0</td><td>3.5</td><td>0.1</td></tr></table><p>The impurity table will be</p><table class="table table"><tr><td>0.929</td><td>0.059</td><td>0.002</td><td>0.000</td></tr><tr><td>0.020</td><td>0.923</td><td>0.056</td><td>0.001</td></tr><tr><td>0.000</td><td>0.030</td><td>0.924</td><td>0.045</td></tr><tr><td>0.000</td><td>0.001</td><td>0.040</td><td>0.923</td></tr></table><p>where, the diagonal is computed as 100 - sum of rows of the
    original table and subsequent cells are directly filled in.</p>
<p>Similarly, for TMT 6-plex tags, we observe</p><table class="table table"><tr><td>reporter</td><td>% of -3</td><td>% of -2</td><td>% of -1</td><td>% of +1 %</td><td>% of +2</td><td>% of +3</td></tr><tr><td>126</td><td>0</td><td>0</td><td>0</td><td>6.1</td><td>0</td><td>0</td></tr><tr><td>127</td><td>0</td><td>0</td><td>0.5</td><td>6.7</td><td>0</td><td>0</td></tr><tr><td>128</td><td>0</td><td>0</td><td>1.1</td><td>4.2</td><td>0</td><td>0</td></tr><tr><td>129</td><td>0</td><td>0</td><td>1.7</td><td>4.1</td><td>0</td><td>0</td></tr><tr><td>130</td><td>0</td><td>0</td><td>1.6</td><td>2.1</td><td>0</td><td>0</td></tr><tr><td>131</td><td>0</td><td>0.2</td><td>3.2</td><td>2.8</td><td>0</td><td>0</td></tr></table><p>and obtain the following impurity correction matrix</p><table class="table table"><tr><td>0.939</td><td>0.061</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>0.005</td><td>0.928</td><td>0.067</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>0.000</td><td>0.011</td><td>0.947</td><td>0.042</td><td>0.000</td><td>0.000</td></tr><tr><td>0.000</td><td>0.000</td><td>0.017</td><td>0.942</td><td>0.041</td><td>0.000</td></tr><tr><td>0.000</td><td>0.000</td><td>0.000</td><td>0.016</td><td>0.963</td><td>0.021</td></tr><tr><td>0.000</td><td>0.000</td><td>0.000</td><td>0.002</td><td>0.032</td><td>0.938</td></tr></table><p>For iTRAQ 8-plex, given the following correction factors (to make
    such a matrix square, if suffices to add -4, -3, +3 and +4 columns
    filled with zeros):</p>
<p></p><table class="table table"><tr><td>TAG</td><td>-2</td><td>-1</td><td>+1</td><td>+2</td></tr><tr><td>113</td><td>0</td><td>2.5</td><td>3</td><td>0.1</td></tr><tr><td>114</td><td>0</td><td>1</td><td>5.9</td><td>0.2</td></tr><tr><td>115</td><td>0</td><td>2</td><td>5.6</td><td>0.1</td></tr><tr><td>116</td><td>0</td><td>3</td><td>4.5</td><td>0.1</td></tr><tr><td>117</td><td>0.1</td><td>4</td><td>3.5</td><td>0.1</td></tr><tr><td>118</td><td>0.1</td><td>2</td><td>3</td><td>0.1</td></tr><tr><td>119</td><td>0.1</td><td>2</td><td>4</td><td>0.1</td></tr><tr><td>121</td><td>0.1</td><td>2</td><td>3</td><td>0.1</td></tr></table><p>we calculate the impurity correction matrix shown below</p>
<p></p><table class="table table"><tr><td></td><td>113</td><td>114</td><td>115</td><td>116</td><td>117</td><td>118</td><td>119</td><td>121</td></tr><tr><td>% reporter 113</td><td>0.944</td><td>0.030</td><td>0.001</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 114</td><td>0.010</td><td>0.929</td><td>0.059</td><td>0.002</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 115</td><td>0.000</td><td>0.020</td><td>0.923</td><td>0.056</td><td>0.001</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 116</td><td>0.000</td><td>0.000</td><td>0.030</td><td>0.924</td><td>0.045</td><td>0.001</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 117</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.040</td><td>0.923</td><td>0.035</td><td>0.001</td><td>0.000</td></tr><tr><td>% reporter 118</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.020</td><td>0.948</td><td>0.030</td><td>0.001</td></tr><tr><td>% reporter 119</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.020</td><td>0.938</td><td>0.040</td></tr><tr><td>% reporter 121</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.020</td><td>0.948</td></tr></table><p>Finally, for a TMT 10-plex impurity matrix (for example lot <a href="https://www.thermofisher.com/document-connect/document-connect.html?url=https://assets.thermofisher.com/TFS-Assets/LSG/certificate/Certificates%20of%20Analysis/RH239932%5C_90309.pdf" class="external-link">RH239932</a>)</p>
<p></p><table class="table table"><tr><td>.</td><td>-2</td><td>-1</td><td>1</td><td>2</td><td>126</td><td>0.0</td><td>0.0</td><td>5.0 (127C)</td><td>0.0 (128C)</td></tr><tr><td>127N</td><td>0.0</td><td>0.2</td><td>5.8 (128N)</td><td>0.0 (129N)</td><td>127C</td><td>0.0</td><td>0.3 (126)</td><td>4.8 (128C)</td><td>0.0 (129C)</td></tr><tr><td>128N</td><td>0.0</td><td>0.4 (127N)</td><td>4.1 (129N)</td><td>0.0 (130N)</td><td>128C</td><td>0.0 (126)</td><td>0.6 (127C)</td><td>3.0 (129C)</td><td>0.0 (130C)</td></tr><tr><td>129N</td><td>0.0 (127N)</td><td>0.8 (128N)</td><td>3.5 (130N)</td><td>0.0 (131)</td><td>129C</td><td>0.0 (127C)</td><td>1.4 (128C)</td><td>2.4 (130C)</td><td>0.0</td></tr><tr><td>130N</td><td>0.1 (128N)</td><td>1.5 (129N)</td><td>2.4 (131)</td><td>3.2</td><td>130C</td><td>0.0 (128C)</td><td>1.7 (129C)</td><td>1.8</td><td>0.0</td></tr></table><p>(Note that a previous example, taken from lot
    <a href="https://www.thermofisher.com/document-connect/document-connect.html?url=https://assets.thermofisher.com/TFS-Assets/LSG/certificate/Certificates%20of%20Analysis/PB199188A%5C_90110.pdf" class="external-link">PB199188A</a>,
    contained a typo.)</p>
<p>the impurity correction matrix is</p>
<p></p><table class="table table"><tr><td>.</td><td>126</td><td>127N</td><td>127C</td><td>128N</td><td>128C</td><td>129N</td><td>129C</td><td>130N</td><td>130C</td><td>131</td></tr><tr><td>% reporter 126</td><td>0.950</td><td>0.000</td><td>0.050</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 127N</td><td>0.000</td><td>0.940</td><td>0.000</td><td>0.058</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 127C</td><td>0.003</td><td>0.000</td><td>0.949</td><td>0.000</td><td>0.048</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 128N</td><td>0.000</td><td>0.004</td><td>0.000</td><td>0.955</td><td>0.000</td><td>0.041</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 128C</td><td>0.000</td><td>0.000</td><td>0.006</td><td>0.000</td><td>0.964</td><td>0.000</td><td>0.030</td><td>0.000</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 129N</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.008</td><td>0.000</td><td>0.957</td><td>0.000</td><td>0.035</td><td>0.000</td><td>0.000</td></tr><tr><td>% reporter 129C</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.014</td><td>0.000</td><td>0.962</td><td>0.000</td><td>0.024</td><td>0.000</td></tr><tr><td>% reporter 130N</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.001</td><td>0.000</td><td>0.015</td><td>0.000</td><td>0.928</td><td>0.000</td><td>0.024</td></tr><tr><td>% reporter 130C</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.017</td><td>0.000</td><td>0.965</td><td>0.000</td></tr><tr><td>% reporter 131</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.000</td><td>0.002</td><td>0.000</td><td>0.020</td><td>0.000</td><td>0.956</td></tr></table><p>These examples are provided as defaults impurity correction matrices
    in <code>makeImpuritiesMatrix</code>.</p>
<p></p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## quantifying full experiment</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">impurities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.929</span>,<span class="fl">0.059</span>,<span class="fl">0.002</span>,<span class="fl">0.000</span>,</span></span>
<span class="r-in"><span>           <span class="fl">0.020</span>,<span class="fl">0.923</span>,<span class="fl">0.056</span>,<span class="fl">0.001</span>,</span></span>
<span class="r-in"><span>           <span class="fl">0.000</span>,<span class="fl">0.030</span>,<span class="fl">0.924</span>,<span class="fl">0.045</span>,</span></span>
<span class="r-in"><span>           <span class="fl">0.000</span>,<span class="fl">0.001</span>,<span class="fl">0.040</span>,<span class="fl">0.923</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>         nrow<span class="op">=</span><span class="fl">4</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## or, using makeImpuritiesMatrix()</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="va">impurities</span> <span class="op">&lt;-</span> <span class="fu">makeImpuritiesMatrix</span><span class="op">(</span><span class="fl">4</span><span class="op">)</span> <span class="co"># \dontrun{}</span></span></span>
<span class="r-in"><span><span class="va">msnset.crct</span> <span class="op">&lt;-</span> <span class="fu">purityCorrect</span><span class="op">(</span><span class="va">msnset</span>, <span class="va">impurities</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X1   1347.6158  2247.3097  3927.6931  7661.1463</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X10   739.9861   799.3501   712.5983   940.6793</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X11 27638.3582 33394.0252 32104.2879 26628.7278</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X12 31892.8928 33634.6980 37674.7272 37227.7119</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X13 26143.7542 29677.4781 29089.0593 27902.5608</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X14  6448.0829  6234.1957  6902.8903  6437.2303</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">msnset.crct</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X1   1402.9442  2214.0346  3762.2549  8114.4429</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X10   779.4666   793.0792   678.8083   985.2003</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X11 29034.3781 33271.0470 31484.7131 27279.1383</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X12 33618.9092 33046.3075 37031.6133 38492.1376</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X13 27508.0038 29440.9296 28390.4561 28814.2463</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> X14  6809.7600  6090.7894  6799.5030  6636.1450</span>
<span class="r-in"><span><span class="fu"><a href="pSet-class.html">processingData</a></span><span class="op">(</span><span class="va">msnset.crct</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> - - - Processing information - - -</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Data loaded: Wed May 11 18:54:39 2011 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iTRAQ4 quantification by trapezoidation: Wed Apr  1 21:41:53 2015 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Purity corrected: Mon Aug  4 19:04:00 2025 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  MSnbase version: 1.1.22 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## default impurity matrix for iTRAQ 8-plex</span></span></span>
<span class="r-in"><span><span class="fu">makeImpuritiesMatrix</span><span class="op">(</span><span class="fl">8</span>, edit <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  113   114   115   116   117   118   119   121</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 113 0.944 0.030 0.001 0.000 0.000 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 114 0.010 0.929 0.059 0.002 0.000 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 115 0.000 0.020 0.923 0.056 0.001 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 116 0.000 0.000 0.030 0.924 0.045 0.001 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 117 0.000 0.000 0.001 0.040 0.923 0.035 0.001 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 118 0.000 0.000 0.000 0.001 0.020 0.948 0.030 0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 119 0.000 0.000 0.000 0.000 0.001 0.020 0.938 0.040</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 121 0.000 0.000 0.000 0.000 0.000 0.001 0.020 0.948</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## default impurity matrix for TMT 10-plex</span></span></span>
<span class="r-in"><span><span class="fu">makeImpuritiesMatrix</span><span class="op">(</span><span class="fl">10</span>, edit <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   126  127N  127C  128N  128C  129N  129C  130N  130C   131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 126  0.950 0.000 0.050 0.000 0.000 0.000 0.000 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 127N 0.000 0.940 0.000 0.058 0.000 0.000 0.000 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 127C 0.003 0.000 0.949 0.000 0.048 0.000 0.000 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 128N 0.000 0.004 0.000 0.955 0.000 0.041 0.000 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 128C 0.000 0.000 0.006 0.000 0.964 0.000 0.030 0.000 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 129N 0.000 0.000 0.000 0.008 0.000 0.957 0.000 0.035 0.000 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 129C 0.000 0.000 0.000 0.000 0.014 0.000 0.962 0.000 0.024 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 130N 0.000 0.000 0.000 0.001 0.000 0.015 0.000 0.928 0.000 0.024</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 130C 0.000 0.000 0.000 0.000 0.000 0.000 0.017 0.000 0.965 0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> % reporter 131  0.000 0.000 0.000 0.000 0.000 0.002 0.000 0.020 0.000 0.956</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.9000.</p>
</div>

    </footer></div>





  </body></html>

