<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MSnbase IO capabilities • MSnbase</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MSnbase IO capabilities">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">2.7.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/v04-benchmarking.html">MSnbase2 benchmarking</a>
    </li>
    <li>
      <a href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/MSnbase">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MSnbase IO capabilities</h1>
                        <h4 class="author">Laurent Gatto</h4>
            <address class="author_afil">
      Computational Proteomics Unit, Cambridge, UK.<br><small class="dont-index">Source: <a href="https://github.com/lgatto/MSnbase/blob/master/vignettes/v02-MSnbase-io.Rmd"><code>vignettes/v02-MSnbase-io.Rmd</code></a></small>
      <div class="hidden name"><code>v02-MSnbase-io.Rmd</code></div>

    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      This vignette describes <em>MSnbase</em>’s input and output capabilities.
    </div>
    
<div id="foreword" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#foreword" class="anchor"></a>Foreword</h1>
<p><a href="http://bioconductor.org/packages/MSnbase"><em>MSnbase</em></a> is under active development; current functionality is evolving and new features will be added. This software is free and open-source software. If you use it, please support the project by citing it in publications:</p>
<blockquote>
<p>Gatto L, Lilley KS. MSnbase-an R/Bioconductor package for isobaric tagged mass spectrometry data visualization, processing and quantitation. Bioinformatics. 2012 Jan 15;28(2):288-9. doi: <a href="https://doi.org/10.1093/bioinformatics/btr645">10.1093/bioinformatics/btr645</a>. PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22113085">22113085</a>.</p>
</blockquote>
</div>
<div id="questions-and-bugs" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#questions-and-bugs" class="anchor"></a>Questions and bugs</h1>
<p>For bugs, typos, suggestions or other questions, please file an issue in our tracking system (<a href="https://github.com/lgatto/MSnbase/issues" class="uri">https://github.com/lgatto/MSnbase/issues</a>) providing as much information as possible, a reproducible example and the output of <code>sessionInfo()</code>.</p>
<p>If you don’t have a GitHub account or wish to reach a broader audience for general questions about proteomics analysis using R, you may want to use the Bioconductor support site: <a href="https://support.bioconductor.org/" class="uri">https://support.bioconductor.org/</a>.</p>
</div>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p><em><a href="https://bioconductor.org/packages/3.8/MSnbase">MSnbase</a></em>’s aims are to facilitate the reproducible analysis of mass spectrometry data within the R environment, from raw data import and processing, feature quantification, quantification and statistical analysis of the results <span class="citation">(Gatto and Lilley 2012)</span>. Data import functions for several formats are provided and intermediate or final results can also be saved or exported. These capabilities are presented below.</p>
</div>
<div id="data-input" class="section level1">
<h1 class="hasAnchor">
<a href="#data-input" class="anchor"></a>Data input</h1>
<div id="raw-data" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#raw-data" class="anchor"></a>Raw data</h4>
<p>Data stored in one of the published <code>XML</code>-based formats. i.e. <code>mzXML</code> <span class="citation">(Pedrioli et al. 2004)</span>, <code>mzData</code> <span class="citation">(Orchard et al. 2007)</span> or <code>mzML</code> <span class="citation">(Martens et al. 2010)</span>, can be imported with the <code>readMSData</code> method, which makes use of the <em><a href="https://bioconductor.org/packages/3.8/mzR">mzR</a></em> package to create <code>MSnExp</code> objects. The files can be in profile or centroided mode. See <code><a href="../reference/readMSData.html">?readMSData</a></code> for details.</p>
<p>Data from <code>mzML</code> files containing chromatographic data (e.g. generated in SRM/MRM experiments) can be imported with the <code>readSRMData</code> function that returns the chromatographic data as a <code>Chromatograms</code> object. See <code><a href="../reference/readSRMData.html">?readSRMData</a></code> for more details.</p>
</div>
<div id="peak-lists" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#peak-lists" class="anchor"></a>Peak lists</h4>
<p>Peak lists in the <code>mgf</code> format<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> can be imported using the <code>readMgfData</code>. In this case, the peak data has generally been pre-processed by other software. See <code><a href="../reference/readMgfData.html">?readMgfData</a></code> for details.</p>
</div>
<div id="quantitation-data" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#quantitation-data" class="anchor"></a>Quantitation data</h4>
<p>Third party software can be used to generate quantitative data and exported as a spreadsheet (generally comma or tab separated format). This data as well as any additional meta-data can be imported with the <code>readMSnSet</code> function. See <code><a href="../reference/readMSnSet.html">?readMSnSet</a></code> for details.</p>
<p><em><a href="https://bioconductor.org/packages/3.8/MSnbase">MSnbase</a></em> also supports the <code>mzTab</code> format<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>, a light-weight, tab-delimited file format for proteomics data developed within the Proteomics Standards Initiative (PSI). <code>mzTab</code> files can be read into R with <code>readMzTabData</code> to create and <code>MSnSet</code> instance.</p>
<div class="figure">
<img src="Figures/MSnbase-io-in.png" alt="MSnbase input capabilities. The white and red boxes represent R functions/methods and objects respectively. The blue boxes represent different disk storage formats."><p class="caption"><em>MSnbase</em> input capabilities. The white and red boxes represent R functions/methods and objects respectively. The blue boxes represent different disk storage formats.</p>
</div>
</div>
</div>
<div id="data-output" class="section level1">
<h1 class="hasAnchor">
<a href="#data-output" class="anchor"></a>Data output</h1>
<div id="rdata-files" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#rdata-files" class="anchor"></a>RData files</h4>
<p>R objects can most easily be stored on disk with the <code>save</code> function. It creates compressed binary images of the data representation that can later be read back from the file with the <code>load</code> function.</p>
</div>
<div id="mzmlmzxml-files" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#mzmlmzxml-files" class="anchor"></a>mzML/mzXML files</h4>
<p><code>MSnExp</code> and <code>OnDiskMSnExp</code> files can be written to MS data files in <code>mzML</code> or <code>mzXML</code> files with the <code>writeMSData</code> method. See <code><a href="../reference/writeMSData.html">?writeMSData</a></code> for details.</p>
</div>
<div id="peak-lists-1" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#peak-lists-1" class="anchor"></a>Peak lists</h4>
<p><code>MSnExp</code> instances as well as individual spectra can be written as <code>mgf</code> files with the <code>writeMgfData</code> method. Note that the meta-data in the original R object can not be included in the file. See <code><a href="../reference/writeMgfData-methods.html">?writeMgfData</a></code> for details.</p>
</div>
<div id="quantitation-data-1" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#quantitation-data-1" class="anchor"></a>Quantitation data</h4>
<p>Quantitation data can be exported to spreadsheet files with the <code>write.exprs</code> method. Feature meta-data can be appended to the feature intensity values. See <code><a href="../reference/writeMgfData-methods.html">?writeMgfData</a></code> for details.</p>
<p><strong>Deprecated</strong> <code>MSnSet</code> instances can also be exported to <code>mzTab</code> files using the <code>writeMzTabData</code> function.</p>
<div class="figure">
<img src="Figures/MSnbase-io-out.png" alt="MSnbase output capabilities. The white and red boxes represent R functions/methods and objects respectively. The blue boxes represent different disk storage formats."><p class="caption"><em>MSnbase</em> output capabilities. The white and red boxes represent R functions/methods and objects respectively. The blue boxes represent different disk storage formats.</p>
</div>
</div>
</div>
<div id="creating-msnset-from-text-spread-sheets" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-msnset-from-text-spread-sheets" class="anchor"></a>Creating <code>MSnSet</code> from text spread sheets</h1>
<p>This section describes the generation of <code>MSnSet</code> objects using data available in a text-based spreadsheet. This entry point into R and <em><a href="https://bioconductor.org/packages/3.8/MSnbase">MSnbase</a></em> allows to import data processed by any of the third party mass-spectrometry processing software available and proceed with data exploration, normalisation and statistical analysis using functions available in and the numerous Bioconductor packages.</p>
<div id="a-complete-work-flow" class="section level2">
<h2 class="hasAnchor">
<a href="#a-complete-work-flow" class="anchor"></a>A complete work flow</h2>
<p>The following section describes a work flow that uses three input files to create the <code>MSnSet</code>. These files respectively describe the quantitative expression data, the sample meta-data and the feature meta-data. It is taken from the <em><a href="https://bioconductor.org/packages/3.8/pRoloc">pRoloc</a></em> tutorial and uses example files from the <em><a href="https://bioconductor.org/packages/3.8/pRolocdat">pRolocdat</a></em> package.</p>
<p>We start by describing the <code>csv</code> to be used as input using the <code>read.csv</code> function.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">## The original data for replicate 1, available</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">## from the pRolocdata package</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3">f0 &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"pRolocdata"</span>), </a>
<a class="sourceLine" id="cb1-4" data-line-number="4">          <span class="dt">full.names =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb1-5" data-line-number="5">          <span class="dt">pattern =</span> <span class="st">"pr800866n_si_004-rep1.csv"</span>)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6">csv &lt;-<span class="st"> </span><span class="kw">read.csv</span>(f0)</a></code></pre></div>
<p>The three first lines of the original spreadsheet, containing the data for replicate one, are illustrated below (using the function <code>head</code>). It contains 888 rows (proteins) and 16 columns, including protein identifiers, database accession numbers, gene symbols, reporter ion quantitation values, information related to protein identification, …</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">head</span>(csv, <span class="dt">n=</span><span class="dv">3</span>)</a></code></pre></div>
<pre><code>##   Protein.ID        FBgn Flybase.Symbol No..peptide.IDs Mascot.score
## 1    CG10060 FBgn0001104    G-ialpha65A               3       179.86
## 2    CG10067 FBgn0000044         Act57B               5       222.40
## 3    CG10077 FBgn0035720        CG10077               5       219.65
##   No..peptides.quantified area.114 area.115 area.116 area.117
## 1                       1 0.379000 0.281000 0.225000 0.114000
## 2                       9 0.420000 0.209667 0.206111 0.163889
## 3                       3 0.187333 0.167333 0.169667 0.476000
##   PLS.DA.classification Peptide.sequence Precursor.ion.mass
## 1                    PM                                    
## 2                    PM                                    
## 3                                                          
##   Precursor.ion.charge pd.2013 pd.markers
## 1                           PM    unknown
## 2                           PM    unknown
## 3                      unknown    unknown</code></pre>
<p>Below read in turn the spread sheets that contain the quantitation data (<code>exprsFile.csv</code>), feature meta-data (<code>fdataFile.csv</code>) and sample meta-data (<code>pdataFile.csv</code>).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">## The quantitation data, from the original data</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">f1 &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"pRolocdata"</span>), </a>
<a class="sourceLine" id="cb4-3" data-line-number="3">          <span class="dt">full.names =</span> <span class="ot">TRUE</span>, <span class="dt">pattern =</span> <span class="st">"exprsFile.csv"</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">exprsCsv &lt;-<span class="st"> </span><span class="kw">read.csv</span>(f1)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">## Feature meta-data, from the original data</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6">f2 &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"pRolocdata"</span>), </a>
<a class="sourceLine" id="cb4-7" data-line-number="7">          <span class="dt">full.names =</span> <span class="ot">TRUE</span>, <span class="dt">pattern =</span> <span class="st">"fdataFile.csv"</span>)</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">fdataCsv &lt;-<span class="st"> </span><span class="kw">read.csv</span>(f2)</a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">## Sample meta-data, a new file</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10">f3 &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"pRolocdata"</span>), </a>
<a class="sourceLine" id="cb4-11" data-line-number="11">          <span class="dt">full.names =</span> <span class="ot">TRUE</span>, <span class="dt">pattern =</span> <span class="st">"pdataFile.csv"</span>)</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">pdataCsv &lt;-<span class="st"> </span><span class="kw">read.csv</span>(f3)</a></code></pre></div>
<p><code>exprsFile.csv</code> contains the quantitation (expression) data for the 888 proteins and 4 reporter tags.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">head</span>(exprsCsv, <span class="dt">n =</span> <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##          FBgn     X114     X115     X116     X117
## 1 FBgn0001104 0.379000 0.281000 0.225000 0.114000
## 2 FBgn0000044 0.420000 0.209667 0.206111 0.163889
## 3 FBgn0035720 0.187333 0.167333 0.169667 0.476000</code></pre>
<p><code>fdataFile.csv</code> contains meta-data for the 888 features (here proteins).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">head</span>(fdataCsv, <span class="dt">n =</span> <span class="dv">3</span>)</a></code></pre></div>
<pre><code>##          FBgn ProteinID FlybaseSymbol NoPeptideIDs MascotScore
## 1 FBgn0001104   CG10060   G-ialpha65A            3      179.86
## 2 FBgn0000044   CG10067        Act57B            5      222.40
## 3 FBgn0035720   CG10077       CG10077            5      219.65
##   NoPeptidesQuantified PLSDA
## 1                    1    PM
## 2                    9    PM
## 3                    3</code></pre>
<p><code>pdataFile.csv</code> contains samples (here fractions) meta-data. This simple file has been created manually.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">pdataCsv</a></code></pre></div>
<pre><code>##   sampleNames Fractions
## 1        X114       4/5
## 2        X115     12/13
## 3        X116        19
## 4        X117        21</code></pre>
<p>The self-contained <code>MSnSet</code> can now easily be generated using the <code>readMSnSet</code> constructor, providing the respective <code>csv</code> file names shown above and specifying that the data is comma-separated (with <code>sep = ","</code>). Below, we call that object <code>res</code> and display its content.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">library</span>(<span class="st">"MSnbase"</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readMSnSet.html">readMSnSet</a></span>(<span class="dt">exprsFile =</span> f1,</a>
<a class="sourceLine" id="cb11-3" data-line-number="3">                  <span class="dt">featureDataFile =</span> f2,</a>
<a class="sourceLine" id="cb11-4" data-line-number="4">                  <span class="dt">phenoDataFile =</span> f3,</a>
<a class="sourceLine" id="cb11-5" data-line-number="5">                  <span class="dt">sep =</span> <span class="st">","</span>)</a>
<a class="sourceLine" id="cb11-6" data-line-number="6">res</a></code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 888 features, 4 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: X114 X115 X116 X117
##   varLabels: Fractions
##   varMetadata: labelDescription
## featureData
##   featureNames: FBgn0001104 FBgn0000044 ... FBgn0001215 (888
##     total)
##   fvarLabels: ProteinID FlybaseSymbol ... PLSDA (6 total)
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## - - - Processing information - - -
## Quantitation data loaded: Mon Oct 15 12:41:32 2018  using readMSnSet. 
##  MSnbase version: 2.7.7</code></pre>
<div id="the-msnset-class" class="section level3">
<h3 class="hasAnchor">
<a href="#the-msnset-class" class="anchor"></a>The <code>MSnSet</code> class</h3>
<p>Although there are additional specific sub-containers for additional meta-data (for instance to make the object MIAPE compliant), the feature (the sub-container, or slot <code>featureData</code>) and sample (the <code>phenoData</code> slot) are the most important ones. They need to meet the following validity requirements (see figure below):</p>
<ul>
<li><p>the number of row in the expression/quantitation data and feature data must be equal and the row names must match exactly, and</p></li>
<li><p>the number of columns in the expression/quantitation data and number of row in the sample meta-data must be equal and the column/row names must match exactly.</p></li>
</ul>
<p>A detailed description of the <code>MSnSet</code> class is available by typing <code><a href="../reference/MSnSet-class.html">?MSnSet</a></code> in the R console.</p>
<div class="figure">
<img src="Figures/msnset.png" alt="Dimension requirements for the respective expression, feature and sample meta-data slots."><p class="caption">Dimension requirements for the respective expression, feature and sample meta-data slots.</p>
</div>
<p>The individual parts of this data object can be accessed with their respective accessor methods:</p>
<ul>
<li>the quantitation data can be retrieved with <code>exprs(res)</code>,</li>
<li>the feature meta-data with <code>fData(res)</code> and</li>
<li>the sample meta-data with <code>pData(res)</code>.</li>
</ul>
</div>
</div>
<div id="a-shorter-work-flow" class="section level2">
<h2 class="hasAnchor">
<a href="#a-shorter-work-flow" class="anchor"></a>A shorter work flow</h2>
<p>The <code>readMSnSet2</code> function provides a simplified import workforce. It takes a single spreadsheet as input (default is <code>csv</code>) and extract the columns identified by <code>ecol</code> to create the expression data, while the others are used as feature meta-data. <code>ecol</code> can be a <code>character</code> with the respective column labels or a numeric with their indices. In the former case, it is important to make sure that the names match exactly. Special characters like <code>'-'</code> or <code>'('</code> will be transformed by R into <code>'.'</code> when the <code>csv</code> file is read in. Optionally, one can also specify a column to be used as feature names. Note that these must be unique to guarantee the final object validity.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">ecol &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"area"</span>, <span class="dv">114</span><span class="op">:</span><span class="dv">117</span>, <span class="dt">sep =</span> <span class="st">"."</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">fname &lt;-<span class="st"> "Protein.ID"</span></a>
<a class="sourceLine" id="cb13-3" data-line-number="3">eset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readMSnSet.html">readMSnSet2</a></span>(f0, ecol, fname)</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">eset</a></code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 888 features, 4 samples 
##   element names: exprs 
## protocolData: none
## phenoData: none
## featureData
##   featureNames: CG10060 CG10067 ... CG9983 (888 total)
##   fvarLabels: Protein.ID FBgn ... pd.markers (12 total)
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.7.7</code></pre>
<p>The <code>ecol</code> columns can also be queried interactively from R using the <code>getEcols</code> and <code>grepEcols</code> function. The former return a character with all column names, given a splitting character, i.e. the separation value of the spreadsheet (typically <code>","</code> for <code>csv</code>, <code>"\t"</code> for <code>tsv</code>, …). The latter can be used to grep a pattern of interest to obtain the relevant column indices.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="../reference/grepEcols.html">getEcols</a></span>(f0, <span class="st">","</span>)</a></code></pre></div>
<pre><code>##  [1] "\"Protein ID\""              "\"FBgn\""                   
##  [3] "\"Flybase Symbol\""          "\"No. peptide IDs\""        
##  [5] "\"Mascot score\""            "\"No. peptides quantified\""
##  [7] "\"area 114\""                "\"area 115\""               
##  [9] "\"area 116\""                "\"area 117\""               
## [11] "\"PLS-DA classification\""   "\"Peptide sequence\""       
## [13] "\"Precursor ion mass\""      "\"Precursor ion charge\""   
## [15] "\"pd.2013\""                 "\"pd.markers\""</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="../reference/grepEcols.html">grepEcols</a></span>(f0, <span class="st">"area"</span>, <span class="st">","</span>)</a></code></pre></div>
<pre><code>## [1]  7  8  9 10</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">e &lt;-<span class="st"> </span><span class="kw"><a href="../reference/grepEcols.html">grepEcols</a></span>(f0, <span class="st">"area"</span>, <span class="st">","</span>)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw"><a href="../reference/readMSnSet.html">readMSnSet2</a></span>(f0, e)</a></code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 888 features, 4 samples 
##   element names: exprs 
## protocolData: none
## phenoData: none
## featureData
##   featureNames: 1 2 ... 888 (888 total)
##   fvarLabels: Protein.ID FBgn ... pd.markers (12 total)
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.7.7</code></pre>
<p>The <code>phenoData</code> slot can now be updated accordingly using the replacement functions <code>phenoData&lt;-</code> or <code>pData&lt;-</code> (see <code><a href="../reference/MSnSet-class.html">?MSnSet</a></code> for details).</p>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 3.5.1 Patched (2018-10-10 r75432)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Manjaro Linux
## 
## Matrix products: default
## BLAS: /usr/lib/libblas.so.3.8.0
## LAPACK: /usr/lib/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
## [1] pRolocdata_1.19.5    MSnbase_2.7.7        ProtGenerics_1.13.0 
## [4] BiocParallel_1.15.15 mzR_2.15.5           Rcpp_0.12.19        
## [7] Biobase_2.41.2       BiocGenerics_0.27.1  BiocStyle_2.9.6     
## 
## loaded via a namespace (and not attached):
##  [1] lattice_0.20-35       assertthat_0.2.0      rprojroot_1.3-2      
##  [4] digest_0.6.18         foreach_1.4.4         R6_2.3.0             
##  [7] plyr_1.8.4            backports_1.1.2       mzID_1.19.0          
## [10] stats4_3.5.1          evaluate_0.12         ggplot2_3.0.0        
## [13] pillar_1.3.0          zlibbioc_1.27.0       rlang_0.2.2          
## [16] lazyeval_0.2.1        S4Vectors_0.19.22     preprocessCore_1.43.0
## [19] rmarkdown_1.10        pkgdown_1.1.0         desc_1.2.0           
## [22] stringr_1.3.1         munsell_0.5.0         compiler_3.5.1       
## [25] xfun_0.3              pkgconfig_2.0.2       pcaMethods_1.73.1    
## [28] htmltools_0.3.6       tidyselect_0.2.5      tibble_1.4.2         
## [31] roxygen2_6.1.0        bookdown_0.7          IRanges_2.15.18      
## [34] codetools_0.2-15      XML_3.98-1.16         crayon_1.3.4         
## [37] dplyr_0.7.6           MASS_7.3-50           commonmark_1.6       
## [40] grid_3.5.1            gtable_0.2.0          affy_1.59.1          
## [43] magrittr_1.5          scales_1.0.0          ncdf4_1.16           
## [46] stringi_1.2.4         impute_1.55.0         fs_1.2.6             
## [49] affyio_1.51.0         doParallel_1.0.14     bindrcpp_0.2.2       
## [52] limma_3.37.7          xml2_1.2.0            iterators_1.0.10     
## [55] tools_3.5.1           glue_1.3.0            purrr_0.2.5          
## [58] yaml_2.2.0            colorspace_1.3-2      BiocManager_1.30.3   
## [61] vsn_3.49.1            MALDIquant_1.18       memoise_1.1.0        
## [64] knitr_1.20            bindr_0.1.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Gatto2012">
<p>Gatto, Laurent, and Kathryn S Lilley. 2012. “MSnbase – an R/Bioconductor Package for Isobaric Tagged Mass Spectrometry Data Visualization, Processing and Quantitation.” <em>Bioinformatics</em> 28 (2): 288–9. <a href="https://doi.org/10.1093/bioinformatics/btr645">https://doi.org/10.1093/bioinformatics/btr645</a>.</p>
</div>
<div id="ref-Martens2010">
<p>Martens, Lennart, Matthew Chambers, Marc Sturm, Darren Kes sner, Fredrik Levander, Jim Shofstahl, Wilfred H Tang, et al. 2010. “MzML - a Community Standard for Mass Spectrometry Data.” <em>Molecular &amp; Cellular Proteomics : MCP</em>. <a href="https://doi.org/10.1074/mcp.R110.000133">https://doi.org/10.1074/mcp.R110.000133</a>.</p>
</div>
<div id="ref-Orchard2007">
<p>Orchard, Sandra, Luisa Montechi-Palazzi, Eric W Deutsch, Pierre-Alain Binz, Andrew R Jones, Norman Paton, Angel Pizarro, David M Creasy, Jérôme Wojcik, and Henning Hermjakob. 2007. “Five Years of Progress in the Standardization of Proteomics Data 4th Annual Spring Workshop of the Hupo-Proteomics Standards Initiative April 23-25, 2007 Ecole Nationale Supérieure (Ens), Lyon, France.” <em>Proteomics</em> 7 (19): 3436–40. <a href="https://doi.org/10.1002/pmic.200700658">https://doi.org/10.1002/pmic.200700658</a>.</p>
</div>
<div id="ref-Pedrioli2004">
<p>Pedrioli, Patrick G A, Jimmy K Eng, Robert Hubley, Mathijs Vogelzang, Eric W Deutsch, Brian Raught, Brian Pratt, et al. 2004. “A Common Open Representation of Mass Spectrometry Data and Its Application to Proteomics Research.” <em>Nat. Biotechnol.</em> 22 (11): 1459–66. <a href="https://doi.org/10.1038/nbt1031">https://doi.org/10.1038/nbt1031</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="http://www.matrixscience.com/help/data_file_help.html" class="uri">http://www.matrixscience.com/help/data_file_help.html</a><a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p><a href="https://github.com/HUPO-PSI/mzTab" class="uri">https://github.com/HUPO-PSI/mzTab</a><a href="#fnref2" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#foreword">Foreword</a></li>
      <li><a href="#questions-and-bugs">Questions and bugs</a></li>
      <li><a href="#overview">Overview</a></li>
      <li><a href="#data-input">Data input</a></li>
      <li><a href="#data-output">Data output</a></li>
      <li>
<a href="#creating-msnset-from-text-spread-sheets">Creating <code>MSnSet</code> from text spread sheets</a><ul class="nav nav-pills nav-stacked">
<li><a href="#a-complete-work-flow">A complete work flow</a></li>
      <li><a href="#a-shorter-work-flow">A shorter work flow</a></li>
      </ul>
</li>
      <li><a href="#session-information">Session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
