<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Quantitative proteomics data imputation — impute-methods • MSnbase</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Quantitative proteomics data imputation — impute-methods" />

<meta property="og:description" content="The impute method performs data imputation on an
  MSnSet instance using a variety of methods (see below).  The
  imputation and the parameters are logged into the
  processingData(object) slot.  
Users should proceed with care when imputing data and take precautions
  to assure that the imputation produce valid results, in particular
  with naive imputations such as replacing missing values with 0." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">2.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
    <li>
      <a href="../articles/MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/benchmarking.html">MSnbase2 benchmarking</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lgatto/MSnbase">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Quantitative proteomics data imputation</h1>
    
    <div class="hidden name"><code>impute-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>The <code>impute</code> method performs data imputation on an
  <code>MSnSet</code> instance using a variety of methods (see below).  The
  imputation and the parameters are logged into the
  <code><a href='pSet-class.html'>processingData(object)</a></code> slot.</p>  
<p>Users should proceed with care when imputing data and take precautions
  to assure that the imputation produce valid results, in particular
  with naive imputations such as replacing missing values with 0.</p>
    
    </div>

        
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    
    <p>There are two types of mechanisms resulting in missing values in
  LC/MSMS experiments.</p>
<p></p><ul>
<li><p>Missing values resulting from absence of detection of a
      feature, despite ions being present at detectable concentrations.
      For example in the case of ion suppression or as a result from the
      stochastic, data-dependent nature of the MS acquisition
      method. These missing value are expected to be randomly
      distributed in the data and are defined as missing at random (MAR)
      or missing completely at random (MCAR).</p></li>
<li><p>Biologically relevant missing values resulting from the
      absence of the low abundance of ions (below the limit of detection
      of the instrument). These missing values are not expected to be
      randomly distributed in the data and are defined as missing not at
      random (MNAR).</p></li>
</ul>
    <p>MNAR features should ideally be imputed with a left-censor method,
  such as <code>QRILC</code> below. Conversely, it is recommended to use host
  deck methods such nearest neighbours, Bayesian missing value
  imputation or maximum likelihood methods when values are missing at
  random.</p>
<p>Currently, the following imputation methods are available:</p>
<p></p><dl class='dl-horizontal'>
    <dt>MLE</dt><dd><p>Maximum likelihood-based imputation method using the EM
      algorithm. Implemented in the <code><a href='http://www.rdocumentation.org/packages/norm/topics/imp.norm'>norm::imp.norm</a></code> function. See
      <code><a href='http://www.rdocumentation.org/packages/norm/topics/imp.norm'>imp.norm</a></code> for details and additional
      parameters. Note that here, <code>...</code> are passed to the
      <code><a href='http://www.rdocumentation.org/packages/norm/topics/em.norm'>em.norm</a></code> function, rather to the actual
      imputation function <code>imp.norm</code>.</p></dd>
    <dt>bpca</dt><dd><p>Bayesian missing value imputation are available, as
    implemented in the and <code><a href='http://www.rdocumentation.org/packages/pcaMethods/topics/pca'>pcaMethods::pca</a></code> functions. See
    <code><a href='http://www.rdocumentation.org/packages/pcaMethods/topics/pca'>pca</a></code> for details and additional
    parameters.</p></dd>
    <dt>knn</dt><dd><p>Nearest neighbour averaging, as implemented in the
      <code><a href='http://www.rdocumentation.org/packages/impute/topics/impute.knn'>impute::impute.knn</a></code> function. See
      <code><a href='http://www.rdocumentation.org/packages/impute/topics/impute.knn'>impute.knn</a></code> for details and additional
      parameters.</p></dd>
    <dt>QRILC</dt><dd><p>A missing data imputation method that performs the
      imputation of left-censored missing data using random draws from a
      truncated distribution with parameters estimated using quantile
      regression. Implemented in the <code><a href='http://www.rdocumentation.org/packages/imputeLCMD/topics/impute.QRILC'>imputeLCMD::impute.QRILC</a></code>
      function. See <code><a href='http://www.rdocumentation.org/packages/imputeLCMD/topics/impute.QRILC'>impute.QRILC</a></code> for details
      and additional parameters.</p></dd>
    <dt>MinDet</dt><dd><p>Performs the imputation of left-censored missing data
      using a deterministic minimal value approach. Considering a
      expression data with <em>n</em> samples and <em>p</em> features, for
      each sample, the missing entries are replaced with a minimal value
      observed in that sample. The minimal value observed is estimated as
      being the q-th quantile (default <code>q = 0.01</code>) of the observed
      values in that sample. Implemented in the
      <code><a href='http://www.rdocumentation.org/packages/imputeLCMD/topics/impute.MinDet'>imputeLCMD::impute.MinDet</a></code> function. See
      <code><a href='http://www.rdocumentation.org/packages/imputeLCMD/topics/impute.MinDet'>impute.MinDet</a></code> for details and additional
      parameters.</p></dd>
    <dt>MinProb</dt><dd><p>Performs the imputation of left-censored missing data
      by random draws from a Gaussian distribution centred to a minimal
      value. Considering an expression data matrix with <em>n</em> samples
      and <em>p</em> features, for each sample, the mean value of the
      Gaussian distribution is set to a minimal observed value in that
      sample. The minimal value observed is estimated as being the q-th
      quantile (default <code>q = 0.01</code>) of the observed values in that
      sample. The standard deviation is estimated as the median of the
      feature standard deviations. Note that when estimating the
      standard deviation of the Gaussian distribution, only the
      peptides/proteins which present more than 50% recorded values are
      considered. Implemented in the <code><a href='http://www.rdocumentation.org/packages/imputeLCMD/topics/impute.MinProb'>imputeLCMD::impute.MinProb</a></code>
      function. See <code><a href='http://www.rdocumentation.org/packages/imputeLCMD/topics/impute.MinProb'>impute.MinProb</a></code> for details
      and additional parameters.</p></dd>
    <dt>min</dt><dd><p>Replaces the missing values by the smallest non-missing
      value in the data.</p></dd>
    <dt>zero</dt><dd><p>Replaces the missing values by 0.</p></dd>
    <dt>mixed</dt><dd><p>A mixed imputation applying two methods (to be defined
      by the user as <code>mar</code> for values missing at random and
      <code>mnar</code> for values missing not at random, see example) on two
      M[C]AR/MNAR subsets of the data (as defined by the user by a
      <code>randna</code> logical, of length equal to <code>nrow(object)</code>).</p></dd>
    <dt>nbavg</dt><dd><p>Average neighbour imputation for fractions collected
      along a fractionation/separation gradient, such as sub-cellular
      fractions. The method assumes that the fraction are ordered along
      the gradient and is invalid otherwise.
    Continuous sets <code>NA</code> value at the beginning and the end of
      the quantitation vectors are set to the lowest observed value in
      the data or to a user defined value passed as argument <code>k</code>.
      Them, when a missing value is flanked by two non-missing
      neighbouring values, it is imputed by the mean of its direct
      neighbours. A stretch of 2 or more missing values will not be
      imputed. See the example below.</p></dd>
    </dl>
    <p>The <code>naset</code> <code><a href='MSnSet-class.html'>MSnSet</a></code> is an real quantitative
  data where quantitative values have been replaced by <code>NA</code>s. See
  <code>script/naset.R</code> for details.</p>
    
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    <p></p><dl class='dl-horizontal'>
    <dt><code>signature(object = "MSnSet", method, ...)</code></dt><dd><p>This method
      performs data imputation on the <code>object</code> <code>MSnSet</code>
      instance using the <code>method</code> algorithm. <code>...</code> is used to
      pass parameters to the imputation function. See the respective
      methods for details and additional parameters.</p></dd>
    </dl>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    
    <p>Olga Troyanskaya, Michael Cantor, Gavin Sherlock, Pat Brown, Trevor
Hastie, Robert Tibshirani, David Botstein and Russ B.  Altman, Missing
value estimation methods for DNA microarrays Bioinformatics (2001) 17
(6): 520-525.</p>
<p>Oba et al., A Bayesian missing value estimation method for gene
expression profile data, Bioinformatics (2003) 19 (16): 2088-2096.</p>
<p>Cosmin Lazar (2015). imputeLCMD: A collection of methods for
left-censored missing data imputation. R package version
2.0. <a href='http://CRAN.R-project.org/package=imputeLCMD'>http://CRAN.R-project.org/package=imputeLCMD</a>.</p>
<p>Lazar C, Gatto L, Ferro M, Bruley C, Burger T. Accounting for the
Multiple Natures of Missing Values in Label-Free Quantitative Proteomics
Data Sets to Compare Imputation Strategies. J Proteome Res. 2016 Apr
1;15(4):1116-25. doi: 10.1021/acs.jproteome.5b00981. PubMed PMID:
26906401.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>naset</span>)
<span class='co'>## table of missing values along the rows</span>
<span class='fu'>table</span>(<span class='fu'>fData</span>(<span class='no'>naset</span>)$<span class='no'>nNA</span>)</div><div class='output co'>#&gt; 
#&gt;   0   1   2   3   4   8   9  10 
#&gt; 301 247  91  13   2  23  10   2 </div><div class='input'><span class='co'>## table of missing values along the columns</span>
<span class='fu'>pData</span>(<span class='no'>naset</span>)$<span class='no'>nNA</span></div><div class='output co'>#&gt;  [1] 34 45 56 39 47 52 49 61 41 42 55 45 51 43 57 53</div><div class='input'>
<span class='co'>## non-random missing values</span>
<span class='no'>notna</span> <span class='kw'>&lt;-</span> <span class='fu'>which</span>(!<span class='fu'>fData</span>(<span class='no'>naset</span>)$<span class='no'>randna</span>)
<span class='fu'><a href='pSet-class.html'>length</a></span>(<span class='no'>notna</span>)</div><div class='output co'>#&gt; [1] 35</div><div class='input'><span class='no'>notna</span></div><div class='output co'>#&gt;  [1]   6  20  79  88 130 187 227 231 238 264 275 317 324 363 373 382 409 437 445
#&gt; [20] 453 456 474 484 485 492 514 516 546 568 580 594 631 648 664 671</div><div class='input'>
<span class='fu'>impute</span>(<span class='no'>naset</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"min"</span>)</div><div class='output co'>#&gt; MSnSet (storageMode: lockedEnvironment)
#&gt; assayData: 689 features, 16 samples 
#&gt;   element names: exprs 
#&gt; protocolData: none
#&gt; phenoData
#&gt;   sampleNames: M1F1A M1F4A ... M2F11B (16 total)
#&gt;   varLabels: nNA
#&gt;   varMetadata: labelDescription
#&gt; featureData
#&gt;   featureNames: AT1G09210 AT1G21750 ... AT4G39080 (689 total)
#&gt;   fvarLabels: nNA randna
#&gt;   fvarMetadata: labelDescription
#&gt; experimentData: use 'experimentData(object)'
#&gt; Annotation:  
#&gt; - - - Processing information - - -
#&gt; Data imputation using min Thu May 17 15:34:27 2018 
#&gt;  MSnbase version: 1.15.6 </div><div class='input'>
<span class='kw'>if</span> (<span class='fu'>require</span>(<span class='st'>"imputeLCMD"</span>)) {
    <span class='fu'>impute</span>(<span class='no'>naset</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"QRILC"</span>)
    <span class='fu'>impute</span>(<span class='no'>naset</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"MinDet"</span>)
}</div><div class='output co'>#&gt; <span class='message'>Loading required package: imputeLCMD</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: tmvtnorm</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: mvtnorm</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Matrix</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘Matrix’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:S4Vectors’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: gmm</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: sandwich</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: norm</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: pcaMethods</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘pcaMethods’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    loadings</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: impute</span></div><div class='output co'>#&gt; MSnSet (storageMode: lockedEnvironment)
#&gt; assayData: 689 features, 16 samples 
#&gt;   element names: exprs 
#&gt; protocolData: none
#&gt; phenoData
#&gt;   sampleNames: M1F1A M1F4A ... M2F11B (16 total)
#&gt;   varLabels: nNA
#&gt;   varMetadata: labelDescription
#&gt; featureData
#&gt;   featureNames: AT1G09210 AT1G21750 ... AT4G39080 (689 total)
#&gt;   fvarLabels: nNA randna
#&gt;   fvarMetadata: labelDescription
#&gt; experimentData: use 'experimentData(object)'
#&gt; Annotation:  
#&gt; - - - Processing information - - -
#&gt; Data imputation using MinDet Thu May 17 15:34:27 2018 
#&gt;   Using default parameters 
#&gt;  MSnbase version: 1.15.6 </div><div class='input'>
<span class='kw'>if</span> (<span class='fu'>require</span>(<span class='st'>"norm"</span>))
    <span class='fu'>impute</span>(<span class='no'>naset</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"MLE"</span>)</div><div class='output co'>#&gt; Iterations of EM: 
#&gt; 1...2...3...4...5...6...7...8...9...10...11...</div><div class='output co'>#&gt; MSnSet (storageMode: lockedEnvironment)
#&gt; assayData: 689 features, 16 samples 
#&gt;   element names: exprs 
#&gt; protocolData: none
#&gt; phenoData
#&gt;   sampleNames: M1F1A M1F4A ... M2F11B (16 total)
#&gt;   varLabels: nNA
#&gt;   varMetadata: labelDescription
#&gt; featureData
#&gt;   featureNames: AT1G09210 AT1G21750 ... AT4G39080 (689 total)
#&gt;   fvarLabels: nNA randna
#&gt;   fvarMetadata: labelDescription
#&gt; experimentData: use 'experimentData(object)'
#&gt; Annotation:  
#&gt; - - - Processing information - - -
#&gt; Data imputation using MLE Thu May 17 15:34:27 2018 
#&gt;   Using default parameters 
#&gt;  MSnbase version: 1.15.6 </div><div class='input'>
<span class='fu'>impute</span>(<span class='no'>naset</span>, <span class='st'>"mixed"</span>,
       <span class='kw'>randna</span> <span class='kw'>=</span> <span class='fu'>fData</span>(<span class='no'>naset</span>)$<span class='no'>randna</span>,
       <span class='kw'>mar</span> <span class='kw'>=</span> <span class='st'>"knn"</span>, <span class='kw'>mnar</span> <span class='kw'>=</span> <span class='st'>"QRILC"</span>)</div><div class='output co'>#&gt; MSnSet (storageMode: lockedEnvironment)
#&gt; assayData: 689 features, 16 samples 
#&gt;   element names: exprs 
#&gt; protocolData: none
#&gt; phenoData
#&gt;   sampleNames: M1F1A M1F4A ... M2F11B (16 total)
#&gt;   varLabels: nNA
#&gt;   varMetadata: labelDescription
#&gt; featureData
#&gt;   featureNames: AT1G09210 AT1G21750 ... AT4G39080 (689 total)
#&gt;   fvarLabels: nNA randna
#&gt;   fvarMetadata: labelDescription
#&gt; experimentData: use 'experimentData(object)'
#&gt; Annotation:  
#&gt; - - - Processing information - - -
#&gt; Data imputation using mixed Thu May 17 15:34:27 2018 
#&gt;   Using default parameters 
#&gt;  MSnbase version: 1.15.6 </div><div class='input'>
<span class='co'>## neighbour averaging</span>

<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='no'>naset</span>[<span class='fl'>1</span>:<span class='fl'>4</span>, <span class='fl'>1</span>:<span class='fl'>6</span>]
<span class='fu'>exprs</span>(<span class='no'>x</span>)[<span class='fl'>1</span>, <span class='fl'>1</span>] <span class='kw'>&lt;-</span> <span class='fl'>NA</span> <span class='co'>## min value</span>
<span class='fu'>exprs</span>(<span class='no'>x</span>)[<span class='fl'>2</span>, <span class='fl'>3</span>] <span class='kw'>&lt;-</span> <span class='fl'>NA</span> <span class='co'>## average</span>
<span class='fu'>exprs</span>(<span class='no'>x</span>)[<span class='fl'>3</span>, <span class='fl'>1</span>:<span class='fl'>2</span>] <span class='kw'>&lt;-</span> <span class='fl'>NA</span> <span class='co'>## min value and average</span>
<span class='co'>## 4th row: no imputation</span>
<span class='fu'>exprs</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;              M1F1A    M1F4A   M1F7A  M1F11A    M1F2B    M1F5B
#&gt; AT1G09210       NA 0.275500 0.21600 0.18525 0.465667 0.199667
#&gt; AT1G21750 0.332000 0.279667      NA 0.16600 0.451500 0.200375
#&gt; AT1G51760       NA       NA 0.16825 0.18825 0.459750 0.214500
#&gt; AT1G56340 0.336733       NA      NA      NA 0.487167 0.201833</div><div class='input'>
<span class='fu'>exprs</span>(<span class='fu'>impute</span>(<span class='no'>x</span>, <span class='st'>"nbavg"</span>))</div><div class='output co'>#&gt; <span class='message'>Assuming values are ordered.</span></div><div class='output co'>#&gt;              M1F1A    M1F4A     M1F7A  M1F11A    M1F2B    M1F5B
#&gt; AT1G09210 0.166000 0.275500 0.2160000 0.18525 0.465667 0.199667
#&gt; AT1G21750 0.332000 0.279667 0.2228335 0.16600 0.451500 0.200375
#&gt; AT1G51760 0.166000 0.167125 0.1682500 0.18825 0.459750 0.214500
#&gt; AT1G56340 0.336733       NA        NA      NA 0.487167 0.201833</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#details">Details</a></li>

      <li><a href="#methods">Methods</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
  Laurent Gatto and Samuel Wieczorek

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

