<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>The "MSnSet" Class for MS Proteomics Expression Data and Meta-Data — MSnSet-class • MSnbase</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="The "MSnSet" Class for MS Proteomics Expression Data and Meta-Data — MSnSet-class" />

<meta property="og:description" content="The MSnSet holds quantified expression data for MS proteomics
  data and the experimental meta-data.
  The MSnSet class is derived from the
  "eSet" class and mimics the
  "ExpressionSet" class classically used for
  microarray data." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">2.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
    <li>
      <a href="../articles/MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/benchmarking.html">MSnbase2 benchmarking</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lgatto/MSnbase">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>The "MSnSet" Class for MS Proteomics Expression Data and Meta-Data</h1>
    
    <div class="hidden name"><code>MSnSet-class.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>The <code>MSnSet</code> holds quantified expression data for MS proteomics
  data and the experimental meta-data.
  The <code>MSnSet</code> class is derived from the
  <code>"eSet"</code> class and mimics the
  <code>"ExpressionSet"</code> class classically used for
  microarray data.</p>
    
    </div>

        
    <h2 class="hasAnchor" id="objects-from-the-class"><a class="anchor" href="#objects-from-the-class"></a>Objects from the Class</h2>

    
    <p>The constructor <code>MSnSet(exprs, fData, pData)</code> can be used to
  create <code>MSnSet</code> instances. Argument <code>exprs</code> is a
  <code>matrix</code> and <code>fData</code> and <code>pData</code> must be of class
  <code>data.frame</code> or <code>"AnnotatedDataFrame"</code> and all
  must meet the dimensions and name validity constrains.</p>
<p>Objects can also be created by calls of the form <code>new("MSnSet",
  exprs, ...)</code>.  See also <code>"ExpressionSet"</code> for
  helpful information.  Expression data produced from other softwares
  can thus make use of this standardized data container to benefit
  <code>R</code> and <code>Bioconductor</code> packages. Proteomics expression data
  available as spreadsheets, as produced by third-party software such as
  Proteome Discoverer, MaxQuant, ... can be imported using the
  <code><a href='readMSnSet.html'>readMSnSet</a></code> and <code><a href='readMSnSet.html'>readMSnSet2</a></code> functions.</p>
<p>Coercion methods are also available to transform <code>MSnSet</code> objects
  to <code>IBSpectra</code>, to <code>data.frame</code> and to/from
  <code>ExpressionSet</code> and <code>SummarizedExperiment</code> objects. In the
  latter case, the metadata available in the <code>protocolData</code>,
  <code>experimentData</code> are completely dropped, and only the logging
  information of the <code>processingData</code> slot is retained. All these
  metadata can be subsequently be added using the
  <code>addMSnSetMetadata</code> (see examples below). When converting a
  <code>SummarizedExperiment</code> to an <code>MSnSet</code>, the respective
  metadata slots will be populated if available in the
  <code>SummarizedExperiment</code> metadata.</p>
<p>In the frame of the <code>MSnbase</code> package, <code>MSnSet</code> instances
  can be generated from <code>"<a href='MSnExp-class.html'>MSnExp</a>"</code> experiments using
  the <code><a href='quantify-methods.html'>quantify</a></code> method).</p>
    
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    
    <p></p><dl class='dl-horizontal'>
    <dt><code>qual</code>:</dt><dd><p>Object of class <code>"data.frame"</code> that records
      peaks data for each of the reporter ions to be used as quality
      metrics.</p></dd>
    <dt><code>processingData</code>:</dt><dd><p>Object of class
      <code>"<a href='MSnProcess-class.html'>MSnProcess</a>"</code> that records all processing.</p></dd>
    <dt><code>assayData</code>:</dt><dd><p>Object of class <code>"assayData"</code>
      containing a matrix with equal with column number equal to
      <code>nrow(phenoData)</code>. <code>assayData</code> must contain a matrix
      <code>exprs</code> with rows represening features (e.g., reporters ions)
      and columns representing samples. See the <code>"AssayData"</code>
      class, <code>exprs</code> and <code>assayData</code> accessor
      for more details. This slot in indirectly inherited from
      <code>"eSet"</code>.</p></dd>
    <dt><code>phenoData</code>:</dt><dd><p>Object of class <code>"AnnotatedDataFrame"</code>
      containing experimenter-supplied variables describing sample (i.e
      the individual tags for an labelled MS experiment) (indireclty
      inherited from <code>"eSet"</code>). See
      <code>phenoData</code> and the <code>"eSet"</code> class
      for more details. This slot can be accessed as a <code>data.frame</code>
      with <code>pData</code> and be replaced by a new valid (i.e. of
      compatible dimensions and row names) <code>data.frame</code> with
      <code>pData()&lt;-</code>.</p></dd>
    <dt><code>featureData</code>:</dt><dd><p>Object of class
      <code>"AnnotatedDataFrame"</code> containing variables describing
      features (spectra in our case), e.g. identificaiton data, peptide
      sequence, identification score,... (inherited indirectly from
      <code>"eSet"</code>). See <code>featureData</code> and
      the <code>"eSet"</code> class for more details. This slot
      can be accessed as a <code>data.frame</code> with <code>fData</code> and be
      replaced by a new valid (i.e. of compatible dimensions and row
      names) <code>data.frame</code> with <code>fData()&lt;-</code>.</p></dd>
    <dt><code>experimentData</code>:</dt><dd><p>Object of class
      <code>"<a href='MIAPE-class.html'>MIAPE</a>"</code>, containing details of experimental
      methods (inherited from <code>"eSet"</code>). See
      <code>experimentData</code> and the <code>"eSet"</code>
      class for more details.</p></dd>
    <dt><code>annotation</code>:</dt><dd><p>not used here.</p></dd>
    <dt><code>protocolData</code>:</dt><dd><p>Object of class
      <code>"AnnotatedDataFrame"</code> containing
      equipment-generated variables (inherited indirectly from
      <code>"eSet"</code>). See
      <code>protocolData</code> and the <code>"eSet"</code>
      class for more details.</p></dd>
    <dt><code>.__classVersion__</code>:</dt><dd><p>Object of class
      <code>"Versions"</code> describing the versions of R,
      the Biobase package, <code>"eSet"</code>,
      <code>"<a href='pSet-class.html'>pSet</a>"</code> and <code>MSnSet</code> of the
      current instance. Intended for developer use and debugging (inherited
      indirectly from <code>"eSet"</code>).</p></dd>
  </dl>
    
    <h2 class="hasAnchor" id="extends"><a class="anchor" href="#extends"></a>Extends</h2>

    <p>Class <code>"eSet"</code>, directly.
  Class <code>"VersionedBiobase"</code>, by class "eSet", distance 2.
  Class <code>"Versioned"</code>, by class "eSet", distance 3.</p>
    
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    <p>MSnSet specific methods or over-riding it's super-class are described
  below. See also more <code>"eSet"</code> for
  inherited methods.</p>
<p></p><dl class='dl-horizontal'>
    %% \item{$name}{\code{signature(x = "pSet")}: Access \code{name} column
    %%   in \code{featureData}. Note that this behaviour is different to
    %%   the \code{eSet}, where \code{$} accesses the \code{phenoData}. }
    <dt>acquisitionNum</dt><dd><p><code><a href='Spectrum-class.html'>acquisitionNum(signature(object = "MSnSet"))</a></code>: Returns the
      a numeric vector with acquisition number of each spectrum. The vector
      names are the corresponding spectrum names.
      The information is extracted from the object's <code>featureData</code> slot.</p></dd>
    <dt>fromFile</dt><dd><p><code><a href='Spectrum-class.html'>fromFile(signature(object = "MSnSet"))</a></code>: get the index of
      the file (in <code><a href='pSet-class.html'>fileNames(object)</a></code>) from which the raw
      spectra from which the corresponding feature were originally
      read. The relevant information is extracted from the object's
      <code>featureData</code> slot.
    Returns a numeric vector with names corresponding to the spectrum
      names.</p></dd>
    <dt>dim</dt><dd><p><code>signature(x = "MSnSet")</code>: Returns the dimensions of
      object's assay data, i.e the number of samples and the number of
      features.</p></dd>
    <dt>fileNames</dt><dd><p><code>signature(object = "MSnSet")</code>: Access file
      names in the <code>processingData</code> slot.</p></dd>
    <dt>msInfo</dt><dd><p><code>signature(object = "MSnSet")</code>: Prints the MIAPE-MS
      meta-data stored in the <code>experimentData</code> slot.</p></dd>
    <dt>processingData</dt><dd><p><code>signature(object = "MSnSet")</code>: Access the
      <code>processingData</code> slot.</p></dd>
    <dt>show</dt><dd><p><code>signature(object = "MSnSet")</code>: Displays object
      content as text.</p></dd>
    <dt>qual</dt><dd><p><code>signature(object = "MSnSet")</code>: Access the reporter
      ion peaks description.</p></dd>
    <dt>purityCorrect</dt><dd><p><code>signature(object = "MSnSet", impurities =
	"matrix")</code>: performs reporter ions purity correction. See
      <code><a href='purityCorrect-methods.html'>purityCorrect</a></code> documentation for more details.</p></dd>
    <dt>normalise</dt><dd><p><code>signature(object = "MSnSet")</code>: Performs
      <code>MSnSet</code> normalisation. See <code><a href='normalise-methods.html'>normalise</a></code> for more
      details.</p></dd>
    <dt>t</dt><dd><p><code>signature(x = "MSnSet")</code>: Returns a transposed
      <code>MSnSet</code> object where features are now aligned along columns
      and samples along rows and the <code>phenoData</code> and
      <code>featureData</code> slots have been swapped. The
      <code>protocolData</code> slot is always dropped.</p></dd>
    <dt>as(,"ExpressionSet")</dt><dd><p><code>signature(x = "MSnSet")</code>: Coerce
      object from <code>MSnSet</code> to
      <code>ExpressionSet-class</code>. The <code>experimentData</code> slot
      is converted to a <code>MIAME</code> instance. It is also possible to
      coerce an <code>ExpressionSet</code> to and <code>MSnSet</code>, in which case
      the <code>experimentData</code> slot is newly initialised.</p></dd>
    <dt>as(,"IBSpectra")</dt><dd><p><code>signature(x = "MSnSet")</code>: Coerce
      object from <code>MSnSet</code> to <code>IBSpectra</code> from the
      <code>isobar</code> package.</p></dd>
    <dt>as(,"data.frame")</dt><dd><p><code>signature(x = "MSnSet")</code>: Coerce
      object from <code>MSnSet</code> to <code>data.frame</code>. The <code>MSnSet</code>
      is transposed and the <code>PhenoData</code> slot is appended.</p></dd>
    <dt>as(,"SummarizedExperiment")</dt><dd><p><code>signature(x = "MSnSet")</code>:
      Coerce object from <code>MSnSet</code> to
      <code>SummarizedExperiment</code>. Only part of the metadata is
      retained. See <code>addMSnSetMetadata</code> and the example below for
      details.</p></dd>
    %% \item{ratios}{\code{signature(object = "MSnSet")}: ... }
    <dt>write.exprs</dt><dd><p>signature(x = "MSnSet")</p></dd>Writes expression values
      to a tab-separated file (default is <code>tmp.txt</code>). The
      <code>fDataCols</code> parameter can be used to specify which
      <code>featureData</code> columns (as column names, column number or
      <code>logical</code>) to append on the right of the expression matrix.
      The following arguments are the same as <code>write.table</code>.
    <dt>combine</dt><dd><p>signature(x = "MSnSet", y = "MSnSet", ...)</p></dd> Combines
      2 or more <code>MSnSet</code> instances according to their feature names.
      Note that the <code>qual</code> slot and the processing information are
      silently dropped.
    <dt>topN</dt><dd><p>signature(object = "MSnSet", groupBy, n = 3, fun, ..., verbose =
      isMSnbaseVerbose())</p></dd>
      Selects the <code>n</code> most intense features (typically peptides or
      spectra) out of all available for each set defined by
      <code>groupBy</code> (typically proteins) and creates a new instance of
      class <code>MSnSet</code>. If less than <code>n</code> features are available,
      all are selected. The <code>ncol(object)</code> features are summerised
      using <code>fun</code> (default is <code>sum</code>) prior to be ordered in
      decreasing order. Additional parameters can be passed to
      <code>fun</code> through <code>...</code>, for instance to control the
      behaviour of <code>topN</code> in case of <code>NA</code> values.
      (Works also with <code>matrix</code> instances.)
    See also the <code><a href='nQuants.html'>nQuants</a></code> function to retrieve the
      actual number of retained peptides out of <code>n</code>.
    A complete use case using <code>topN</code> and <code>nQuants</code> is
      detailed in the <code>synapter</code> package vignette.
    <dt>filterNA</dt><dd><p>signature(object = "MSnSet", pNA = "numeric",
      pattern = "character", droplevels = "logical")</p></dd> This method
      subsets <code>object</code> by removing features that have (strictly)
      more than <code>pNA</code> percent of NA values. Default <code>pNA</code> is
      0, which removes any feature that exhibits missing data.
      The method can also be used with a character pattern composed of
      <code>0</code> or <code>1</code> characters only. A <code>0</code> represent a
      column/sample that is allowed a missing values, while
      columns/samples with and <code>1</code> must not have <code>NA</code>s.
    This method also accepts <code>matrix</code>
      instances. <code>droplevels</code> defines whether unused levels in the
      feature meta-data ought to be lost. Default is <code>TRUE</code>.
      See the <code>droplevels</code> method below.
    See also the <code><a href='plotNA-methods.html'>is.na.MSnSet</a></code> and <code><a href='plotNA-methods.html'>plotNA</a></code>
      methods for missing data exploration.
    <dt>filterZero</dt><dd><p>signature(object = "MSnSet", pNA = "numeric",
      pattern = "character", droplevels = "logical")</p></dd> As
      <code>filterNA</code>, but for zeros.
    <dt>filterMsLevel</dt><dd><p>signature(object = "MSnSet", msLevel. =
      "numeric", fcol = "character")</p></dd> Keeps only spectra with level
      <code>msLevel.</code>, as defined by the <code>fcol</code> feature variable
      (default is <code>"msLevel"</code>).
    <dt>log</dt><dd><p>signature(object = "MSnSet", base = "numeric")</p></dd> Log
      transforms <code>exprs(object)</code> using
      <code><a href='http://www.rdocumentation.org/packages/base/topics/Log'>base::log</a></code>. <code>base</code> (defaults is <code>e='exp(1)'</code>) must
      be a positive or complex number, the base with respect to which
      logarithms are computed.
    <dt>droplevels</dt><dd><p>signature(x = "MSnSet", ...)</p></dd>Drops the unused
      factor levels in the <code>featureData</code> slot. See
      <code>droplevels</code> for details.
    <dt>exprsToRatios</dt><dd><p><code>signature(object = "MSnSet", log =
	"logical")</code></p></dd>calculates all possible ratios between
      <code>object</code>'s columns/samples.
      See <code><a href='exprsToRatios-methods.html'>exprsToRatios</a></code> for more details.
    <dt>impute</dt><dd><p><code>signature(object = "MSnSet", ...)</code></p></dd>
      Performs data imputation on the <code>MSnSet</code> object.
      See <code><a href='impute-methods.html'>impute</a></code> for more details.
    <dt>trimws</dt><dd><p>signature(object = "MSnSet", ...)</p></dd>Trim leading and/or
      trailing white spaces in the feature data slot. Also available for
      <code>data.frame</code> objects. See <code>?base::<a href='http://www.rdocumentation.org/packages/base/topics/trimws'>trimws</a></code>
      for details.
    </dl>
    <p>Additional accessors for the experimental metadata
  (<code>experimentData</code> slot) are defined. See
  <code>"<a href='MIAPE-class.html'>MIAPE</a>"</code> for details.</p>
    
    <h2 class="hasAnchor" id="plotting"><a class="anchor" href="#plotting"></a>Plotting</h2>

    <p></p><dl class='dl-horizontal'>
    <dt>meanSdPlot</dt><dd><p><code>signature(object = "MSnSet")</code></p></dd> Plots row
      standard deviations versus row means. See
      <code>meanSdPlot</code> (<code>vsn</code> package) for more details.
    <dt>image</dt><dd><p><code>signature(x = "MSnSet", facetBy = "character",
	sOrderBy = "character", legend = "character", low = "character",
	high = "character", fnames = "logical", nmax =
	"numeric")</code></p></dd>Produces an heatmap of expression values in the
	<code>x</code> object. Simple horizontal facetting is enabled by
	passing a single character as <code>facetBy</code>. Arbitrary
	facetting can be performed manually by saving the return value
	of the method (see example below). Re-ordering of the samples is
	possible by providing the name of a phenotypic variable to
	<code>sOrderBy</code>. The title of the legend can be set with
	<code>legend</code> and the colours with the <code>low</code> and
	<code>high</code> arguments. If any negative value is detected in the
	data, the values are considered as log fold-changes and a
	divergent colour scale is used. Otherwise, a gradient from low
	to high is used. To scale the quantitative data in <code>x</code>
	prior to plotting, please see the <code>scale</code> method.
    When there are more than <code>nmax</code> (default is 50)
	features/rows, these are not printed. This behaviour can be
	controlled by setting <code>fnames</code> to <code>TRUE</code> (always
	print) or <code>FALSE</code> (never print). See examples below.
    The code is based on Vlad Petyuk's
	<code>vp.misc::image_msnset</code>. The previous version of this
	method is still available through the <code>image2</code> function.
    <dt>plotNA</dt><dd><p><code>signature(object = "MSnSet", pNA =
	"numeric")</code></p></dd>
	Plots missing data for an <code>MSnSet</code> instance. <code>pNA</code> is a
	<code>numeric</code> of length 1 that specifies the percentage
	of accepted missing data values per features. This value will be
	highlighted with a point on the figure, illustrating the overall
	percentage of NA values in the full data set and the number of
	proteins retained. Default is 1/2. See also
	<code><a href='plotNA-methods.html'>plotNA</a></code>.
    <dt>MAplot</dt><dd><p><code>signature(object = "MSnSet", log.it = "logical",
	base = "numeric", ...)</code></p></dd>
      Produces MA plots (Ratio as a function
      of average intensity) for the samples in <code>object</code>. If
      <code>ncol(object) == 2</code>, then one MA plot is produced using the
      <code>ma.plot</code> function from the <code>affy</code> package. If
      <code>object</code> has more than 2 columns, then
      <code>mva.pairs</code>. <code>log.it</code> specifies is the data
      should be log-transformed (default is <code>TRUE</code>) using
      <code>base</code>. Further <code>...</code> arguments will be passed to the
      respective functions.
    <dt>addIdentificationData</dt><dd><p><code>signature(object = "MSnSet", ...)</code>:
      Adds identification data to a <code>MSnSet</code> instance.
      See <code><a href='addIdentificationData-methods.html'>addIdentificationData</a></code> documentation for
      more details and examples.</p></dd>
    <dt>removeNoId</dt><dd><p><code>signature(object = "MSnSet", fcol =
	"pepseq", keep = NULL)</code>: Removes non-identified features. See
      <code><a href='removeNoId-methods.html'>removeNoId</a></code> documentation for more details and
      examples.</p></dd>
    <dt>removeMultipleAssignment</dt><dd><p><code>signature(object = "MSnSet",
	fcol = "nprot")</code>: Removes protein groups (or feature belong to
	protein groups) with more than one member. The latter is defined
	by extracting a feature variable (default is
	<code>"nprot"</code>). Also removes non-identified features/</p></dd>
    <dt>idSummary</dt><dd><p><code>signature(object = "MSnSet", ...)</code>: Prints a
      summary that lists the percentage of identified features per file
      (called <code>coverage</code>).</p></dd>
    </dl>
    
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    <p></p><dl class='dl-horizontal'>
    <dt>updateFvarLabels</dt><dd><p>signature(object, label, sep)</p></dd> This
      function updates <code>object</code>'s featureData variable labels by
      appending <code>label</code>. By default, <code>label</code> is the variable
      name and the separator <code>sep</code> is <code>.</code>.
    <dt>updateSampleNames</dt><dd><p>signature(object, label, sep)</p></dd> This
      function updates <code>object</code>'s sample names by appending
      <code>label</code>. By default, <code>label</code> is the variable name and
      the separator <code>sep</code> is <code>.</code>.
    <dt>updateFeatureNames</dt><dd><p>signature(object, label, sep)</p></dd> This
      function updates <code>object</code>'s feature names by appending
      <code>label</code>. By default, <code>label</code> is the variable name and
      the separator <code>sep</code> is <code>.</code>.
    <dt>ms2df</dt><dd><p>signature(x, fcols)</p></dd>Coerces the <code>MSnSet</code> instance
      to a <code>data.frame</code>. The direction of the data is retained and
      the feature variable labels that match <code>fcol</code> are appended to
      the expression values. See also <code><a href='mzRident2dfr.html'>as(x, "data.frame")</a></code> above.
    <dt>addMSnSetMetadata</dt><dd><p>signature(x, y)</p></dd>When coercing an
      <code>MSnSet</code> <code>y</code> to a <code>SummarizedExperiment</code> <code>x</code>
      with <code>x &lt;- as(y, "SummarizedExperiment")</code>, most of <code>y</code>'s
      metadata is lost. Only the file names, the processing log and the
      MSnbase version from the <code>processingData</code> slots are passed
      along. The <code>addMSnSetMetadata</code> function can be used to add
      the complete <code>processingData</code>, <code>experimentData</code> and
      <code>protocolData</code> slots. The downside of this is that MSnbase is
      now required to use the <code>SummarizedExperiment</code> object.
    </dl>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'></div>
    <div class='dont-index'><p><code>"eSet"</code>, <code>"ExpressionSet"</code> and
  <code><a href='quantify-methods.html'>quantify</a></code>. <code>MSnSet</code> quantitation values and
  annotation can be exported to a file with
  <code>write.exprs</code>. See <code><a href='readMSnSet.html'>readMSnSet</a></code> to
  create and <code>MSnSet</code> using data available in a spreadsheet or
  <code>data.frame</code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>msnset</span>)
<span class='no'>msnset</span> <span class='kw'>&lt;-</span> <span class='no'>msnset</span>[<span class='fl'>10</span>:<span class='fl'>15</span>]

<span class='fu'>exprs</span>(<span class='no'>msnset</span>)[<span class='fl'>1</span>, <span class='fu'>c</span>(<span class='fl'>1</span>, <span class='fl'>4</span>)] <span class='kw'>&lt;-</span> <span class='fl'>NA</span>
<span class='fu'>exprs</span>(<span class='no'>msnset</span>)[<span class='fl'>2</span>, <span class='fu'>c</span>(<span class='fl'>1</span>, <span class='fl'>2</span>)] <span class='kw'>&lt;-</span> <span class='fl'>NA</span>
<span class='fu'>is.na</span>(<span class='no'>msnset</span>)</div><div class='output co'>#&gt;     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117
#&gt; X18       TRUE      FALSE      FALSE       TRUE
#&gt; X19       TRUE       TRUE      FALSE      FALSE
#&gt; X2       FALSE      FALSE      FALSE      FALSE
#&gt; X20      FALSE      FALSE      FALSE      FALSE
#&gt; X21      FALSE      FALSE      FALSE      FALSE
#&gt; X22      FALSE      FALSE      FALSE      FALSE</div><div class='input'><span class='fu'>featureNames</span>(<span class='fu'>filterNA</span>(<span class='no'>msnset</span>, <span class='kw'>pNA</span> <span class='kw'>=</span> <span class='fl'>1</span>/<span class='fl'>4</span>))</div><div class='output co'>#&gt; [1] "X2"  "X20" "X21" "X22"</div><div class='input'><span class='fu'>featureNames</span>(<span class='fu'>filterNA</span>(<span class='no'>msnset</span>, <span class='kw'>pattern</span> <span class='kw'>=</span> <span class='st'>"0110"</span>))</div><div class='output co'>#&gt; [1] "X18" "X2"  "X20" "X21" "X22"</div><div class='input'>
<span class='no'>M</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='fl'>12</span>), <span class='fl'>4</span>)
<span class='no'>pd</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>otherpdata</span> <span class='kw'>=</span> <span class='no'>letters</span>[<span class='fl'>1</span>:<span class='fl'>3</span>])
<span class='no'>fd</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>otherfdata</span> <span class='kw'>=</span> <span class='no'>letters</span>[<span class='fl'>1</span>:<span class='fl'>4</span>])
<span class='no'>x0</span> <span class='kw'>&lt;-</span> <span class='fu'>MSnSet</span>(<span class='no'>M</span>, <span class='no'>fd</span>, <span class='no'>pd</span>)
<span class='fu'>sampleNames</span>(<span class='no'>x0</span>)</div><div class='output co'>#&gt; [1] "1" "2" "3"</div><div class='input'>
<span class='no'>M</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='fl'>12</span>), <span class='fl'>4</span>)
<span class='fu'>colnames</span>(<span class='no'>M</span>) <span class='kw'>&lt;-</span> <span class='no'>LETTERS</span>[<span class='fl'>1</span>:<span class='fl'>3</span>]
<span class='fu'>rownames</span>(<span class='no'>M</span>) <span class='kw'>&lt;-</span> <span class='fu'>paste0</span>(<span class='st'>"id"</span>, <span class='no'>LETTERS</span>[<span class='fl'>1</span>:<span class='fl'>4</span>])
<span class='no'>pd</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>otherpdata</span> <span class='kw'>=</span> <span class='no'>letters</span>[<span class='fl'>1</span>:<span class='fl'>3</span>])
<span class='fu'>rownames</span>(<span class='no'>pd</span>) <span class='kw'>&lt;-</span> <span class='fu'>colnames</span>(<span class='no'>M</span>)
<span class='no'>fd</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>otherfdata</span> <span class='kw'>=</span> <span class='no'>letters</span>[<span class='fl'>1</span>:<span class='fl'>4</span>])
<span class='fu'>rownames</span>(<span class='no'>fd</span>) <span class='kw'>&lt;-</span> <span class='fu'>rownames</span>(<span class='no'>M</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>MSnSet</span>(<span class='no'>M</span>, <span class='no'>fd</span>, <span class='no'>pd</span>)
<span class='fu'>sampleNames</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "A" "B" "C"</div><div class='input'>

<span class='co'>## Visualisation</span>

<span class='fu'>library</span>(<span class='st'>"pRolocdata"</span>)</div><div class='output co'>#&gt; <span class='error'>Error: package ‘MSnbase’ required by ‘pRolocdata’ could not be found</span></div><div class='input'><span class='fu'>data</span>(<span class='no'>dunkley2006</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: data set ‘dunkley2006’ not found</span></div><div class='input'><span class='fu'>image</span>(<span class='no'>dunkley2006</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006): object 'dunkley2006' not found</span></div><div class='input'><span class='co'>## Changing colours</span>
<span class='fu'>image</span>(<span class='no'>dunkley2006</span>, <span class='kw'>high</span> <span class='kw'>=</span> <span class='st'>"darkgreen"</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006, high = "darkgreen"): object 'dunkley2006' not found</span></div><div class='input'><span class='fu'>image</span>(<span class='no'>dunkley2006</span>, <span class='kw'>high</span> <span class='kw'>=</span> <span class='st'>"darkgreen"</span>, <span class='kw'>low</span> <span class='kw'>=</span> <span class='st'>"yellow"</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006, high = "darkgreen", low = "yellow"): object 'dunkley2006' not found</span></div><div class='input'><span class='co'>## Forcing feature names</span>
<span class='fu'>image</span>(<span class='no'>dunkley2006</span>, <span class='kw'>fnames</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006, fnames = TRUE): object 'dunkley2006' not found</span></div><div class='input'><span class='co'>## Facetting</span>
<span class='fu'>image</span>(<span class='no'>dunkley2006</span>, <span class='kw'>facetBy</span> <span class='kw'>=</span> <span class='st'>"replicate"</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006, facetBy = "replicate"): object 'dunkley2006' not found</span></div><div class='input'><span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fu'>image</span>(<span class='no'>dunkley2006</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006): object 'dunkley2006' not found</span></div><div class='input'><span class='fu'>library</span>(<span class='st'>"ggplot2"</span>) <span class='co'>## for facet_grid</span>
<span class='no'>p</span> + <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid'>facet_grid</a></span>(<span class='no'>replicate</span> ~ <span class='no'>membrane.prep</span>, <span class='kw'>scales</span> <span class='kw'>=</span> <span class='st'>'free'</span>, <span class='kw'>space</span> <span class='kw'>=</span> <span class='st'>'free'</span>)</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'p' not found</span></div><div class='input'><span class='no'>p</span> + <span class='fu'><a href='http://www.rdocumentation.org/packages/ggplot2/topics/facet_grid'>facet_grid</a></span>(<span class='no'>markers</span> ~ <span class='no'>replicate</span>)</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'p' not found</span></div><div class='input'><span class='co'>## Fold-changes</span>
<span class='no'>dd</span> <span class='kw'>&lt;-</span> <span class='no'>dunkley2006</span></div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'dunkley2006' not found</span></div><div class='input'><span class='fu'>exprs</span>(<span class='no'>dd</span>) <span class='kw'>&lt;-</span> <span class='fu'>exprs</span>(<span class='no'>dd</span>) - <span class='fl'>0.25</span></div><div class='output co'>#&gt; <span class='error'>Error in exprs(dd): object 'dd' not found</span></div><div class='input'><span class='fu'>image</span>(<span class='no'>dd</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dd): object 'dd' not found</span></div><div class='input'><span class='fu'>image</span>(<span class='no'>dd</span>, <span class='kw'>low</span> <span class='kw'>=</span> <span class='st'>"green"</span>, <span class='kw'>high</span> <span class='kw'>=</span> <span class='st'>"red"</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dd, low = "green", high = "red"): object 'dd' not found</span></div><div class='input'><span class='co'>## Feature names are displayed by default for smaller data</span>
<span class='no'>dunkley2006</span> <span class='kw'>&lt;-</span> <span class='no'>dunkley2006</span>[<span class='fl'>1</span>:<span class='fl'>25</span>, ]</div><div class='output co'>#&gt; <span class='error'>Error in eval(expr, envir, enclos): object 'dunkley2006' not found</span></div><div class='input'><span class='fu'>image</span>(<span class='no'>dunkley2006</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006): object 'dunkley2006' not found</span></div><div class='input'><span class='fu'>image</span>(<span class='no'>dunkley2006</span>, <span class='kw'>legend</span> <span class='kw'>=</span> <span class='st'>"hello"</span>)</div><div class='output co'>#&gt; <span class='error'>Error in image(dunkley2006, legend = "hello"): object 'dunkley2006' not found</span></div><div class='input'>

<span class='co'>## Coercion</span>
<span class='kw'>if</span> (<span class='fu'>require</span>(<span class='st'>"SummarizedExperiment"</span>)) {
    <span class='fu'>data</span>(<span class='no'>msnset</span>)
    <span class='no'>se</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='mzRident2dfr.html'>as</a></span>(<span class='no'>msnset</span>, <span class='st'>"SummarizedExperiment"</span>)
    <span class='fu'>metadata</span>(<span class='no'>se</span>) <span class='co'>## only logging</span>
    <span class='no'>se</span> <span class='kw'>&lt;-</span> <span class='fu'>addMSnSetMetadata</span>(<span class='no'>se</span>, <span class='no'>msnset</span>)
    <span class='fu'>metadata</span>(<span class='no'>se</span>) <span class='co'>## all metadata</span>
    <span class='no'>msnset2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='mzRident2dfr.html'>as</a></span>(<span class='no'>se</span>, <span class='st'>"MSnSet"</span>)
    <span class='fu'><a href='pSet-class.html'>processingData</a></span>(<span class='no'>msnset2</span>)
}</div><div class='output co'>#&gt; <span class='message'>Loading required package: SummarizedExperiment</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘S4Vectors’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomeInfoDb</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: DelayedArray</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: matrixStats</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘matrixStats’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:Biobase’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyMissing, rowMedians</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘DelayedArray’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:matrixStats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    aperm, apply</span></div><div class='output co'>#&gt; - - - Processing information - - -
#&gt; Data loaded: Wed May 11 18:54:39 2011 
#&gt; iTRAQ4 quantification by trapezoidation: Wed Apr  1 21:41:53 2015 
#&gt; Subset [55,4][6,4] Thu May 17 15:33:26 2018 
#&gt;  MSnbase version: 1.1.22 </div><div class='input'>
<span class='fu'><a href='mzRident2dfr.html'>as</a></span>(<span class='no'>msnset</span>, <span class='st'>"ExpressionSet"</span>)</div><div class='output co'>#&gt; ExpressionSet (storageMode: lockedEnvironment)
#&gt; assayData: 6 features, 4 samples 
#&gt;   element names: exprs 
#&gt; protocolData: none
#&gt; phenoData
#&gt;   sampleNames: iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117
#&gt;   varLabels: mz reporters
#&gt;   varMetadata: labelDescription
#&gt; featureData
#&gt;   featureNames: X18 X19 ... X22 (6 total)
#&gt;   fvarLabels: spectrum ProteinAccession ... collision.energy (15 total)
#&gt;   fvarMetadata: labelDescription
#&gt; experimentData: use 'experimentData(object)'
#&gt; Annotation: No annotation </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#objects-from-the-class">Objects from the Class</a></li>

      <li><a href="#slots">Slots</a></li>

      <li><a href="#extends">Extends</a></li>

      <li><a href="#methods">Methods</a></li>

      <li><a href="#plotting">Plotting</a></li>

      <li><a href="#functions">Functions</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
  Laurent Gatto &lt;lg390@cam.ac.uk&gt;

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

