<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MSnbase: MS data processing, visualisation and quantification • MSnbase</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MSnbase: MS data processing, visualisation and quantification">
<meta property="og:description" content="MSnbase">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.29.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/v04-benchmarking.html">MSnbase benchmarking</a>
    </li>
    <li>
      <a href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/MSnbase/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>MSnbase: MS data processing, visualisation and
quantification</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
                                    <h4 data-toc-skip class="author">Johannes
Rainer</h4>
                                    <h4 data-toc-skip class="author">Sebastian
Gibb</h4>
                        
      
      <small class="dont-index">Source: <a href="https://github.com/lgatto/MSnbase/blob/HEAD/vignettes/v01-MSnbase-demo.Rmd" class="external-link"><code>vignettes/v01-MSnbase-demo.Rmd</code></a></small>
      <div class="hidden name"><code>v01-MSnbase-demo.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette describes the functionality implemented in the
      <em>MSnbase</em> package. <em>MSnbase</em> aims at (1)
      facilitating the import, processing, visualisation and
      quantification of mass spectrometry data into the R environment
      <span class="citation">(R Development Core Team 2011)</span> by
      providing specific data classes and methods and (2) enabling the
      utilisation of throughput-high data analysis pipelines provided by
      the Bioconductor <span class="citation">(Gentleman et al.
      2004)</span> project.</p>
    </div>
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 class="unnumbered" id="foreword">Foreword<a class="anchor" aria-label="anchor" href="#foreword"></a>
</h2>
<p>This software is free and open-source software. If you use it, please
support the project by citing it in publications:</p>
<blockquote>
<p>Gatto L, Lilley KS. MSnbase-an R/Bioconductor package for isobaric
tagged mass spectrometry data visualization, processing and
quantitation. Bioinformatics. 2012 Jan 15;28(2):288-9. doi: <a href="https://doi.org/10.1093/bioinformatics/btr645" class="external-link">10.1093/bioinformatics/btr645</a>.
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22113085" class="external-link">22113085</a>.</p>
</blockquote>
<blockquote>
<p><em><code>MSnbase</code>, efficient and elegant R-based processing
and visualisation of raw mass spectrometry data</em>. Laurent Gatto,
Sebastian Gibb, Johannes Rainer. bioRxiv 2020.04.29.067868; doi: <a href="https://doi.org/10.1101/2020.04.29.067868" class="external-link uri">https://doi.org/10.1101/2020.04.29.067868</a></p>
</blockquote>
</div>
<div class="section level2">
<h2 class="unnumbered" id="questions-and-bugs">Questions and bugs<a class="anchor" aria-label="anchor" href="#questions-and-bugs"></a>
</h2>
<p>For bugs, typos, suggestions or other questions, please file an issue
in our tracking system (<a href="https://github.com/lgatto/MSnbase/issues" class="external-link uri">https://github.com/lgatto/MSnbase/issues</a>) providing as
much information as possible, a reproducible example and the output of
<code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code>.</p>
<p>If you don’t have a GitHub account or wish to reach a broader
audience for general questions about proteomics analysis using R, you
may want to use the Bioconductor support site: <a href="https://support.bioconductor.org/" class="external-link uri">https://support.bioconductor.org/</a>.</p>
</div>
<div class="section level2">
<h2 id="sec:intro">Introduction<a class="anchor" aria-label="anchor" href="#sec:intro"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
<span class="citation">(Laurent Gatto and Lilley 2012)</span> aims are
providing a reproducible research framework to proteomics data analysis.
It should allow researcher to easily mine mass spectrometry data,
explore the data and its statistical properties and visually display
these.</p>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
also aims at being compatible with the infrastructure implemented in
Bioconductor, in particular <em><a href="https://bioconductor.org/packages/3.18/Biobase" class="external-link">Biobase</a></em>.
As such, classes developed specifically for proteomics mass spectrometry
data are based on the <em>eSet</em> and <em>ExpressionSet</em> classes.
The main goal is to assure seamless compatibility with existing meta
data structure, accessor methods and normalisation techniques.</p>
<p>This vignette illustrates <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
utility using a dummy data sets provided with the package without
describing the underlying data structures. More details can be found in
the package, classes, method and function documentations. A description
of the classes is provided in the <em>MSnbase-development</em>
vignette<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<div class="section level3">
<h3 id="speed-and-memory-requirements">Speed and memory requirements<a class="anchor" aria-label="anchor" href="#speed-and-memory-requirements"></a>
</h3>
<p>Raw mass spectrometry file are generally several hundreds of MB large
and most of this is used for binary raw spectrum data. As such, data
containers can easily grow very large and thus require large amounts of
RAM. This requirement is being tackled by avoiding to load the raw data
into memory and using on-disk random access to the content of
<code>mzXML</code>/<code>mzML</code> data files on demand. When focusing
on reporter ion quantitation, a direct solution for this is to trim the
spectra using the <code>trimMz</code> method to select the area of
interest and thus substantially reduce the size of the
<code>Spectrum</code> objects. This is illustrated in section
@ref(sec:trim).</p>
<p><strong>Parallel processing</strong> The independent handling of
spectra is ideally suited for parallel processing. The
<code>quantify</code> method for example performs reporter peaks
quantitation in parallel.</p>
<p>Parallel support is provided by the <em><a href="https://bioconductor.org/packages/3.18/BiocParallel" class="external-link">BiocParallel</a></em>
and various backends including multicore (forking, default on Linux),
simple networf network of workstations (SNOW, default on Windows) using
sockets, forking or MPI among others. We refer readers to the
documentation in <em><a href="https://bioconductor.org/packages/3.18/BiocParallel" class="external-link">BiocParallel</a></em>.
Automatic parallel processing of spectra is only established for a
certain number of spectra (per file). This value (default is 1000) can
be set with the <code>setMSnbaseParallelThresh</code> function.</p>
<p>In sock-based parallel processing, the main worker process has to
start new R instances and connect to them via sock. Sometimes these
connections can not be established and the processes get stuck. To test
this, users can disable parallel processing by disabling parallel
processing with <code>register(SerialParam())</code>. To avoid these
deadlocks, it is possible to initiate the parallel processing setup
explicitly at the beginning of the script using, for example</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/RevolutionAnalytics/doparallel" class="external-link">"doParallel"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">registerDoParallel</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span> <span class="co">## using 3 slave nodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/DoparParam-class.html" class="external-link">DoparParam</a></span><span class="op">(</span><span class="op">)</span>, default <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## rest of script comes below</span></span></code></pre></div>
<p><strong>On-disk access</strong> Developmenets in version 2 of the
package have solved the memory issue by implementing and
<em>on-disk</em> version the of data class storing raw data
(<em>MSnExp</em>, see section @ref(sec:msnexp)), where the spectra a
accessed on-disk only when required. The <em>benchmarking</em> vignette
compares the on-disk and in-memory implemenatations<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. See details
below.</p>
</div>
</div>
<div class="section level2">
<h2 id="sec:data">Data structure and content<a class="anchor" aria-label="anchor" href="#sec:data"></a>
</h2>
<div class="section level3">
<h3 id="sec:io">Importing experiments<a class="anchor" aria-label="anchor" href="#sec:io"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
is able to import raw MS data stored in one of the
<code>XML</code>-based formats as well as peak lists in the
<code>mfg</code> format<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>.</p>
<p><strong>Raw data</strong> The <code>XML</code>-based formats,
<code>mzXML</code> <span class="citation">(Pedrioli et al. 2004)</span>,
<code>mzData</code> <span class="citation">(Orchard et al. 2007)</span>
and <code>mzML</code> <span class="citation">(Martens et al.
2010)</span> can be imported with the <code>readMSData</code> function,
as illustrated below (see <code><a href="../reference/readMSData.html">?readMSData</a></code> for more details). To
make use of the new <em>on-disk</em> implementation, set
<code>mode = "onDisk"</code> in <code>readMSData</code> rather than
using the default <code>mode = "inMemory"</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"MSnbase"</span>, dir <span class="op">=</span> <span class="st">"extdata"</span><span class="op">)</span>,</span>
<span>            full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"mzXML$"</span><span class="op">)</span></span>
<span><span class="va">rawdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMSData.html">readMSData</a></span><span class="op">(</span><span class="va">file</span>, msLevel <span class="op">=</span> <span class="fl">2</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Only spectra of a given MS level can be loaded at a time by setting
the <code>msLevel</code> parameter accordingly in
<code>readMSData</code> and <em>in-memory</em> data. In this document,
we will use the <code>itraqdata</code> data set, provided with <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>.
It includes feature metadata, accessible with the <code>fData</code>
accessor. The metadata includes identification data for the 55 MS2
spectra.</p>
<p>Version 2.0 and later of <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
provide a new <em>on-disk</em> data storage model (see the
<em>benchmarking</em> vignette for more details). The new data backend
is compatible with the orignal <em>in-memory</em> model. To make use of
the new infrastructure, read your raw data by setting the
<code>mode</code> argument to <code>"onDisk"</code> (the default is
still <code>"inMemory"</code> but is likely to change in the future).
The new <em>on-disk</em> implementation supports several MS levels in a
single raw data object. All existing operations work irrespective of the
backend.</p>
<p><strong>Peak lists</strong> can often be exported after spectrum
processing from vendor-specific software and are also used as input to
search engines. Peak lists in <code>mgf</code> format can be imported
with the function <code>readMgfData</code> (see
<code><a href="../reference/readMgfData.html">?readMgfData</a></code> for details) to create experiment objects.
Experiments or individual spectra can be exported to an <code>mgf</code>
file with the <code>writeMgfData</code> methods (see
<code><a href="../reference/writeMgfData-methods.html">?writeMgfData</a></code> for details and examples).</p>
<p><strong>Experiments with multiple runs</strong> Although it is
possible to load and process multiple files serially and later merge the
resulting quantitation data as show in section @ref(sec:combine), it is
also feasible to load several raw data files at once. Here, we report
the analysis of an LC-MSMS experiment were 14 liquid chromatography (LC)
fractions were loaded in memory using <code>readMSData</code> on a
32-cores servers with 128 Gb of RAM. It took about 90 minutes to read
the 14 uncentroided <code>mzXML</code> raw files (4.9 Gb on disk in
total) and create a 3.3 Gb raw data object (an <em>MSnExp</em> instance,
see next section). Quantitation of 9 reporter ions (<em>iTRAQ9</em>
object, see @ref(sec:reporterions)) for 88690 features was performed in
parallel on 16 processors and took 76 minutes. The resulting
quantitation data was only 22.1 Mb and could easily be further
processed. These number are based on the older <em>in-memory</em>
implementation. As shown in the <em>benchmarking</em> vignette, using
<em>on-disk</em> data greatly reduces memory requirement and computation
time.</p>
<p>See also section @ref(sec:io2) to import quantitative data stored in
spreadsheets into R for further processing using <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>.
The <em>MSnbase-io</em>vignette[in R, open it with
<code>vignette("MSnbase-io")</code> or read it online <a href="https://bioconductor.org/packages/devel/bioc/vignettes/MSnbase/inst/doc/MSnbase-io.html" class="external-link">here</a>]
gives a general overview of <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>’s
input/ouput capabilites.</p>
<p>See section @ref(sec:io3) for importing chromatographic data of
SRM/MRM experiments.</p>
</div>
<div class="section level3">
<h3 id="exporting-experimentsms-data">Exporting experiments/MS data<a class="anchor" aria-label="anchor" href="#exporting-experimentsms-data"></a>
</h3>
<p><code>MSnbase</code> supports also to write <code>MSnExp</code> or
<code>OnDiskMSnExp</code> objects to <code>mzML</code> or
<code>mzXML</code> files using the <code>writeMSData</code> function.
This is specifically useful in workflows in which the MS data was
heavily manipulated. Presently, each sample/file is exported into one
file.</p>
<p>Below we write the data in <code>mzML</code> format to a temporary
file. By setting the optional parameter <code>copy = TRUE</code> general
metadata (such as instrument info or all data processing descriptions)
are copied over from the originating file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/writeMSData.html">writeMSData</a></span><span class="op">(</span><span class="va">rawdata</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">".mzML"</span><span class="op">)</span>, copy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sec:msnexp">MS experiments<a class="anchor" aria-label="anchor" href="#sec:msnexp"></a>
</h3>
<p>Raw data is contained in <em>MSnExp</em> objects, that stores all the
spectra of an experiment, as defined by one or multiple raw data
files.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://lgatto.github.io/MSnbase" class="external-link">"MSnbase"</a></span><span class="op">)</span></span>
<span><span class="va">itraqdata</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn experiment data ("MSnExp")</span></span>
<span><span class="co">## Object size in memory: 1.9 Mb</span></span>
<span><span class="co">## - - - Spectra data - - -</span></span>
<span><span class="co">##  MS level(s): 2 </span></span>
<span><span class="co">##  Number of spectra: 55 </span></span>
<span><span class="co">##  MSn retention times: 19:09 - 50:18 minutes</span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data loaded: Wed May 11 18:54:39 2011 </span></span>
<span><span class="co">## Updated from version 0.3.0 to 0.3.1 [Fri Jul  8 20:23:25 2016] </span></span>
<span><span class="co">##  MSnbase version: 1.1.22 </span></span>
<span><span class="co">## - - - Meta data  - - -</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   rowNames: 1</span></span>
<span><span class="co">##   varLabels: sampleNames sampleNumbers</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## Loaded from:</span></span>
<span><span class="co">##   dummyiTRAQ.mzXML </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: X1 X10 ... X9 (55 total)</span></span>
<span><span class="co">##   fvarLabels: spectrum ProteinAccession ProteinDescription</span></span>
<span><span class="co">##     PeptideSequence</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     spectrum ProteinAccession                       ProteinDescription</span></span>
<span><span class="co">## X1         1              BSA                     bovine serum albumin</span></span>
<span><span class="co">## X10       10          ECA1422 glucose-1-phosphate cytidylyltransferase</span></span>
<span><span class="co">## X11       11          ECA4030         50S ribosomal subunit protein L4</span></span>
<span><span class="co">## X12       12          ECA3882                   chaperone protein DnaK</span></span>
<span><span class="co">## X13       13          ECA1364      succinyl-CoA synthetase alpha chain</span></span>
<span><span class="co">## X14       14          ECA0871              NADP-dependent malic enzyme</span></span>
<span><span class="co">##     PeptideSequence</span></span>
<span><span class="co">## X1           NYQEAK</span></span>
<span><span class="co">## X10 VTLVDTGEHSMTGGR</span></span>
<span><span class="co">## X11           SPIWR</span></span>
<span><span class="co">## X12        TAIDDALK</span></span>
<span><span class="co">## X13          SILINK</span></span>
<span><span class="co">## X14    DFEVVNNESDPR</span></span></code></pre>
<p>As illustrated above, showing the experiment textually displays it’s
content:</p>
<ul>
<li><p>Information about the raw data, i.e. the spectra.</p></li>
<li><p>Specific information about the experiment processing<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a> and
package version. This slot can be accessed with the
<code>processingData</code> method.</p></li>
<li>
<p>Other meta data, including experimental phenotype, file name(s)
used to import the data, protocol data, information about features
(individual spectra here) and experiment data. Most of these are
implemented as in the <em>eSet</em> class and are described in more
details in their respective manual pages. See <code><a href="../reference/MSnExp-class.html">?MSnExp</a></code> and
references therein for additional background information.</p>
<p>The experiment meta data associated with an <em>MSnExp</em>
experiment is of class <em>MIAPE</em>. It stores general information
about the experiment as well as MIAPE (Minimum Information About a
Proteomics Experiment) information <span class="citation">Chris F.
Taylor et al. (2008)</span>. This meta-data can be accessed with the
<code>experimentData</code> method. When available, a summary of
MIAPE-MS data can be printed with the <code>msInfo</code> method. See
<code><a href="../reference/MIAPE-class.html">?MIAPE</a></code> for more details.</p>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="sec:spectra">Spectra objects<a class="anchor" aria-label="anchor" href="#sec:spectra"></a>
</h3>
<p>The raw data is composed of the 55 MS spectra. The spectra are named
individually (X1, X10, X11, X12, X13, X14, …) and stored in a
<code>environment</code>. They can be accessed individually with
<code>itraqdata[["X1"]]</code> or <code>itraqdata[[1]]</code>, or as a
list with <code>spectra(itraqdata)</code>. As we have loaded our
experiment specifying <code>msLevel=2</code>, the spectra will all be of
level 2 (or higher, if available).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="va">itraqdata</span><span class="op">[[</span><span class="st">"X1"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sp</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "Spectrum2"</span></span>
<span><span class="co">##  Precursor: 520.7833 </span></span>
<span><span class="co">##  Retention time: 19:09 </span></span>
<span><span class="co">##  Charge: 2 </span></span>
<span><span class="co">##  MSn level: 2 </span></span>
<span><span class="co">##  Peaks count: 1922 </span></span>
<span><span class="co">##  Total ion count: 26413754</span></span></code></pre>
<p>Attributes of individual spectra or of all spectra of an experiment
can be accessed with their respective methods:
<code>precursorCharge</code> for the precursor charge,
<code>rtime</code> for the retention time, <code>mz</code> for the MZ
values, <code>intensity</code> for the intensities, … see the
<em>Spectrum</em>, <em>Spectrum1</em> and <em>Spectrum2</em> manuals for
more details.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">peaksCount</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1922</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/Spectrum-class.html">peaksCount</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   X1  X10  X11  X12  X13  X14 </span></span>
<span><span class="co">## 1922 1376 1571 2397 2574 1829</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">rtime</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1149.31</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/Spectrum-class.html">rtime</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      X1     X10     X11     X12     X13     X14 </span></span>
<span><span class="co">## 1149.31 1503.03 1663.61 1663.86 1664.08 1664.32</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="sec:reporterions">Reporter ions<a class="anchor" aria-label="anchor" href="#sec:reporterions"></a>
</h3>
<p>Reporter ions are defined with the <em>ReporterIons</em> class.
Specific peaks of interest are defined by a MZ value, a with around the
expected MZ and a name (and optionally a colour for plotting, see
section @ref(sec:plotting)). <em>ReporterIons</em> instances are
required to quantify reporter peaks in <em>MSnExp</em> experiments.
Instances for the most commonly used isobaric tags like iTRAQ 4-plex and
8-plex and TMT 6- and 10-plex tags are already defined in <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>.
See <code><a href="../reference/ReporterIons-class.html">?ReporterIons</a></code> for details about how to generate new
<em>ReporterIons</em> objects.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">iTRAQ4</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "ReporterIons"</span></span>
<span><span class="co">## iTRAQ4: '4-plex iTRAQ' with 4 reporter ions</span></span>
<span><span class="co">##  - [iTRAQ4.114] 114.1112 +/- 0.05 (red)</span></span>
<span><span class="co">##  - [iTRAQ4.115] 115.1083 +/- 0.05 (green)</span></span>
<span><span class="co">##  - [iTRAQ4.116] 116.1116 +/- 0.05 (blue)</span></span>
<span><span class="co">##  - [iTRAQ4.117] 117.115 +/- 0.05 (yellow)</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">TMT16</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "ReporterIons"</span></span>
<span><span class="co">## TMT16HCD: '16-plex TMT HCD' with 16 reporter ions</span></span>
<span><span class="co">##  - [126] 126.1277 +/- 0.002 (#E55400)</span></span>
<span><span class="co">##  - [127N] 127.1248 +/- 0.002 (#E28100)</span></span>
<span><span class="co">##  - [127C] 127.1311 +/- 0.002 (#DFAC00)</span></span>
<span><span class="co">##  - [128N] 128.1281 +/- 0.002 (#DDD700)</span></span>
<span><span class="co">##  - [128C] 128.1344 +/- 0.002 (#B5DA00)</span></span>
<span><span class="co">##  - [129N] 129.1315 +/- 0.002 (#87D8000)</span></span>
<span><span class="co">##  - [129C] 129.1378 +/- 0.002 (#5BD500)</span></span>
<span><span class="co">##  - [130N] 130.1348 +/- 0.002 (#30D300)</span></span>
<span><span class="co">##  - [130C] 130.1411 +/- 0.002 (#05D000)</span></span>
<span><span class="co">##  - [131N] 131.1382 +/- 0.002 (#00CE23)</span></span>
<span><span class="co">##  - [131C] 131.1445 +/- 0.002 (#00CB4B)</span></span>
<span><span class="co">##  - [132N] 132.1415 +/- 0.002 (#00C972)</span></span>
<span><span class="co">##  - [132C] 132.1479 +/- 0.002 (#00C699)</span></span>
<span><span class="co">##  - [133N] 133.1449 +/- 0.002 (#00C4BE)</span></span>
<span><span class="co">##  - [133C] 133.1512 +/- 0.002 (#00A0C1)</span></span>
<span><span class="co">##  - [134N] 134.1482 +/- 0.002 (#0078BF)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="sec:chromatograms">Chromatogram objects<a class="anchor" aria-label="anchor" href="#sec:chromatograms"></a>
</h3>
<p>Chromatographic data, i.e. intensity values along the retention time
dimension for a given <span class="math inline">\(m/z\)</span>
range/slice, can be extracted with the <code>chromatogram</code> method.
Below we read a file from the <code>msdata</code> package and extract
the (MS level 1) chromatogram. Without providing an <span class="math inline">\(m/z\)</span> and a retention time range the
function returns the total ion chromatogram (TIC) for each file within
the <code>MSnExp</code> or <code>OnDiskMSnExp</code> object. See also
section @ref(sec:io3) for importing chromatographic data from SRM/MRM
experiments.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"microtofq/MM14.mzML"</span>, package <span class="op">=</span> <span class="st">"msdata"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mtof</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMSData.html">readMSData</a></span><span class="op">(</span><span class="va">f</span>, mode <span class="op">=</span> <span class="st">"onDisk"</span><span class="op">)</span></span>
<span><span class="va">mtof_tic</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-MSnExp-method.html">chromatogram</a></span><span class="op">(</span><span class="va">mtof</span><span class="op">)</span></span>
<span><span class="va">mtof_tic</span></span></code></pre></div>
<pre><code><span><span class="co">## MChromatograms with 1 row and 1 column</span></span>
<span><span class="co">##           MM14.mzML</span></span>
<span><span class="co">##      &lt;Chromatogram&gt;</span></span>
<span><span class="co">## [1,]    length: 112</span></span>
<span><span class="co">## phenoData with 1 variables</span></span>
<span><span class="co">## featureData with 1 variables</span></span></code></pre>
<p>Chromatographic data, represented by the intensity-retention time
duplets, is stored in the <code>Chromatogram</code> object. The
<code>chromatogram</code> method returns a <code>Chromatograms</code>
object (note the <em>s</em>) which holds multiple
<code>Chromatogram</code> objects and arranges them in a two-dimensional
grid with columns representing files/samples of the <code>MSnExp</code>
or <code>OnDiskMSnExp</code> object and rows <span class="math inline">\(m/z\)</span>-retention time ranges. In the example
above the <code>Chromatograms</code> object contains only a single
<code>Chromatogram</code> object. Below we access this chromatogram
object. Similar to the <code>Spectrum</code> objects,
<code>Chromatogram</code> objects provide the accessor functions
<code>intensity</code> and <code>rtime</code> to access the data, as
well as the <code>mz</code> function, that returns the <span class="math inline">\(m/z\)</span> range of the chromatogram.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtof_tic</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class: Chromatogram</span></span>
<span><span class="co">## Intensity values aggregated using: sum </span></span>
<span><span class="co">## length of object: 112</span></span>
<span><span class="co">## from file: 1</span></span>
<span><span class="co">## mz range: [94.80679, 1004.962]</span></span>
<span><span class="co">## rt range: [270.334, 307.678]</span></span>
<span><span class="co">## MS level: 1</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/Spectrum-class.html">intensity</a></span><span class="op">(</span><span class="va">mtof_tic</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S001 F1.S002 F1.S003 F1.S004 F1.S005 F1.S006 </span></span>
<span><span class="co">##   64989   67445   77843  105097  155609  212760</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/Spectrum-class.html">rtime</a></span><span class="op">(</span><span class="va">mtof_tic</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S001 F1.S002 F1.S003 F1.S004 F1.S005 F1.S006 </span></span>
<span><span class="co">## 270.334 270.671 271.007 271.343 271.680 272.016</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">mz</a></span><span class="op">(</span><span class="va">mtof_tic</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]   94.80679 1004.96155</span></span></code></pre>
<p>To extract the base peak chromatogram (the largest peak along the
<span class="math inline">\(m/z\)</span> dimension for each retention
time/spectrum) we set the <code>aggregationFun</code> argument to
<code>"max"</code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtof_bpc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-MSnExp-method.html">chromatogram</a></span><span class="op">(</span><span class="va">mtof</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span></code></pre></div>
<p>See the <code>Chromatogram</code> help page and the vignettes from
the <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em> package
for more details and use cases, also on how to extract chromatograms for
specific ions.</p>
</div>
</div>
<div class="section level2">
<h2 id="sec:plotting">Plotting raw data<a class="anchor" aria-label="anchor" href="#sec:plotting"></a>
</h2>
<div class="section level3">
<h3 id="sec:msmaps">MS data space<a class="anchor" aria-label="anchor" href="#sec:msmaps"></a>
</h3>
<p>The <em>MSmap</em> class can be used to isolate specific slices of
interest from a complete MS acquisition by specifying <span class="math inline">\(m/z\)</span> and retention time ranges. One needs
a raw data file in a format supported by <em><a href="https://bioconductor.org/packages/3.18/mzR" class="external-link">mzR</a></em>’s
<code>openMSfile</code> (<code>mzML</code>, <code>mzXML</code>, …).
Below we first download a raw data file from the PRIDE repository and
create an <em>MSmap</em> containing all the MS1 spectra between acquired
between 30 and 35 minutes and peaks between 521 and 523 <span class="math inline">\(m/z\)</span>. See <code><a href="../reference/MSmap-class.html">?MSmap</a></code> for
details.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## downloads the data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/rpx" class="external-link">"rpx"</a></span><span class="op">)</span></span>
<span><span class="va">px1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html" class="external-link">PXDataset</a></span><span class="op">(</span><span class="st">"PXD000001"</span><span class="op">)</span></span>
<span><span class="va">mzf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpx/man/PXDataset2.html" class="external-link">pxget</a></span><span class="op">(</span><span class="va">px1</span>, <span class="fl">7</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## reads the data</span></span>
<span><span class="va">ms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mzR/man/openMSfile.html" class="external-link">openMSfile</a></span><span class="op">(</span><span class="va">mzf</span><span class="op">)</span></span>
<span><span class="va">hd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pSet-class.html">header</a></span><span class="op">(</span><span class="va">ms</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## a set of spectra of interest: MS1 spectra eluted</span></span>
<span><span class="co">## between 30 and 35 minutes retention time</span></span>
<span><span class="va">ms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">hd</span><span class="op">$</span><span class="va">msLevel</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">rtsel</span> <span class="op">&lt;-</span> <span class="va">hd</span><span class="op">$</span><span class="va">retentionTime</span><span class="op">[</span><span class="va">ms1</span><span class="op">]</span> <span class="op">/</span> <span class="fl">60</span> <span class="op">&gt;</span> <span class="fl">30</span> <span class="op">&amp;</span></span>
<span>    <span class="va">hd</span><span class="op">$</span><span class="va">retentionTime</span><span class="op">[</span><span class="va">ms1</span><span class="op">]</span> <span class="op">/</span> <span class="fl">60</span> <span class="op">&lt;</span> <span class="fl">35</span></span>
<span></span>
<span><span class="co">## the map</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSmap-class.html">MSmap</a></span><span class="op">(</span><span class="va">ms</span>, <span class="va">ms1</span><span class="op">[</span><span class="va">rtsel</span><span class="op">]</span>, <span class="fl">521</span>, <span class="fl">523</span>, <span class="fl">.005</span>, <span class="va">hd</span>, zeroIsNA <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "MSmap"</span></span>
<span><span class="co">##  Map [75, 401]</span></span>
<span><span class="co">##   [1]  Retention time: 30:01 - 34:58 </span></span>
<span><span class="co">##   [2]  M/Z: 521 - 523 (res 0.005)</span></span></code></pre>
<p>The <code>M</code> map object can be rendered as a heatmap with
<code>plot</code>, as shown on figure @ref(fig:mapheat).</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot-methods.html">plot</a></span><span class="op">(</span><span class="va">M</span>, aspect <span class="op">=</span> <span class="fl">1</span>, allTicks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/mapheat-1.png" alt="Heat map of a chunk of the MS data." width="700"><p class="caption">
Heat map of a chunk of the MS data.
</p>
</div>
<p>One can also render the data in 3 dimension with the
<code>plot3D</code> function, as show on figure @ref(fig:map3d).</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MSmap-class.html">plot3D</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/map3d-1.png" alt="3 dimensional represention of MS map data." width="700"><p class="caption">
3 dimensional represention of MS map data.
</p>
</div>
<p>To produce figure @ref(fig:map3d2), we create a second <em>MSmap</em>
object containing the first two MS1 spectra of the first map (object
<code>M</code> above) and all intermediate MS2 spectra and display <span class="math inline">\(m/z\)</span> values between 100 and 1000.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">i</span> <span class="op">&lt;-</span> <span class="va">ms1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">rtsel</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">j</span> <span class="op">&lt;-</span> <span class="va">ms1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">rtsel</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">M2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSmap-class.html">MSmap</a></span><span class="op">(</span><span class="va">ms</span>, <span class="va">i</span><span class="op">:</span><span class="va">j</span>, <span class="fl">100</span>, <span class="fl">1000</span>, <span class="fl">1</span>, <span class="va">hd</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M2</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "MSmap"</span></span>
<span><span class="co">##  Map [12, 901]</span></span>
<span><span class="co">##   [1]  Retention time: 30:01 - 30:05 </span></span>
<span><span class="co">##   [2]  M/Z: 100 - 1000 (res 1)</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MSmap-class.html">plot3D</a></span><span class="op">(</span><span class="va">M2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/map3d2-1.png" alt="3 dimensional represention of MS map data. MS1 and MS2 spectra are coloured in blue and magenta respectively." width="700"><p class="caption">
3 dimensional represention of MS map data. MS1 and MS2 spectra are
coloured in blue and magenta respectively.
</p>
</div>
</div>
<div class="section level3">
<h3 id="sec:specplots">MS Spectra<a class="anchor" aria-label="anchor" href="#sec:specplots"></a>
</h3>
<p>Spectra can be plotted individually or as part of (subset)
experiments with the <code>plot</code> method. Full spectra can be
plotted (using <code>full=TRUE</code>), specific reporter ions of
interest (by specifying with reporters with
<code>reporters=iTRAQ4</code> for instance) or both (see figure
@ref(fig:spectrumPlot)).</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot-methods.html">plot</a></span><span class="op">(</span><span class="va">sp</span>, reporters <span class="op">=</span> <span class="va">iTRAQ4</span>, full <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/spectrumPlot-1.png" alt="Raw MS2 spectrum with details about reporter ions." width="700"><p class="caption">
Raw MS2 spectrum with details about reporter ions.
</p>
</div>
<p>It is also possible to plot all spectra of an experiment (figure
@ref(fig:msnexpPlot)). Lets start by subsetting the
<code>itraqdata</code> experiment using the protein accession numbers
included in the feature metadata, and keep the 6 from the <em>BSA</em>
protein.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span><span class="op">$</span><span class="va">ProteinAccession</span> <span class="op">==</span> <span class="st">"BSA"</span></span>
<span><span class="va">bsa</span> <span class="op">&lt;-</span> <span class="va">itraqdata</span><span class="op">[</span><span class="va">sel</span><span class="op">]</span></span>
<span><span class="va">bsa</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn experiment data ("MSnExp")</span></span>
<span><span class="co">## Object size in memory: 0.11 Mb</span></span>
<span><span class="co">## - - - Spectra data - - -</span></span>
<span><span class="co">##  MS level(s): 2 </span></span>
<span><span class="co">##  Number of spectra: 3 </span></span>
<span><span class="co">##  MSn retention times: 19:09 - 36:17 minutes</span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data loaded: Wed May 11 18:54:39 2011 </span></span>
<span><span class="co">## Updated from version 0.3.0 to 0.3.1 [Fri Jul  8 20:23:25 2016] </span></span>
<span><span class="co">## Data [logically] subsetted 3 spectra: Tue Apr 30 12:23:45 2024 </span></span>
<span><span class="co">##  MSnbase version: 1.1.22 </span></span>
<span><span class="co">## - - - Meta data  - - -</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   rowNames: 1</span></span>
<span><span class="co">##   varLabels: sampleNames sampleNumbers</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## Loaded from:</span></span>
<span><span class="co">##   dummyiTRAQ.mzXML </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: X1 X52 X53</span></span>
<span><span class="co">##   fvarLabels: spectrum ProteinAccession ProteinDescription</span></span>
<span><span class="co">##     PeptideSequence</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">bsa</span><span class="op">)</span><span class="op">$</span><span class="va">ProteinAccession</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BSA" "BSA" "BSA"</span></span></code></pre>
<p>These can then be visualised together by plotting the <em>MSnExp</em>
object, as illustrated on figure @ref(fig:msnexpPlot).</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot-methods.html">plot</a></span><span class="op">(</span><span class="va">bsa</span>, reporters <span class="op">=</span> <span class="va">iTRAQ4</span>, full <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_gray</a></span><span class="op">(</span><span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/msnexpPlot-1.png" alt="Experiment-wide raw MS2 spectra. The y-axes of the individual spectra are automatically rescaled to the same range. See section \@ref(sec:norm) to rescale peaks identically." width="700"><p class="caption">
Experiment-wide raw MS2 spectra. The y-axes of the individual spectra
are automatically rescaled to the same range. See section @ref(sec:norm)
to rescale peaks identically.
</p>
</div>
<p><strong>Customising your plots</strong> The <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
<code>plot</code> methods have a logical <code>plot</code> parameter
(default is <code>TRUE</code>), that specifies if the plot should be
printed to the current device. A plot object is also (invisibly)
returned, so that it can be saved as a variable for later use or for
customisation.</p>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
uses the package to generate plots, which can subsequently easily be
customised. More details about can be found in <span class="citation">(Wickham 2009)</span> (especially chapter 8) and on <a href="http://had.co.nz/ggplot2/" class="external-link uri">http://had.co.nz/ggplot2/</a>. Finally, if a plot object has
been saved in a variable <code>p</code>, it is possible to obtain a
summary of the object with <code>summary(p)</code>. To view the data
frame used to generate the plot, use <code>p$data</code>.</p>
</div>
<div class="section level3">
<h3 id="sec:chromplots">MS Chromatogram<a class="anchor" aria-label="anchor" href="#sec:chromplots"></a>
</h3>
<p>Chromatographic data can be plotted using the <code>plot</code>
method which, in contrast to the <code>plot</code> method for
<code>Spectrum</code> classes, uses R base graphics. The
<code>plot</code> method is implemented for <code>Chromatogram</code>
and <code>MChromatograms</code> classes. The latter plots all
chromatograms for the same <span class="math inline">\(m/z\)</span>-rt
range of all files in an experiment (i.e. for one row in the
<code>MChromatograms</code> object) into one plot.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot-methods.html">plot</a></span><span class="op">(</span><span class="va">mtof_bpc</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/chromPlot-1.png" alt="Base peak chromatogram." width="700"><p class="caption">
Base peak chromatogram.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="sec:id">Tandem MS identification data<a class="anchor" aria-label="anchor" href="#sec:id"></a>
</h2>
<p>Typically, identification data is produced by a search engine and
serialised to disk in the <code>mzIdentML</code> (or <code>mzid</code>)
file format. This format can be parsed by <code>openIDfile</code> from
the <em><a href="https://bioconductor.org/packages/3.18/mzR" class="external-link">mzR</a></em> package
or <code>mzID</code> from the <em><a href="https://bioconductor.org/packages/3.18/mzID" class="external-link">mzID</a></em>
package. The <code>MSnbase</code> package relies on the former (which is
faster) and offers a simplified interface by converting the dedicated
identification data objects into <code>data.frames</code>.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"msdata"</span><span class="op">)</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="st">"TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid"</span></span>
<span><span class="va">idf</span> <span class="op">&lt;-</span> <span class="fu">msdata</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html" class="external-link">ident</a></span><span class="op">(</span>full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="va">f</span><span class="op">)</span></span>
<span><span class="va">iddf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMzIdData.html">readMzIdData</a></span><span class="op">(</span><span class="va">idf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">iddf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    5802 obs. of  35 variables:</span></span>
<span><span class="co">##  $ sequence                : chr  "RQCRTDFLNYLR" "ESVALADQVTCVDWRNRKATKK" "KELLCLAMQIIR" "QRMARTSDKQQSIRFLERLCGR" ...</span></span>
<span><span class="co">##  $ spectrumID              : chr  "controllerType=0 controllerNumber=1 scan=2949" "controllerType=0 controllerNumber=1 scan=6534" "controllerType=0 controllerNumber=1 scan=5674" "controllerType=0 controllerNumber=1 scan=4782" ...</span></span>
<span><span class="co">##  $ chargeState             : int  3 2 2 3 3 3 2 3 3 2 ...</span></span>
<span><span class="co">##  $ rank                    : int  1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">##  $ passThreshold           : logi  TRUE TRUE TRUE TRUE TRUE TRUE ...</span></span>
<span><span class="co">##  $ experimentalMassToCharge: num  548 1288 744 913 927 ...</span></span>
<span><span class="co">##  $ calculatedMassToCharge  : num  548 1288 744 913 926 ...</span></span>
<span><span class="co">##  $ peptideRef              : chr  "Pep77" "Pep108" "Pep136" "Pep163" ...</span></span>
<span><span class="co">##  $ modNum                  : int  1 1 1 1 1 1 1 2 2 1 ...</span></span>
<span><span class="co">##  $ isDecoy                 : logi  FALSE FALSE TRUE FALSE TRUE FALSE ...</span></span>
<span><span class="co">##  $ post                    : chr  "V" "G" "Q" "D" ...</span></span>
<span><span class="co">##  $ pre                     : chr  "R" "R" "R" "R" ...</span></span>
<span><span class="co">##  $ start                   : int  574 69 131 182 135 310 182 201 201 121 ...</span></span>
<span><span class="co">##  $ end                     : int  585 90 142 203 158 334 203 233 233 140 ...</span></span>
<span><span class="co">##  $ DatabaseAccess          : chr  "ECA2006" "ECA1676" "XXX_ECA2855" "ECA3009" ...</span></span>
<span><span class="co">##  $ DBseqLength             : int  1295 110 157 437 501 477 437 1204 1204 210 ...</span></span>
<span><span class="co">##  $ DatabaseSeq             : chr  "" "" "" "" ...</span></span>
<span><span class="co">##  $ DatabaseDescription     : chr  "ECA2006 ATP-dependent helicase" "ECA1676 putative growth inhibitory protein" "" "ECA3009 putative coproporphyrinogen oxidase" ...</span></span>
<span><span class="co">##  $ scan.number.s.          : num  2949 6534 5674 4782 5839 ...</span></span>
<span><span class="co">##  $ acquisitionNum          : num  2949 6534 5674 4782 5839 ...</span></span>
<span><span class="co">##  $ spectrumFile            : chr  "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML" "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML" "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML" "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML" ...</span></span>
<span><span class="co">##  $ idFile                  : chr  "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid" "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid" "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid" "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid" ...</span></span>
<span><span class="co">##  $ MS.GF.RawScore          : num  10 12 8 -5 8 7 21 -31 -31 -3 ...</span></span>
<span><span class="co">##  $ MS.GF.DeNovoScore       : num  101 121 74 160 241 214 196 165 165 59 ...</span></span>
<span><span class="co">##  $ MS.GF.SpecEValue        : num  4.62e-08 7.26e-08 9.34e-08 1.27e-07 1.32e-07 ...</span></span>
<span><span class="co">##  $ MS.GF.EValue            : num  0.132 0.209 0.267 0.366 0.379 ...</span></span>
<span><span class="co">##  $ MS.GF.QValue            : num  0.525 0.61 0.625 0.717 0.736 ...</span></span>
<span><span class="co">##  $ MS.GF.PepQValue         : num  0.549 0.623 0.636 0.724 0.745 ...</span></span>
<span><span class="co">##  $ modPeptideRef           : chr  "Pep77" "Pep108" "Pep136" "Pep163" ...</span></span>
<span><span class="co">##  $ modName                 : chr  "Carbamidomethyl" "Carbamidomethyl" "Carbamidomethyl" "Carbamidomethyl" ...</span></span>
<span><span class="co">##  $ modMass                 : num  57 57 57 57 57 ...</span></span>
<span><span class="co">##  $ modLocation             : int  3 11 5 20 20 21 20 1 28 4 ...</span></span>
<span><span class="co">##  $ subOriginalResidue      : chr  NA NA NA NA ...</span></span>
<span><span class="co">##  $ subReplacementResidue   : chr  NA NA NA NA ...</span></span>
<span><span class="co">##  $ subLocation             : int  NA NA NA NA NA NA NA NA NA NA ...</span></span></code></pre>
<p>The spectra along the rows are duplicated when the PSM can be
assigned to multiple proteins, such as</p>
<pre><code><span><span class="co">##                                         spectrumID    sequence DatabaseAccess</span></span>
<span><span class="co">## 3794 controllerType=0 controllerNumber=1 scan=5291  RKAYLLRMRR    XXX_ECA2052</span></span>
<span><span class="co">## 4886 controllerType=0 controllerNumber=1 scan=5291 ILLHPLRTLMR        ECA1281</span></span></code></pre>
<p>of when there are multiple modifications in a PSM, such as</p>
<pre><code><span><span class="co">##                                        spectrumID</span></span>
<span><span class="co">## 411 controllerType=0 controllerNumber=1 scan=4936</span></span>
<span><span class="co">## 412 controllerType=0 controllerNumber=1 scan=4936</span></span>
<span><span class="co">##                                    sequence         modName modLocation</span></span>
<span><span class="co">## 411 ICSAILRIISPEWWGRKLWRLRCEWRENQFRAIGVIHKK Carbamidomethyl           2</span></span>
<span><span class="co">## 412 ICSAILRIISPEWWGRKLWRLRCEWRENQFRAIGVIHKK Carbamidomethyl          23</span></span></code></pre>
<p>At this stage, it is useful to perform some exploratory data analysis
and visualisation on the identification data. For example</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">iddf</span><span class="op">$</span><span class="va">isDecoy</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##  2906  2896</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">iddf</span><span class="op">$</span><span class="va">chargeState</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    2    3    4    5    6 </span></span>
<span><span class="co">## 3312 2064  400   23    3</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">iddf</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">MS.GF.RawScore</span>, colour <span class="op">=</span> <span class="va">isDecoy</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">chargeState</span><span class="op">)</span></span></code></pre></div>
<p><img src="v01-MSnbase-demo_files/figure-html/idvis-1.png" width="700"></p>
<p>The <code>filterIdentificationDataFrame</code> function can be used
to remove - PSMs that match decoy entries - PSMs of rank &gt; 1 - PSMs
that match non-proteotypic proteins</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">iddf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterIdentificationDataFrame.html">filterIdentificationDataFrame</a></span><span class="op">(</span><span class="va">iddf</span><span class="op">)</span></span></code></pre></div>
<p>This <code>data.frame</code> can be now be further reduced so that
individual rows represent unique spectra, which can be done with the
<code>reduce</code> method.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">iddf2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">iddf</span>, key <span class="op">=</span> <span class="st">"spectrumID"</span><span class="op">)</span></span></code></pre></div>
<p>This reduces the number of rows from 2710 to 2646.</p>
<p>The first duplicated spectrum mentioned above is now unique as is
matched a decoy protein that was filtered out with
<code>filterIdentificationDataFrame</code>.</p>
<pre><code><span><span class="co">##                                         spectrumID    sequence DatabaseAccess</span></span>
<span><span class="co">## 1808 controllerType=0 controllerNumber=1 scan=5291 ILLHPLRTLMR        ECA1281</span></span></code></pre>
<p>The matches to multiple modification in the same peptide are now
combined into a single row and documented as semicolon-separated
values.</p>
<pre><code><span><span class="co">##                                         spectrumID</span></span>
<span><span class="co">## 1659 controllerType=0 controllerNumber=1 scan=4936</span></span>
<span><span class="co">##                                                                             sequence</span></span>
<span><span class="co">## 1659 ICSAILRIISPEWWGRKLWRLRCEWRENQFRAIGVIHKK;ICSAILRIISPEWWGRKLWRLRCEWRENQFRAIGVIHKK</span></span>
<span><span class="co">##                              modName modLocation</span></span>
<span><span class="co">## 1659 Carbamidomethyl;Carbamidomethyl        2;23</span></span></code></pre>
<p>This is the form that is used when combined to raw data, as described
in the next section.</p>
<div class="section level3">
<h3 id="adding-identification-data">Adding identification data<a class="anchor" aria-label="anchor" href="#adding-identification-data"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
is able to integrate identification data from <code>mzIdentML</code>
<span class="citation">(Jones et al. 2012)</span> files.</p>
<p>We first load two example files shipped with the <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
containing raw data (as above) and the corresponding identification
results respectively. The raw data is read with the
<code>readMSData</code>, as demonstrated above. As can be seen, the
default feature data only contain spectra numbers. More data about the
spectra is of course available in an <em>MSnExp</em> object, as
illustrated in the previous sections. See also <code><a href="../reference/pSet-class.html">?pSet</a></code> and
<code><a href="../reference/MSnExp-class.html">?MSnExp</a></code> for more details.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## find path to a mzXML file</span></span>
<span><span class="va">quantFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"MSnbase"</span>, dir <span class="op">=</span> <span class="st">"extdata"</span><span class="op">)</span>,</span>
<span>                 full.name <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"mzXML$"</span><span class="op">)</span></span>
<span><span class="co">## find path to a mzIdentML file</span></span>
<span><span class="va">identFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"MSnbase"</span>, dir <span class="op">=</span> <span class="st">"extdata"</span><span class="op">)</span>,</span>
<span>                 full.name <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"dummyiTRAQ.mzid"</span><span class="op">)</span></span>
<span><span class="co">## create basic MSnExp</span></span>
<span><span class="va">msexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMSData.html">readMSData</a></span><span class="op">(</span><span class="va">quantFile</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       spectrum</span></span>
<span><span class="co">## F1.S1        1</span></span>
<span><span class="co">## F1.S2        2</span></span></code></pre>
<p>The <code>addIdentificationData</code> method takes an
<em>MSnExp</em> instance (or an <em>MSnSet</em> instance storing
quantitation data, see section @ref(sec:quant)) as first argument and
one or multiple <code>mzIdentML</code> file names (as a character
vector) as second one<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a> and updates the <em>MSnExp</em> feature
data using the identification data read from the <code>mzIdentML</code>
file(s).</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addIdentificationData-methods.html">addIdentificationData</a></span><span class="op">(</span><span class="va">msexp</span>, id <span class="op">=</span> <span class="va">identFile</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       spectrum acquisition.number          sequence chargeState rank</span></span>
<span><span class="co">## F1.S1        1                  1 VESITARHGEVLQLRPK           3    1</span></span>
<span><span class="co">## F1.S2        2                  2     IDGQWVTHQWLKK           3    1</span></span>
<span><span class="co">##       passThreshold experimentalMassToCharge calculatedMassToCharge peptideRef</span></span>
<span><span class="co">## F1.S1          TRUE                 645.3741               645.0375       Pep2</span></span>
<span><span class="co">## F1.S2          TRUE                 546.9586               546.9633       Pep1</span></span>
<span><span class="co">##       modNum isDecoy post pre start end DatabaseAccess DBseqLength DatabaseSeq</span></span>
<span><span class="co">## F1.S1      0   FALSE    A   R   170 186        ECA0984         231            </span></span>
<span><span class="co">## F1.S2      0   FALSE    A   K    50  62        ECA1028         275            </span></span>
<span><span class="co">##                                                              DatabaseDescription</span></span>
<span><span class="co">## F1.S1                                        ECA0984 DNA mismatch repair protein</span></span>
<span><span class="co">## F1.S2 ECA1028 2,3,4,5-tetrahydropyridine-2,6-dicarboxylate N-succinyltransferase</span></span>
<span><span class="co">##       scan.number.s.          idFile MS.GF.RawScore MS.GF.DeNovoScore</span></span>
<span><span class="co">## F1.S1              1 dummyiTRAQ.mzid            -39                77</span></span>
<span><span class="co">## F1.S2              2 dummyiTRAQ.mzid            -30                39</span></span>
<span><span class="co">##       MS.GF.SpecEValue MS.GF.EValue modPeptideRef modName modMass modLocation</span></span>
<span><span class="co">## F1.S1     5.527468e-05     79.36958          &lt;NA&gt;    &lt;NA&gt;      NA          NA</span></span>
<span><span class="co">## F1.S2     9.399048e-06     13.46615          &lt;NA&gt;    &lt;NA&gt;      NA          NA</span></span>
<span><span class="co">##       subOriginalResidue subReplacementResidue subLocation nprot npep.prot</span></span>
<span><span class="co">## F1.S1               &lt;NA&gt;                  &lt;NA&gt;          NA     1         1</span></span>
<span><span class="co">## F1.S2               &lt;NA&gt;                  &lt;NA&gt;          NA     1         1</span></span>
<span><span class="co">##       npsm.prot npsm.pep</span></span>
<span><span class="co">## F1.S1         1        1</span></span>
<span><span class="co">## F1.S2         1        1</span></span></code></pre>
<p>Finally we can use <code>idSummary</code> to summarise the percentage
of identified features per quantitation/identification pairs.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MSnSet-class.html">idSummary</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       spectrumFile          idFile coverage</span></span>
<span><span class="co">## 1 dummyiTRAQ.mzXML dummyiTRAQ.mzid      0.6</span></span></code></pre>
<p>When identification data is present, and hence peptide sequences, one
can annotation fragment peaks on the MS2 figure by passing the peptide
sequence to the <code>plot</code> method.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">itraqdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pickPeaks-method.html">pickPeaks</a></span><span class="op">(</span><span class="va">itraqdata</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fl">14</span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">itraqdata2</span><span class="op">)</span><span class="op">[</span><span class="va">i</span>, <span class="st">"PeptideSequence"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot-methods.html">plot</a></span><span class="op">(</span><span class="va">itraqdata2</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">s</span>, main <span class="op">=</span> <span class="va">s</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/fragplot-1.png" alt="Annotated MS2 spectrum." width="700"><p class="caption">
Annotated MS2 spectrum.
</p>
</div>
<p>The fragment ions are calculated with the
<code>calculateFragments</code>, described in section
@ref(sec:calcfrag).</p>
</div>
<div class="section level3">
<h3 id="filtering-identification-data">Filtering identification data<a class="anchor" aria-label="anchor" href="#filtering-identification-data"></a>
</h3>
<p>One can remove the features that have not been identified using
<code>removeNoId</code>. This function uses by default the
<code>pepseq</code> feature variable to search the presence of missing
data (<code>NA</code> values) and then filter these non-identified
spectra.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">$</span><span class="va">sequence</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "VESITARHGEVLQLRPK" "IDGQWVTHQWLKK"     NA                 </span></span>
<span><span class="co">## [4] NA                  "LVILLFR"</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removeNoId-methods.html">removeNoId</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">$</span><span class="va">sequence</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "VESITARHGEVLQLRPK" "IDGQWVTHQWLKK"     "LVILLFR"</span></span></code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/MSnSet-class.html">idSummary</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       spectrumFile          idFile coverage</span></span>
<span><span class="co">## 1 dummyiTRAQ.mzXML dummyiTRAQ.mzid        1</span></span></code></pre>
<p>Similarly, the <code>removeMultipleAssignment</code> method can be
used to filter out non-unique features, i.e. that have been assigned to
protein groups with more than one member. This function uses by default
the <code>nprot</code> feature variable.</p>
<p>Note that <code>removeNoId</code> and
<code>removeMultipleAssignment</code> methods can also be called on
<em>MSnExp</em> instances.</p>
</div>
<div class="section level3">
<h3 id="sec:calcfrag">Calculate Fragments<a class="anchor" aria-label="anchor" href="#sec:calcfrag"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
is able to calculate theoretical peptide fragments via
<code>calculateFragments</code>.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateFragments-methods.html">calculateFragments</a></span><span class="op">(</span><span class="st">"ACEK"</span>,</span>
<span>                   type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>, <span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           mz ion type pos z seq</span></span>
<span><span class="co">## 1   44.04947  a1    a   1 1   A</span></span>
<span><span class="co">## 2  204.08012  a2    a   2 1  AC</span></span>
<span><span class="co">## 3  333.12271  a3    a   3 1 ACE</span></span>
<span><span class="co">## 4   72.04439  b1    b   1 1   A</span></span>
<span><span class="co">## 5  232.07504  b2    b   2 1  AC</span></span>
<span><span class="co">## 6  361.11763  b3    b   3 1 ACE</span></span>
<span><span class="co">## 7   89.07094  c1    c   1 1   A</span></span>
<span><span class="co">## 8  249.10159  c2    c   2 1  AC</span></span>
<span><span class="co">## 9  378.14417  c3    c   3 1 ACE</span></span>
<span><span class="co">## 10 173.09207  x1    x   1 1   K</span></span>
<span><span class="co">## 11 302.13466  x2    x   2 1  EK</span></span>
<span><span class="co">## 12 462.16531  x3    x   3 1 CEK</span></span>
<span><span class="co">## 13 147.11280  y1    y   1 1   K</span></span>
<span><span class="co">## 14 276.15539  y2    y   2 1  EK</span></span>
<span><span class="co">## 15 436.18604  y3    y   3 1 CEK</span></span>
<span><span class="co">## 16 130.08625  z1    z   1 1   K</span></span>
<span><span class="co">## 17 259.12884  z2    z   2 1  EK</span></span>
<span><span class="co">## 18 419.15949  z3    z   3 1 CEK</span></span>
<span><span class="co">## 19 284.12409 x2_   x_   2 1  EK</span></span>
<span><span class="co">## 20 258.14483 y2_   y_   2 1  EK</span></span>
<span><span class="co">## 21 241.11828 z2_   z_   2 1  EK</span></span>
<span><span class="co">## 22 155.08150 x1_   x_   1 1   K</span></span>
<span><span class="co">## 23 444.15474 x3_   x_   3 1 CEK</span></span>
<span><span class="co">## 24 129.10224 y1_   y_   1 1   K</span></span>
<span><span class="co">## 25 418.17548 y3_   y_   3 1 CEK</span></span>
<span><span class="co">## 26 112.07569 z1_   z_   1 1   K</span></span>
<span><span class="co">## 27 401.14893 z3_   z_   3 1 CEK</span></span></code></pre>
<p>It is also possible to match these fragments against an
<em>Spectrum2</em> object.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pepseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">$</span><span class="va">sequence</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/calculateFragments-methods.html">calculateFragments</a></span><span class="op">(</span><span class="va">pepseq</span>, <span class="va">msexp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"b"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           mz  intensity  ion type pos z             seq       error</span></span>
<span><span class="co">## 1   100.0005       0.00   b1    b   1 1               V  0.07522824</span></span>
<span><span class="co">## 2   429.2563 1972344.00   b4    b   4 1            VESI -0.02189010</span></span>
<span><span class="co">## 3   512.3044  684918.00  b5_   b_   5 1           VESIT -0.03290132</span></span>
<span><span class="co">## 4   513.3047 2574137.00   y4    y   4 1            LRPK  0.04598246</span></span>
<span><span class="co">## 5   583.3300 1440833.75  b6_   b_   6 1          VESITA -0.02142609</span></span>
<span><span class="co">## 6   754.4504  537234.81   y6    y   6 1          LQLRPK  0.04293155</span></span>
<span><span class="co">## 7   836.6139   82364.42  y7*   y*   7 1         VLQLRPK -0.07865960</span></span>
<span><span class="co">## 8   982.5354  500159.06   y8    y   8 1        EVLQLRPK  0.06897061</span></span>
<span><span class="co">## 9  1080.5867  209363.69  b10    b  10 1      VESITARHGE -0.04344392</span></span>
<span><span class="co">## 10 1672.8380   76075.02 b15*   b*  15 1 VESITARHGEVLQLR  0.07488430</span></span>
<span><span class="co">## 11 1688.0375  136748.83 y15*   y*  15 1 SITARHGEVLQLRPK -0.07729359</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="sec:qc">Quality control<a class="anchor" aria-label="anchor" href="#sec:qc"></a>
</h2>
<p>The current section is not executed dynamically for package size and
processing time constrains. The figures and tables have been generated
with the respective methods and included statically in the vignette for
illustration purposes.</p>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
allows easy and flexible access to the data, which allows to visualise
data features to assess it’s quality. Some methods are readily
available, although many QC approaches will be experiment specific and
users are encourage to explore their data.</p>
<p>The <code>plot2d</code> method takes one <em>MSnExp</em> instance as
first argument to produce retention time <em>vs.</em> precursor MZ
scatter plots. Points represent individual MS2 spectra and can be
coloured based on precursor charge (with second argument
<code>z="charge"</code>), total ion count (<code>z="ionCount"</code>),
number of peaks in the MS2 spectra <code>z="peaks.count"</code>) or,
when multiple data files were loaded, file <code>z="file"</code>), as
illustrated on the <a href="#fig:plot2d">next figure</a>. The lower
right panel is produced for only a subset of proteins. See the method
documentation for more details.</p>
<div class="float" id="fig:plot2d">
<img src="Figures/plot2d-figure.png" alt="Illustration of the plot2d output."><div class="figcaption">Illustration of the <code>plot2d</code>
output.</div>
</div>
<p>The <code>plotDensity</code> method illustrates the distribution of
several parameters of interest (see <a href="#fig:plotDensity">figure
below</a>). Similarly to <code>plot2d</code>, the first argument is an
<em>MSnExp</em> instance. The second is one of
<code>precursor.mz</code>, <code>peaks.count</code> or
<code>ionCount</code>, whose density will be plotted. An optional third
argument specifies whether the x axes should be logged.</p>
<div class="float" id="fig:plotDensity">
<img src="Figures/plotDensity-figure.png" alt="Illustration of the plotDensity output."><div class="figcaption">Illustration of the <code>plotDensity</code>
output.</div>
</div>
<p>The <code>plotMzDelta</code> method<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a> implements the <span class="math inline">\(m/z\)</span> delta plot from <span class="citation">(Foster et al. 2011)</span> The <span class="math inline">\(m/z\)</span> delta plot illustrates the
suitability of MS2 spectra for identification by plotting the <span class="math inline">\(m/z\)</span> differences of the most intense
peaks. The resulting histogram should optimally shown outstanding bars
at amino acid residu masses. More details and parameters are described
in the method documentation (<code><a href="../reference/plotMzDelta-methods.html">?plotMzDelta</a></code>). The <a href="#fig:plotMzDelta">next figure</a> has been generated using the
PRIDE experiment 12011, as in <span class="citation">(Foster et al.
2011)</span>.</p>
<div class="float" id="fig:plotMzDelta">
<img src="Figures/plotMzDelta-pride12011.png" alt="Illustration of the plotMzDelta output for the PRIDE experiment 12011, as in figure 4A from (Foster et al. 2011)."><div class="figcaption">Illustration of the <code>plotMzDelta</code>
output for the PRIDE experiment 12011, as in figure 4A from <span class="citation">(Foster et al. 2011)</span>.</div>
</div>
<p>In section @ref(sec:incompdissoc), we illustrate how to assess
incomplete reporter ion dissociation.</p>
</div>
<div class="section level2">
<h2 id="sec:rawprocessing">Raw data processing<a class="anchor" aria-label="anchor" href="#sec:rawprocessing"></a>
</h2>
<div class="section level3">
<h3 id="sec:clean">Cleaning spectra<a class="anchor" aria-label="anchor" href="#sec:clean"></a>
</h3>
<p>There are several methods implemented to perform basic raw data
processing and manipulation. Low intensity peaks can be set to 0 with
the <code>removePeaks</code> method from spectra or whole experiments.
The intensity threshold below which peaks are removed is defined by the
<code>t</code> parameter. <code>t</code> can be specified directly as a
numeric. The default value is the character <code>"min"</code>, that
will remove all peaks equal to the lowest non null intensity in any
spectrum. We observe the effect of the <code>removePeaks</code> method
by comparing total ion count (i.e. the total intensity in a spectrum)
with the <code>ionCount</code> method before (object
<code>itraqdata</code>) and after (object <code>experiment</code>) for
spectrum <code>X55</code>. The respective spectra are shown on figure
@ref(fig:spectrum-clean-plot).</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/removePeaks-methods.html">removePeaks</a></span><span class="op">(</span><span class="va">itraqdata</span>, t <span class="op">=</span> <span class="fl">400</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Spectrum-class.html">ionCount</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">[[</span><span class="st">"X55"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 555408.8</span></span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">ionCount</a></span><span class="op">(</span><span class="va">experiment</span><span class="op">[[</span><span class="st">"X55"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 499769.6</span></span></code></pre>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/spectrum-clean-plot-1.png" alt="Same spectrum before (left) and after setting peaks &lt;= 400 to 0." width="700"><p class="caption">
Same spectrum before (left) and after setting peaks &lt;= 400 to 0.
</p>
</div>
<p>Unlike the name might suggest, the <code>removePeaks</code> method
does not actually remove peaks from the spectrum; they are set to 0.
This can be checked using the <code>peaksCount</code> method, that
returns the number of peaks (including 0 intensity peaks) in a spectrum.
To effectively remove 0 intensity peaks from spectra, and reduce the
size of the data set, one can use the <code>clean</code> method. The
effect of the <code>removePeaks</code> and <code>clean</code> methods
are illustrated on figure @ref(fig:preprocPlot).</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">peaksCount</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">[[</span><span class="st">"X55"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1726</span></span></code></pre>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">peaksCount</a></span><span class="op">(</span><span class="va">experiment</span><span class="op">[[</span><span class="st">"X55"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1726</span></span></code></pre>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clean-methods.html">clean</a></span><span class="op">(</span><span class="va">experiment</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Spectrum-class.html">peaksCount</a></span><span class="op">(</span><span class="va">experiment</span><span class="op">[[</span><span class="st">"X55"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 440</span></span></code></pre>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/preprocPlot-1.png" alt="This figure illustrated the effect or the `removePeaks` and `clean` methods. The left-most spectrum displays two peaks, of max height 3 and 7 respectively. The middle spectrum shows the result of calling `removePeaks` with argument `t=3`, which sets all data points of the first peak, whose maximum height is smaller or equal to `t` to 0. The second peak is unaffected. Calling `clean` after `removePeaks` effectively deletes successive 0 intensities from the spectrum, as shown on the right plot." width="700"><p class="caption">
This figure illustrated the effect or the <code>removePeaks</code> and
<code>clean</code> methods. The left-most spectrum displays two peaks,
of max height 3 and 7 respectively. The middle spectrum shows the result
of calling <code>removePeaks</code> with argument <code>t=3</code>,
which sets all data points of the first peak, whose maximum height is
smaller or equal to <code>t</code> to 0. The second peak is unaffected.
Calling <code>clean</code> after <code>removePeaks</code> effectively
deletes successive 0 intensities from the spectrum, as shown on the
right plot.
</p>
</div>
</div>
<div class="section level3">
<h3 id="sec:trim">Focusing on specific MZ values<a class="anchor" aria-label="anchor" href="#sec:trim"></a>
</h3>
<p>Another useful manipulation method is <code>trimMz</code>, that takes
as parameters and <em>MSnExp</em> (or a <em>Spectrum</em>) and a numeric
<code>mzlim</code>. MZ values smaller then <code>min(mzlim)</code> or
greater then <code>max(mzmax)</code> are discarded. This method is
particularly useful when one wants to concentrate on a specific MZ
range, as for reporter ions quantification, and generally results in
substantial reduction of data size. Compare the size of the full trimmed
experiment to the original 1.9 Mb.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="../reference/Spectrum-class.html">mz</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">[[</span><span class="st">"X55"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 100.0002 977.6636</span></span></code></pre>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trimMz-methods.html">filterMz</a></span><span class="op">(</span><span class="va">experiment</span>, mzlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">112</span>,<span class="fl">120</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="../reference/Spectrum-class.html">mz</a></span><span class="op">(</span><span class="va">experiment</span><span class="op">[[</span><span class="st">"X55"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 102.0612 473.3372</span></span></code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">experiment</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn experiment data ("MSnExp")</span></span>
<span><span class="co">## Object size in memory: 1.18 Mb</span></span>
<span><span class="co">## - - - Spectra data - - -</span></span>
<span><span class="co">##  MS level(s): 2 </span></span>
<span><span class="co">##  Number of spectra: 55 </span></span>
<span><span class="co">##  MSn retention times: 19:09 - 50:18 minutes</span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data loaded: Wed May 11 18:54:39 2011 </span></span>
<span><span class="co">## Updated from version 0.3.0 to 0.3.1 [Fri Jul  8 20:23:25 2016] </span></span>
<span><span class="co">## Curves &lt;= 400 set to '0': Tue Apr 30 12:23:51 2024 </span></span>
<span><span class="co">## Spectra cleaned: Tue Apr 30 12:23:51 2024 </span></span>
<span><span class="co">##  MSnbase version: 1.1.22 </span></span>
<span><span class="co">## - - - Meta data  - - -</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   rowNames: 1</span></span>
<span><span class="co">##   varLabels: sampleNames sampleNumbers</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## Loaded from:</span></span>
<span><span class="co">##   dummyiTRAQ.mzXML </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: X1 X10 ... X9 (55 total)</span></span>
<span><span class="co">##   fvarLabels: spectrum ProteinAccession ProteinDescription</span></span>
<span><span class="co">##     PeptideSequence</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span></code></pre>
<p>As can be seen above, all processing performed on the experiment is
recorded and displayed as integral part of the experiment object.</p>
</div>
<div class="section level3">
<h3 id="sec:specproc">Spectrum processing<a class="anchor" aria-label="anchor" href="#sec:specproc"></a>
</h3>
<p><em>MSnExp</em> and <em>Spectrum2</em> instances also support
standard MS data processing such as smoothing and peak picking, as
described in the <code>smooth</code> and <code>pickPeak</code> manual
pages. The methods that either single spectra of experiments, process
the spectrum/spectra, and return a updated, processed, object. The
implementations originate from the package <span class="citation">(Gibb
and Strimmer 2012)</span>.</p>
</div>
</div>
<div class="section level2">
<h2 id="sec:isoquant">MS2 isobaric tagging quantitation<a class="anchor" aria-label="anchor" href="#sec:isoquant"></a>
</h2>
<div class="section level3">
<h3 id="sec:quant">Reporter ions quantitation<a class="anchor" aria-label="anchor" href="#sec:quant"></a>
</h3>
<p>Quantitation is performed on fixed peaks in the spectra, that are
specified with an <em>ReporterIons</em> object. A specific peak is
defined by it’s expected <code>mz</code> value and is searched for
within <code>mz</code> <span class="math inline">\(\pm\)</span>
<code>width</code>. If no data is found, <code>NA</code> is
returned.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">mz</a></span><span class="op">(</span><span class="va">iTRAQ4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 114.1112 115.1083 116.1116 117.1150</span></span></code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ReporterIons-class.html">width</a></span><span class="op">(</span><span class="va">iTRAQ4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.05</span></span></code></pre>
<p>The <code>quantify</code> method takes the following parameters: an
<em>MSnExp</em> experiment, a character describing the quantification
<code>method</code>, the <code>reporters</code> to be quantified and a
<code>strict</code> logical defining whether data points ranging outside
of <code>mz</code> <span class="math inline">\(\pm\)</span>
<code>width</code> should be considered for quantitation. Additionally,
a progress bar can be displaying when setting the <code>verbose</code>
parameter to <code>TRUE</code>. Three quantification methods are
implemented, as illustrated on figure @ref(fig:quantitationPlot).
Quantitation using <code>sum</code> sums all the data points in the
peaks to produce, for this example, 7, whereas method <code>max</code>
only uses the peak’s maximum intensity, 3. <code>Trapezoidation</code>
calculates the area under the peak taking the full with into account
(using <code>strict = FALSE</code> gives 0.375) or only the width as
defined by the reporter (using <code>strict = TRUE</code> gives 0.1).
See <code><a href="../reference/quantify-methods.html">?quantify</a></code> for more details.</p>
<pre><code><span><span class="co">## Warning in geom_point(aes(x = 114.1, y = 3), alpha = I(1/18), colour = "red", : All aesthetics have length 1, but the data has 7 rows.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please consider using `annotate()` or provide this layer with data containing</span></span>
<span><span class="co">##   a single row.</span></span></code></pre>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/quantitationPlot-1.png" alt="The different quantitation methods. See text for details." width="700"><p class="caption">
The different quantitation methods. See text for details.
</p>
</div>
<p>The <code>quantify</code> method returns <em>MSnSet</em> objects,
that extend the well-known <em>eSet</em> class defined in the <em><a href="https://bioconductor.org/packages/3.18/Biobase" class="external-link">Biobase</a></em>
package. <em>MSnSet</em> instances are very similar to
<em>ExpressionSet</em> objects, except for the experiment meta-data that
captures MIAPE specific information. The assay data is a matrix of
dimensions <span class="math inline">\(n \times m\)</span>, where <span class="math inline">\(m\)</span> is the number of features/spectra
originally in the <em>MSnExp</em> used as parameter in
<code>quantify</code> and <span class="math inline">\(m\)</span> is the
number of reporter ions, that can be accessed with the
<code>exprs</code> method. The meta data is directly inherited from the
<em>MSnExp</em> instance.</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">experiment</span>,</span>
<span>                method <span class="op">=</span> <span class="st">"trap"</span>,</span>
<span>                reporters <span class="op">=</span> <span class="va">iTRAQ4</span>,</span>
<span>                strict <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">qnt</span></span></code></pre></div>
<pre><code><span><span class="co">## MSnSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 55 features, 4 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span></span>
<span><span class="co">##   varLabels: mz reporters</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: X1 X10 ... X9 (55 total)</span></span>
<span><span class="co">##   fvarLabels: spectrum ProteinAccession ... collision.energy (15 total)</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation: No annotation </span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data loaded: Wed May 11 18:54:39 2011 </span></span>
<span><span class="co">## Updated from version 0.3.0 to 0.3.1 [Fri Jul  8 20:23:25 2016] </span></span>
<span><span class="co">## Curves &lt;= 400 set to '0': Tue Apr 30 12:23:51 2024 </span></span>
<span><span class="co">## Spectra cleaned: Tue Apr 30 12:23:51 2024 </span></span>
<span><span class="co">## iTRAQ4 quantification by trapezoidation: Tue Apr 30 12:23:53 2024 </span></span>
<span><span class="co">##  MSnbase version: 1.1.22</span></span></code></pre>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">qnt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span></span>
<span><span class="co">## X1   1347.6158  2247.3097  3927.6931  7661.1463</span></span>
<span><span class="co">## X10   739.9861   799.3501   712.5983   940.6793</span></span>
<span><span class="co">## X11 27638.3582 33394.0252 32104.2879 26628.7278</span></span>
<span><span class="co">## X12 31892.8928 33634.6980 37674.7272 37227.7119</span></span>
<span><span class="co">## X13 26143.7542 29677.4781 29089.0593 27902.5608</span></span>
<span><span class="co">## X14  6448.0829  6234.1957  6902.8903  6437.2303</span></span></code></pre>
<p>The <a href="#fig:tmt10">next figure</a> illustrates the quantitation
of the TMT 10-plex isobaric tags using the <code>quantify</code> method
and the <code>TMT10</code> reporter instance. The data on the <span class="math inline">\(x\)</span> axis has been quantified using
<code>method = "max"</code> and centroided data (as generated using
ProteoWizard’s <code>msconvert</code> with vendor libraries’ peak
picking); on the <span class="math inline">\(y\)</span> axis, the
quantitation method was <code>trapezoidation</code> and
<code>strict = TRUE</code> (that’s important for TMT 10-plex) and the
profile data. We observe a very good correlation.</p>
<div class="float" id="fig:tmt10">
<img src="Figures/tmt10comp.png" alt="TMT 10-plex quantitation."><div class="figcaption">TMT 10-plex quantitation.</div>
</div>
<p>If no peak is detected for a reporter ion peak, the respective
quantitation value is set to <code>NA</code>. In our case, there is 1
such case in row 41. We will remove the offending line using the
<code>filterNA</code> method. The <code>pNA</code> argument defines the
percentage of accepted missing values per feature. As we do not expect
any missing peaks, we set it to be 0 (which is also the detault
value).</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">qnt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##   219     1</span></span></code></pre>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qnt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSnSet-class.html">filterNA</a></span><span class="op">(</span><span class="va">qnt</span>, pNA <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">qnt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0</span></span></code></pre>
<p>The filtering criteria for <code>filterNA</code> can also be defined
as a pattern of columns that can have missing values and columns that
must not exhibit any. See <code><a href="../reference/MSnSet-class.html">?filterNA</a></code> for details and
examples.</p>
<p>The infrastructure around the <em>MSnSet</em> class allows flexible
filtering using the <code>[</code> sub-setting operator. Below, we mimic
the behaviour of <code>filterNA(, pNA = 0)</code> by calculating the row
indices that should be removed, i.e. those that have at least one
<code>NA</code> value and explicitly remove these rows. This method
allows one to devise and easily apply any filtering strategy.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">whichRow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="op">(</span><span class="va">qnt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">qnt</span><span class="op">)</span></span>
<span><span class="va">qnt</span> <span class="op">&lt;-</span> <span class="va">qnt</span><span class="op">[</span><span class="op">-</span><span class="va">whichRow</span>, <span class="op">]</span></span></code></pre></div>
<p>See also the <code>plotNA</code> method to obtain a graphical
overview of the completeness of a data set.</p>
</div>
<div class="section level3">
<h3 id="sec:io2">Importing quantitation data<a class="anchor" aria-label="anchor" href="#sec:io2"></a>
</h3>
<p>If quantitation data is already available as a spreadsheet, it can be
imported, along with additional optional feature and sample (pheno) meta
data, with the <code>readMSnSet</code> function. This function takes the
respective text-based spreadsheet (comma- or tab-separated) file names
as argument to create a valid <em>MSnSet</em> instance.</p>
<p>Note that the quantitation data of <em>MSnSet</em> objects can also
be exported to a text-based spreadsheet file using the
<code>write.exps</code> method.</p>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
also supports the <code>mzTab</code> format, a light-weight,
tab-delimited file format for proteomics data. <code>mzTab</code> files
can be read into R with <code>readMzTabData</code> to create and
<em>MSnSet</em> instance.</p>
<p>See the <em>MSnbase-io</em> vignette for a general overview of <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>’s
input/ouput capabilites.</p>
</div>
<div class="section level3">
<h3 id="sec:io3">Importing chromatographic data from SRM/MRM experiments<a class="anchor" aria-label="anchor" href="#sec:io3"></a>
</h3>
<p>Data from SRM/MRM experiments can be imported from <code>mzML</code>
files using the <code>readSRMData</code> function. The <code>mzML</code>
files are expected to contain chromatographic data for the same
precursor and product m/z values. The function returns a
<code>MChromatograms</code> object that arranges the data in a
two-dimensional array, each column representing the data of one file
(sample) and each row the chromatographic data for the same polarity,
precursor and product m/z. In the example code below we load a single
SRM file using <code>readSRMData</code>.</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msdata/man/proteomics.html" class="external-link">proteomics</a></span><span class="op">(</span>full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"MRM"</span><span class="op">)</span></span>
<span><span class="va">srm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSRMData.html">readSRMData</a></span><span class="op">(</span><span class="va">fl</span><span class="op">)</span></span>
<span></span>
<span><span class="va">srm</span></span></code></pre></div>
<pre><code><span><span class="co">## MChromatograms with 137 rows and 1 column</span></span>
<span><span class="co">##                     1</span></span>
<span><span class="co">##        &lt;Chromatogram&gt;</span></span>
<span><span class="co">## [1,]      length: 523</span></span>
<span><span class="co">## [2,]      length: 523</span></span>
<span><span class="co">## ...              ... </span></span>
<span><span class="co">## [136,]    length: 962</span></span>
<span><span class="co">## [137,]    length: 962</span></span>
<span><span class="co">## phenoData with 1 variables</span></span>
<span><span class="co">## featureData with 10 variables</span></span></code></pre>
<p>The precursor and product m/z values can be extracted with the
<code>precursorMz</code> and <code>productMz</code> functions. These
functions always return a matrix, each row providing the lower and upper
m/z value of the isolation window (in most cases minimal and maximal m/z
will be identical).</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/Spectrum2-class.html">precursorMz</a></span><span class="op">(</span><span class="va">srm</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      mzmin mzmax</span></span>
<span><span class="co">## [1,]   115   115</span></span>
<span><span class="co">## [2,]   115   115</span></span>
<span><span class="co">## [3,]   117   117</span></span>
<span><span class="co">## [4,]   117   117</span></span>
<span><span class="co">## [5,]   133   133</span></span>
<span><span class="co">## [6,]   133   133</span></span></code></pre>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">productMz</a></span><span class="op">(</span><span class="va">srm</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        mzmin   mzmax</span></span>
<span><span class="co">## [1,]  26.996  26.996</span></span>
<span><span class="co">## [2,]  70.996  70.996</span></span>
<span><span class="co">## [3,]  72.996  72.996</span></span>
<span><span class="co">## [4,]  98.996  98.996</span></span>
<span><span class="co">## [5,] 114.996 114.996</span></span>
<span><span class="co">## [6,]  70.996  70.996</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="sec:purcor">Peak adjustments<a class="anchor" aria-label="anchor" href="#sec:purcor"></a>
</h3>
<p><strong>Single peak adjustment</strong> In certain cases, peak
intensities need to be adjusted as a result of peak interferance. For
example, the <span class="math inline">\(+1\)</span> peak of the
phenylalanine (F, Phe) immonium ion (with <span class="math inline">\(m/z\)</span> 120.03) inteferes with the 121.1 TMT
reporter ion. Below, we calculate the relative intensity of the +1 peaks
compared to the main peak using the <em><a href="https://bioconductor.org/packages/3.18/Rdisop" class="external-link">Rdisop</a></em>
package.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/Rdisop" class="external-link">Rdisop</a></span><span class="op">)</span></span>
<span><span class="co">## Phenylalanine immonium ion</span></span>
<span><span class="va">Fim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rdisop/man/getMolecule.html" class="external-link">getMolecule</a></span><span class="op">(</span><span class="st">"C8H10N"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rdisop/man/getMolecule.html" class="external-link">getMass</a></span><span class="op">(</span><span class="va">Fim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 120.0813</span></span></code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">isotopes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rdisop/man/getMolecule.html" class="external-link">getIsotope</a></span><span class="op">(</span><span class="va">Fim</span><span class="op">)</span></span>
<span><span class="va">F1</span> <span class="op">&lt;-</span> <span class="va">isotopes</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">F1</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.08573496</span></span></code></pre>
<p>If desired, one can thus specifically quantify the F immonium ion in
the MS2 spectrum, estimate the intensity of the +1 ion (0.0857% of the F
peak) and substract this calculated value from the 121.1 TMT reporter
intensity.</p>
<p>The above principle can also be generalised for a set of overlapping
peaks, as described below.</p>
<p><strong>Reporter ions purity correction</strong> Impurities in the
reporter reagents can also bias the results and can be corrected when
manufacturers provide correction coefficients. These generally come as
percentages of each reporter ion that have masses differing by -2, -1,
+1 and +2 Da from the nominal reporter ion mass due to isotopic
variants. The <code>purityCorrect</code> method applies such correction
to <em>MSnSet</em> instances. It also requires a square matrix as second
argument, <code>impurities</code>, that defines the relative percentage
of reporter in the quantified each peak. See <code><a href="../reference/purityCorrect-methods.html">?purityCorrect</a></code>
for more details.</p>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">impurities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.929</span>, <span class="fl">0.059</span>, <span class="fl">0.002</span>, <span class="fl">0.000</span>,</span>
<span>                       <span class="fl">0.020</span>, <span class="fl">0.923</span>, <span class="fl">0.056</span>, <span class="fl">0.001</span>,</span>
<span>                       <span class="fl">0.000</span>, <span class="fl">0.030</span>, <span class="fl">0.924</span>, <span class="fl">0.045</span>,</span>
<span>                       <span class="fl">0.000</span>, <span class="fl">0.001</span>, <span class="fl">0.040</span>, <span class="fl">0.923</span><span class="op">)</span>,</span>
<span>                     nrow <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">qnt.crct</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/purityCorrect-methods.html">purityCorrect</a></span><span class="op">(</span><span class="va">qnt</span>, <span class="va">impurities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">qnt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span></span>
<span><span class="co">## X1   1347.6158  2247.3097  3927.6931  7661.1463</span></span>
<span><span class="co">## X10   739.9861   799.3501   712.5983   940.6793</span></span>
<span><span class="co">## X11 27638.3582 33394.0252 32104.2879 26628.7278</span></span>
<span><span class="co">## X12 31892.8928 33634.6980 37674.7272 37227.7119</span></span>
<span><span class="co">## X13 26143.7542 29677.4781 29089.0593 27902.5608</span></span>
<span><span class="co">## X14  6448.0829  6234.1957  6902.8903  6437.2303</span></span></code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">qnt.crct</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span></span>
<span><span class="co">## X1   1304.7675  2168.1082  3784.2244  8133.9211</span></span>
<span><span class="co">## X10   743.8159   806.5647   696.9024   988.0787</span></span>
<span><span class="co">## X11 27547.6515 33592.3997 32319.1803 27413.1833</span></span>
<span><span class="co">## X12 32127.1898 33408.8353 37806.0787 38658.7865</span></span>
<span><span class="co">## X13 26187.3141 29788.6254 29105.2485 28936.6871</span></span>
<span><span class="co">## X14  6533.1862  6184.1103  6945.2074  6666.5633</span></span></code></pre>
<p>The <code>makeImpuritiesMatrix</code> can be used to create impurity
matrices. It opens a rudimentary spreadsheet that can be directly
edited.</p>
</div>
</div>
<div class="section level2">
<h2 id="sec:qproc">Processing quantitative data<a class="anchor" aria-label="anchor" href="#sec:qproc"></a>
</h2>
<div class="section level3">
<h3 id="sec:imp">Data imputation<a class="anchor" aria-label="anchor" href="#sec:imp"></a>
</h3>
<p>A set of imputation methods are available in the <code>impute</code>
method: it takes an <em>MSnSet</em> instance as input, the name of the
imputation method to be applied (one of bpca, knn, QRILC, MLE, MLE2,
MinDet, MinProb, min, zero, mixed, nbavg, with, RF, none), possible
additional parameters and returns an updated for <em>MSnSet</em> without
any missing values. Below, we apply a deterministic minimum value
imputation on the <code>naset</code> example data:</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## an example MSnSet containing missing values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">naset</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">naset</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">## 10254   770</span></span></code></pre>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## number of NAs per protein</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">naset</span><span class="op">)</span><span class="op">$</span><span class="va">nNA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##   0   1   2   3   4   8   9  10 </span></span>
<span><span class="co">## 301 247  91  13   2  23  10   2</span></span></code></pre>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">naset</span>, <span class="st">"min"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pSet-class.html">processingData</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data imputation using min Tue Apr 30 12:23:55 2024 </span></span>
<span><span class="co">##  MSnbase version: 1.15.6</span></span></code></pre>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE </span></span>
<span><span class="co">## 11024</span></span></code></pre>
<p>As described in more details in <span class="citation">(Lazar et al.
2016)</span>, there are two types of mechanisms resulting in missing
values in LC/MSMS experiments.</p>
<ul>
<li><p>Missing values resulting from absence of detection of a feature,
despite ions being present at detectable concentrations. For example in
the case of ion suppression or as a result from the stochastic,
data-dependent nature of the MS acquisition method. These missing value
are expected to be randomly distributed in the data and are defined as
<em>missing at random</em> (MAR) or <em>missing completely at
random</em> (MCAR).</p></li>
<li><p>Biologically relevant missing values, resulting from the
<em>absence</em> of the low abundance of ions (below the limit of
detection of the instrument). These missing values are not expected to
be randomly distributed in the data and are defined as <em>missing not
at random</em> (MNAR).</p></li>
</ul>
<p>MAR and MCAR values can be reasonably well tackled by many imputation
methods. MNAR data, however, requires some knowledge about the
underlying mechanism that generates the missing data, to be able to
attempt data imputation. MNAR features should ideally be imputed with a
<em>left-censor</em> (for example using a deterministic or probabilistic
minimum value) method. Conversely, it is recommended to use <em>hot
deck</em> methods (for example nearest neighbour, maximum likelihood,
etc) when data are missing at random.</p>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/miximp-1.png" alt="Mixed imputation method. Black cells represent presence of quantitation values and light grey corresponds to missing data. The two groups of interest are depicted in green and blue along the heatmap columns. Two classes of proteins are annotated on the left: yellow are proteins with randomly occurring missing values (if any) while proteins in brown are candidates for non-random missing value imputation." width="700"><p class="caption">
Mixed imputation method. Black cells represent presence of quantitation
values and light grey corresponds to missing data. The two groups of
interest are depicted in green and blue along the heatmap columns. Two
classes of proteins are annotated on the left: yellow are proteins with
randomly occurring missing values (if any) while proteins in brown are
candidates for non-random missing value imputation.
</p>
</div>
<p>It is anticipated that the identification of both classes of missing
values will depend on various factors, such as feature intensities and
experimental design. Below, we use perform mixed imputation, applying
nearest neighbour imputation on the 654 features that are assumed to
contain randomly distributed missing values (if any) (yellow on figure
@ref(fig:miximp)) and a deterministic minimum value imputation on the 35
proteins that display a non-random pattern of missing values (brown on
figure @ref(fig:miximp)).</p>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">naset</span>, method <span class="op">=</span> <span class="st">"mixed"</span>,</span>
<span>            randna <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">naset</span><span class="op">)</span><span class="op">$</span><span class="va">randna</span>,</span>
<span>            mar <span class="op">=</span> <span class="st">"knn"</span>, mnar <span class="op">=</span> <span class="st">"min"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Imputing along margin 1 (features/rows).</span></span></code></pre>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span></span></code></pre></div>
<pre><code><span><span class="co">## MSnSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 689 features, 16 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: M1F1A M1F4A ... M2F11B (16 total)</span></span>
<span><span class="co">##   varLabels: nNA</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: AT1G09210 AT1G21750 ... AT4G39080 (689 total)</span></span>
<span><span class="co">##   fvarLabels: nNA randna</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:  </span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data imputation using mixed Tue Apr 30 12:23:55 2024 </span></span>
<span><span class="co">##  MSnbase version: 1.15.6</span></span></code></pre>
<p>Please read <code>?MsCoreUtils::impute_matix()</code> for a
description of the different methods.</p>
</div>
<div class="section level3">
<h3 id="sec:norm">Normalisation<a class="anchor" aria-label="anchor" href="#sec:norm"></a>
</h3>
<p>A <em>MSnSet</em> object is meant to be compatible with further
downstream packages for data normalisation and statistical analysis.
There is also a <code>normalise</code> (also available as
<code>normalize</code>) method for expression sets. The method takes and
instance of class <em>MSnSet</em> as first argument, and a character to
describe the <code>method</code> to be used:</p>
<ul>
<li><p><code>quantiles</code>: Applies quantile normalisation <span class="citation">(Bolstad et al. 2003)</span> as implemented in the
<code>normalize.quantiles</code> function of the <em><a href="https://bioconductor.org/packages/3.18/preprocessCore" class="external-link">preprocessCore</a></em>
package.</p></li>
<li><p><code>quantiles.robust</code>: Applies robust quantile
normalisation <span class="citation">(Bolstad et al. 2003)</span> as
implemented in the <code>normalize.quantiles.robust</code> function of
the <em><a href="https://bioconductor.org/packages/3.18/preprocessCore" class="external-link">preprocessCore</a></em>
package.</p></li>
<li><p><code>vsn</code>: Applies variance stabilisation normalization
<span class="citation">(Huber et al. 2002)</span> as implemented in the
<code>vsn2</code> function of the <em><a href="https://bioconductor.org/packages/3.18/vsn" class="external-link">vsn</a></em>
package.</p></li>
<li><p><code>max</code>: Each feature’s reporter intensity is divided by
the maximum of the reporter ions intensities.</p></li>
<li><p><code>sum</code>: Each feature’s reporter intensity is divided by
the sum of the reporter ions intensities.</p></li>
</ul>
<p>See <code><a href="../reference/normalise-methods.html">?normalise</a></code> for more methods. A <code>scale</code>
method for <em>MSnSet</em> instances, that relies on the
<code><a href="https://rdrr.io/r/base/scale.html" class="external-link">base::scale</a></code> function.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qnt.max</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalise-methods.html">normalise</a></span><span class="op">(</span><span class="va">qnt</span>, <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="va">qnt.sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalise-methods.html">normalise</a></span><span class="op">(</span><span class="va">qnt</span>, <span class="st">"sum"</span><span class="op">)</span></span>
<span><span class="va">qnt.quant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalise-methods.html">normalise</a></span><span class="op">(</span><span class="va">qnt</span>, <span class="st">"quantiles"</span><span class="op">)</span></span>
<span><span class="va">qnt.qrob</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalise-methods.html">normalise</a></span><span class="op">(</span><span class="va">qnt</span>, <span class="st">"quantiles.robust"</span><span class="op">)</span></span>
<span><span class="va">qnt.vsn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalise-methods.html">normalise</a></span><span class="op">(</span><span class="va">qnt</span>, <span class="st">"vsn"</span><span class="op">)</span></span></code></pre></div>
<p>The effect of these are illustrated on figure @ref(fig:normPlot) and
figure @ref(fig:cvPlot) reproduces figure 3 of <span class="citation">(Karp et al. 2010)</span> that described the
application of vsn on iTRAQ reporter data.</p>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/normPlot-1.png" alt="Comparison of the normalisation *MSnSet* methods. Note that vsn also glog-transforms the intensities." width="700"><p class="caption">
Comparison of the normalisation <em>MSnSet</em> methods. Note that vsn
also glog-transforms the intensities.
</p>
</div>
<pre><code><span><span class="co">## Warning: `qplot()` was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/cvPlot-1.png" alt="CV versus signal intensity comparison for log2 and vsn transformed data. Lines indicate running CV medians." width="700"><p class="caption">
CV versus signal intensity comparison for log2 and vsn transformed data.
Lines indicate running CV medians.
</p>
</div>
<p>Note that it is also possible to normalise individual spectra or
whole <em>MSnExp</em> experiments with the <code>normalise</code> method
using the <code>max</code> method. This will rescale all peaks between 0
and 1. To visualise the relative reporter peaks, one should this first
trim the spectra using method <code>trimMz</code> as illustrated in
section @ref(sec:rawprocessing), then normalise the <em>MSnExp</em> with
<code>normalise</code> using <code>method="max"</code> as illustrated
above and plot the data using <code>plot</code> (figure
@ref(fig:msnexpNormPlot)).</p>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/msnexpNormPlot-1.png" alt="Experiment-wide normalised MS2 spectra. The y-axes of the individual spectra is now rescaled between 0 and 1 (highest peak), as opposed to figure \@ref(fig:msnexpPlot)." width="700"><p class="caption">
Experiment-wide normalised MS2 spectra. The y-axes of the individual
spectra is now rescaled between 0 and 1 (highest peak), as opposed to
figure @ref(fig:msnexpPlot).
</p>
</div>
<p>Additional dedicated normalisation method are available for MS2
label-free quantitation, as described in section @ref(sec:lf) and in the
<code>quantify</code> documentation.</p>
</div>
</div>
<div class="section level2">
<h2 id="sec:feataggregation">Feature aggregation<a class="anchor" aria-label="anchor" href="#sec:feataggregation"></a>
</h2>
<p>The above quantitation and normalisation has been performed on
quantitative data obtained from individual spectra. However, the
biological unit of interest is not the spectrum but the peptide or the
protein. As such, it is important to be able to summarise features that
belong to a same group, i.e. spectra from one peptide, peptides that
originate from one protein, or directly combine all spectra that have
been uniquely associated to one protein.</p>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
provides one function, <code>combineFeatures</code>, that allows to
aggregate features stored in an <em>MSnSet</em> using build-in or user
defined summary function and return a new <em>MSnSet</em> instance. The
three main arguments are described below. Additional details can be
found in the method documentation.</p>
<p><code>combineFeatures</code>’s first argument, <code>object</code>,
is an instance of class <em>MSnSet</em>, as has been created in the
section @ref(sec:quant) for instance. The second argument,
<code>groupBy</code>, is a <code>factor</code> than has as many elements
as there are features in the <em>MSnSet</em> <code>object</code>
argument. The features corresponding to the <code>groupBy</code> levels
will be aggregated so that the resulting <em>MSnSet</em> output will
have <code>length(levels(groupBy))</code> features. Here, we will
combine individual MS2 spectra based on the protein they originate from.
As shown below, this will result in 40 new aggregated features.</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">qnt</span><span class="op">)</span><span class="op">$</span><span class="va">ProteinAccession</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">gb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## gb</span></span>
<span><span class="co">##     BSA ECA0172 ECA0435 ECA0452 ECA0469 ECA0621 ECA0631 ECA0691 ECA0871 ECA0978 </span></span>
<span><span class="co">##       3       1       2       1       2       1       1       1       1       1 </span></span>
<span><span class="co">## ECA1032 ECA1093 ECA1104 ECA1294 ECA1362 ECA1363 ECA1364 ECA1422 ECA1443 ECA2186 </span></span>
<span><span class="co">##       1       1       1       1       1       1       1       1       1       1 </span></span>
<span><span class="co">## ECA2391 ECA2421 ECA2831 ECA3082 ECA3175 ECA3349 ECA3356 ECA3377 ECA3566 ECA3882 </span></span>
<span><span class="co">##       1       1       1       1       1       2       1       1       2       1 </span></span>
<span><span class="co">## ECA3929 ECA3969 ECA4013 ECA4026 ECA4030 ECA4037 ECA4512 ECA4513 ECA4514     ENO </span></span>
<span><span class="co">##       1       1       1       2       1       1       1       1       6       3</span></span></code></pre>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pSet-class.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 40</span></span></code></pre>
<p>The third argument, <code>method</code>, defined how to combine the
features. Predefined functions are readily available and can be
specified as strings (<code>method="mean"</code>,
<code>method="median"</code>, <code>method="sum"</code>,
<code>method="weighted.mean"</code> or
<code>method="medianpolish"</code> to compute respectively the mean,
media, sum, weighted mean or median polish of the features to be
aggregated). Alternatively, is is possible to supply user defined
functions with <code>method=function(x) { ... }</code>. We will use the
<code>median</code> here.</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qnt2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineFeatures.html">combineFeatures</a></span><span class="op">(</span><span class="va">qnt</span>, groupBy <span class="op">=</span> <span class="va">gb</span>, method <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span>
<span><span class="va">qnt2</span></span></code></pre></div>
<pre><code><span><span class="co">## MSnSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 40 features, 4 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117</span></span>
<span><span class="co">##   varLabels: mz reporters</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: BSA ECA0172 ... ENO (40 total)</span></span>
<span><span class="co">##   fvarLabels: spectrum ProteinAccession ... CV.iTRAQ4.117 (19 total)</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:  </span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data loaded: Wed May 11 18:54:39 2011 </span></span>
<span><span class="co">## Updated from version 0.3.0 to 0.3.1 [Fri Jul  8 20:23:25 2016] </span></span>
<span><span class="co">## Curves &lt;= 400 set to '0': Tue Apr 30 12:23:51 2024 </span></span>
<span><span class="co">## Spectra cleaned: Tue Apr 30 12:23:51 2024 </span></span>
<span><span class="co">## iTRAQ4 quantification by trapezoidation: Tue Apr 30 12:23:53 2024 </span></span>
<span><span class="co">## Subset [55,4][54,4] Tue Apr 30 12:23:54 2024 </span></span>
<span><span class="co">## Removed features with more than 0 NAs: Tue Apr 30 12:23:54 2024 </span></span>
<span><span class="co">## Dropped featureData's levels Tue Apr 30 12:23:54 2024 </span></span>
<span><span class="co">## Combined 54 features into 40 using median: Tue Apr 30 12:23:57 2024 </span></span>
<span><span class="co">##  MSnbase version: 2.29.5</span></span></code></pre>
<p>Of interest is also the <code>iPQF</code> spectra-to-protein
summarisation method, which integrates peptide spectra characteristics
and quantitative values for protein quantitation estimation. See
<code><a href="../reference/iPQF.html">?iPQF</a></code> and references therein for details.</p>
</div>
<div class="section level2">
<h2 id="sec:lf">Label-free MS2 quantitation<a class="anchor" aria-label="anchor" href="#sec:lf"></a>
</h2>
<div class="section level3">
<h3 id="peptide-counting">Peptide counting<a class="anchor" aria-label="anchor" href="#peptide-counting"></a>
</h3>
<p>Note that if samples are not multiplexed, label-free MS2 quantitation
by spectral counting is possible using <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>.
Once individual spectra have been assigned to peptides and proteins (see
section @ref(sec:id)), it becomes straightforward to estimate protein
quantities using the simple peptide counting method, as illustrated in
section @ref(sec:feataggregation).</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">msexp</span>, method <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span>
<span><span class="co">## lets modify out data for demonstration purposes</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span><span class="op">$</span><span class="va">DatabaseAccess</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span><span class="op">$</span><span class="va">DatabaseAccess</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span><span class="op">$</span><span class="va">DatabaseAccess</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ECA1028" "ECA1028" "ECA0510"</span></span></code></pre>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineFeatures.html">combineFeatures</a></span><span class="op">(</span><span class="va">sc</span>, groupBy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span><span class="op">$</span><span class="va">DatabaseAccess</span>,</span>
<span>                      method <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">sc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         dummyiTRAQ.mzXML</span></span>
<span><span class="co">## ECA0510                1</span></span>
<span><span class="co">## ECA1028                2</span></span></code></pre>
<p>Such count data could then be further analyses using dedicated count
methods (originally developed for high-throughput sequencing) and
directly available for <em>MSnSet</em> instances in the <em><a href="https://bioconductor.org/packages/3.18/msmsTests" class="external-link">msmsTests</a></em>
Bioconductor package.</p>
</div>
<div class="section level3">
<h3 id="spectral-counting-and-intensity-methods">Spectral counting and intensity methods<a class="anchor" aria-label="anchor" href="#spectral-counting-and-intensity-methods"></a>
</h3>
<p>The spectral abundance factor (SAF) and the normalised form (NSAF)
<span class="citation">(Paoletti et al. 2006)</span> as well as the
spectral index (SI) and other normalised variations (SI<span class="math inline">\(_{GI}\)</span> and SI<span class="math inline">\(_N\)</span>) <span class="citation">(Griffin et
al. 2010)</span> are also available. Below, we illustrate how to apply
the normalised SI<span class="math inline">\(_N\)</span> to the
experiment containing identification data produced in section
@ref(sec:id).</p>
<p>The spectra that did not match any peptide have already been remove
with the <code>removeNoId</code> method. As can be seen in the following
code chunk, the first spectrum could not be matched to any single
protein. Non-identified spectra and those matching multiple proteins are
removed automatically prior to any label-free quantitation. Once can
also remove peptide that do not match uniquely to proteins (as defined
by the <code>nprot</code> feature variable column) with the
<code>removeMultipleAssignment</code> method.</p>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">msexp</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DatabaseAccess"</span>, <span class="st">"nprot"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       DatabaseAccess nprot</span></span>
<span><span class="co">## F1.S1        ECA0984     1</span></span>
<span><span class="co">## F1.S2        ECA1028     1</span></span>
<span><span class="co">## F1.S5        ECA0510     1</span></span></code></pre>
<p>Note that the label-free methods implicitely apply feature
aggregation (section @ref(sec:feataggregation)) and normalise (section
@ref(sec:norm)) the quantitation values based on the total sample
intensity and or the protein lengths (see <span class="citation">(Paoletti et al. 2006)</span> and <span class="citation">(Griffin et al. 2010)</span> for details).</p>
<p>Let’s now proceed with the quantitation using the
<code>quantify</code>, as in section @ref(sec:quant), this time however
specifying the method of interest, <code>SIn</code> (the
<code>reporters</code> argument can of course be ignored here). The
required peptide-protein mapping and protein lengths are extracted
automatically from the feature meta-data using the default
<code>accession</code> and <code>length</code> feature variables.</p>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">siquant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">msexp</span>, method <span class="op">=</span> <span class="st">"SIn"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pSet-class.html">processingData</a></span><span class="op">(</span><span class="va">siquant</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Data loaded: Tue Apr 30 12:23:50 2024 </span></span>
<span><span class="co">## Filtered 2 unidentified peptides out [Tue Apr 30 12:23:51 2024] </span></span>
<span><span class="co">## Quantitation by total ion current [Tue Apr 30 12:23:57 2024] </span></span>
<span><span class="co">## Combined 3 features into 3 using sum: Tue Apr 30 12:23:57 2024 </span></span>
<span><span class="co">## Quantification by SIn [Tue Apr 30 12:23:57 2024] </span></span>
<span><span class="co">##  MSnbase version: 2.29.5</span></span></code></pre>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">siquant</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         dummyiTRAQ.mzXML</span></span>
<span><span class="co">## ECA0510     0.0006553518</span></span>
<span><span class="co">## ECA0984     0.0035384487</span></span>
<span><span class="co">## ECA1028     0.0002684726</span></span></code></pre>
<p>Other label-free methods can be applied by specifiying the
appropriate <code>method</code> argument. See <code><a href="../reference/quantify-methods.html">?quantify</a></code> for
more details.</p>
</div>
</div>
<div class="section level2">
<h2 id="spectra-comparison">Spectra comparison<a class="anchor" aria-label="anchor" href="#spectra-comparison"></a>
</h2>
<div class="section level3">
<h3 id="plotting-two-spectra">Plotting two spectra<a class="anchor" aria-label="anchor" href="#plotting-two-spectra"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
provides functionality to compare spectra against each other. The first
notable function is <code>plot</code>. If two <em>Spectrum2</em> objects
are provided <code>plot</code> will draw two plots: the upper and lower
panel contain respectively the first and second spectrum. Common peaks
are drawn in a slightly darker colour.</p>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">centroided</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pickPeaks-method.html">pickPeaks</a></span><span class="op">(</span><span class="va">itraqdata</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">centroided</span><span class="op">)</span><span class="op">[</span>, <span class="st">"PeptideSequence"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"TAGIQIVADDLTVTNPK"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 41 42</span></span></code></pre>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Spectrum2-class.html">precursorMz</a></span><span class="op">(</span><span class="va">centroided</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span></span>
<span><span class="va">zk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Spectrum2-class.html">precursorCharge</a></span><span class="op">(</span><span class="va">centroided</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span></span>
<span><span class="va">mzk</span> <span class="op">*</span> <span class="va">zk</span></span></code></pre></div>
<pre><code><span><span class="co">##      X46      X47 </span></span>
<span><span class="co">## 2046.175 2045.169</span></span></code></pre>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot-methods.html">plot</a></span><span class="op">(</span><span class="va">centroided</span><span class="op">[[</span><span class="va">k</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, <span class="va">centroided</span><span class="op">[[</span><span class="va">k</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/compms2plot-1.png" alt="Comparing two MS2 spectra." width="700"><p class="caption">
Comparing two MS2 spectra.
</p>
</div>
</div>
<div class="section level3">
<h3 id="comparison-metrics">Comparison metrics<a class="anchor" aria-label="anchor" href="#comparison-metrics"></a>
</h3>
<p>Currently <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
supports three different metrics to compare spectra against each other:
<code>common</code> to calculate the number of common peaks,
<code>cor</code> to calculate the Pearson correlation and
<code>dotproduct</code> to calculate the dot product. See
<code><a href="../reference/compareSpectra-methods.html">?compareSpectra</a></code> to apply other arbitrary metrics.</p>
<div class="sourceCode" id="cb159"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compareSpectra-methods.html">compareSpectra</a></span><span class="op">(</span><span class="va">centroided</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">centroided</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>,</span>
<span>               fun <span class="op">=</span> <span class="st">"common"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 8</span></span></code></pre>
<div class="sourceCode" id="cb161"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compareSpectra-methods.html">compareSpectra</a></span><span class="op">(</span><span class="va">centroided</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">centroided</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>,</span>
<span>               fun <span class="op">=</span> <span class="st">"cor"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.1105021</span></span></code></pre>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/compareSpectra-methods.html">compareSpectra</a></span><span class="op">(</span><span class="va">centroided</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">centroided</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>,</span>
<span>               fun <span class="op">=</span> <span class="st">"dotproduct"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.1185025</span></span></code></pre>
<p><code>compareSpectra</code> supports <em>MSnExp</em> objects as
well.</p>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compareSpectra-methods.html">compareSpectra</a></span><span class="op">(</span><span class="va">centroided</span>, fun<span class="op">=</span><span class="st">"cor"</span><span class="op">)</span></span>
<span><span class="va">compmat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##             X1        X10        X11        X12        X13</span></span>
<span><span class="co">## X1          NA 0.07672973 0.38024702 0.51579989 0.46647324</span></span>
<span><span class="co">## X10 0.07672973         NA 0.11050214 0.11162512 0.08611906</span></span>
<span><span class="co">## X11 0.38024702 0.11050214         NA 0.47184437 0.47905818</span></span>
<span><span class="co">## X12 0.51579989 0.11162512 0.47184437         NA 0.57909089</span></span>
<span><span class="co">## X13 0.46647324 0.08611906 0.47905818 0.57909089         NA</span></span>
<span><span class="co">## X14 0.09999703 0.01558385 0.12165400 0.12057251 0.11853321</span></span>
<span><span class="co">## X15 0.03314059 0.00416184 0.01733228 0.04796236 0.03196115</span></span>
<span><span class="co">## X16 0.39140514 0.06634870 0.42259036 0.45624614 0.45469020</span></span>
<span><span class="co">## X17 0.37945538 0.07188420 0.52292845 0.44791250 0.43679447</span></span>
<span><span class="co">## X18 0.55367861 0.10286983 0.56621755 0.66884285 0.64262061</span></span></code></pre>
<p>Below, we illustrate how to compare a set of spectra using a
hierarchical clustering.</p>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot-methods.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="va">compmat</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="v01-MSnbase-demo_files/figure-html/dendo-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="sec:incompdissoc">Quantitative assessment of incomplete dissociation<a class="anchor" aria-label="anchor" href="#sec:incompdissoc"></a>
</h2>
<p>Quantitation using isobaric reporter tags assumes complete
dissociation between the reporter group (red on the <a href="#fig:itraqchem">figure below</a>), balance group (blue) and
peptide (the peptide reactive group is drawn in green). However,
incomplete dissociation does occur and results in an isobaric tag (i.e
reporter and balance groups) specific peaks.</p>
<div class="float" id="fig:itraqchem">
<img src="Figures/itraqchem.png" alt="iTRAQ 4-plex isobaric tags reagent consist of three parts: (1) a charged reporter group (MZ of 114, 115, 116 and 117) that is unique to each of the four reagents (red), (2) an uncharged mass balance group (28-31 Da) (blue)and (3) a peptide reactive group (NHS ester) that binds to the peptide. In case of incomplete dissociation, the reporter and balance groups produce a specific peaks at MZ 145."><div class="figcaption">iTRAQ 4-plex isobaric tags reagent consist of
three parts: (1) a charged reporter group (MZ of 114, 115, 116 and 117)
that is unique to each of the four reagents (red), (2) an uncharged mass
balance group (28-31 Da) (blue)and (3) a peptide reactive group (NHS
ester) that binds to the peptide. In case of incomplete dissociation,
the reporter and balance groups produce a specific peaks at MZ
145.</div>
</div>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
provides, among others, a <em>ReporterIons</em> object for iTRAQ 4-plex
that includes the 145 peaks, called <em>iTRAQ5</em>. This can then be
used to quantify the experiment as show in section @ref(sec:quant) to
estimate incomplete dissociation for each spectrum.</p>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">iTRAQ5</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "ReporterIons"</span></span>
<span><span class="co">## iTRAQ5: '4-plex iTRAQ and reporter + balance group' with 5 reporter ions</span></span>
<span><span class="co">##  - [iTRAQ5.114] 114.1112 +/- 0.05 (red)</span></span>
<span><span class="co">##  - [iTRAQ5.115] 115.1083 +/- 0.05 (green)</span></span>
<span><span class="co">##  - [iTRAQ5.116] 116.1116 +/- 0.05 (blue)</span></span>
<span><span class="co">##  - [iTRAQ5.117] 117.115 +/- 0.05 (yellow)</span></span>
<span><span class="co">##  - [iTRAQ5.145] 145.1 +/- 0.05 (grey)</span></span></code></pre>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">incompdiss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">itraqdata</span>,</span>
<span>                       method <span class="op">=</span> <span class="st">"trap"</span>,</span>
<span>                       reporters <span class="op">=</span> <span class="va">iTRAQ5</span>,</span>
<span>                       strict <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">incompdiss</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     iTRAQ5.114 iTRAQ5.115 iTRAQ5.116 iTRAQ5.117 iTRAQ5.145</span></span>
<span><span class="co">## X1   1347.6158  2247.3097  3927.6931  7661.1463  2063.8947</span></span>
<span><span class="co">## X10   739.9861   799.3501   712.5983   940.6793   467.3615</span></span>
<span><span class="co">## X11 27638.3582 33394.0252 32104.2879 26628.7278 13543.4565</span></span>
<span><span class="co">## X12 31892.8928 33634.6980 37674.7272 37227.7119 11839.2558</span></span>
<span><span class="co">## X13 26143.7542 29677.4781 29089.0593 27902.5608 12206.5508</span></span>
<span><span class="co">## X14  6448.0829  6234.1957  6902.8903  6437.2303   427.6654</span></span></code></pre>
<p>Figure @ref(fig:incompdissPlot) compares these intensities for the
whole experiment.</p>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/incompdissPlot-1.png" alt="Boxplot and scatterplot comparing intensities of the 4 reporter ions (or their sum, on the right) and the incomplete dissociation specific peak." width="700"><p class="caption">
Boxplot and scatterplot comparing intensities of the 4 reporter ions (or
their sum, on the right) and the incomplete dissociation specific peak.
</p>
</div>
</div>
<div class="section level2">
<h2 id="sec:combine">Combining <em>MSnSet</em> instances<a class="anchor" aria-label="anchor" href="#sec:combine"></a>
</h2>
<p>Combining mass spectrometry runs can be done in two different ways
depending on the nature of these runs. If the runs represent repeated
measures of identical samples, for instance multiple fractions, the data
has to be combine along the row of the quantitation matrix: all the
features (along the rows) represent measurements of the same set of
samples (along the columns). In this situation, described in section
@ref(sec:comb1), two experiments of dimensions <span class="math inline">\(n_1\)</span> (rows) by <span class="math inline">\(m\)</span> (columns and <span class="math inline">\(n_2\)</span> by <span class="math inline">\(m\)</span> will produce a new experiment of
dimensions <span class="math inline">\(n_1 + n_2\)</span> by <span class="math inline">\(m\)</span>.</p>
<p>When however, different sets of samples have been analysed in
different mass spectrometry runs, the data has to be combined along the
columns of the quantitation matrix: some features will be shared across
experiments and should thus be aligned on a same row in the new data
set, whereas unique features to one experiment should be set as missing
in the other one. In this situation, described in section
@ref(sec:comb2), two experiments of dimensions <span class="math inline">\(n_1\)</span> by <span class="math inline">\(m_1\)</span> and <span class="math inline">\(n_2\)</span> by <span class="math inline">\(m_2\)</span> will produce a new experiment of
dimensions <span class="math inline">\(unique_{n_1} + unique_{n_2} +
shared_{n_1, n_2}\)</span> by <span class="math inline">\(m_1 +
m_2\)</span>. The two first terms of the first dimension will be
complemented by <code>NA</code> values.</p>
<p>Default <em>MSnSet</em> feature names (<code>X1</code>,
<code>X2</code>, …) and sample names (<code>iTRAQ4.114</code>,
<code>iTRAQ4.115</code>, <code>iTRAQ4.116</code>, …) are not
informative. The features and samples of these anonymous quantitative
data-sets should be updated before being combined, to guide how to
meaningfully merge them.</p>
<div class="section level3">
<h3 id="sec:comb1">Combining identical samples<a class="anchor" aria-label="anchor" href="#sec:comb1"></a>
</h3>
<p>To simulate this situation, let us use quantiation data from the
<code>itraqdata</code> object that is provided with the package as
experiment 1 and the data from the <code>rawdata</code> <em>MSnExp</em>
instance created at the very beginning of this document. Both
experiments share the <em>same</em> default iTRAQ 4-plex reporter names
as default sample names, and will thus automatically be combined along
rows.</p>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">itraqdata</span>, reporters <span class="op">=</span> <span class="va">iTRAQ4</span>,</span>
<span>                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">sampleNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "iTRAQ4.114" "iTRAQ4.115" "iTRAQ4.116" "iTRAQ4.117"</span></span></code></pre>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Spectrum-class.html">centroided</a></span><span class="op">(</span><span class="va">rawdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">rawdata</span>, reporters <span class="op">=</span> <span class="va">iTRAQ4</span>,</span>
<span>                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">sampleNames</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "iTRAQ4.114" "iTRAQ4.115" "iTRAQ4.116" "iTRAQ4.117"</span></span></code></pre>
<p>It important to note that the features of these independent
experiments share the same default anonymous names: X1, X2, X3, …, that
however represent quantitation of distinct physical analytes. If the
experiments were to be combined as is, it would result in an error
because data points for the same <em>feature</em> name (say
<code>X1</code>) and the same <em>sample name</em> (say
<code>iTRAQ4.114</code>) have different values. We thus first update the
feature names to explicitate that they originate from different
experiment and represent quantitation from different spectra using the
convenience function <code>updateFeatureNames</code>. Note that updating
the names of one experiment would suffice here.</p>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "X1"  "X10" "X11" "X12" "X13" "X14"</span></span></code></pre>
<div class="sourceCode" id="cb179"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSnSet-class.html">updateFeatureNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "X1.exp1"  "X10.exp1" "X11.exp1" "X12.exp1" "X13.exp1" "X14.exp1"</span></span></code></pre>
<div class="sourceCode" id="cb181"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "F1.S1" "F1.S2" "F1.S3" "F1.S4" "F1.S5"</span></span></code></pre>
<div class="sourceCode" id="cb183"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSnSet-class.html">updateFeatureNames</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "F1.S1.exp2" "F1.S2.exp2" "F1.S3.exp2" "F1.S4.exp2" "F1.S5.exp2"</span></span></code></pre>
<p>The two experiments now share the same sample names and have
different feature names and will be combined along the row. Note that
all meta-data is correctly combined along the quantitation values.</p>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/combine.html" class="external-link">combine</a></span><span class="op">(</span><span class="va">exp1</span>, <span class="va">exp2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in combine(experimentData(x), experimentData(y)): </span></span>
<span><span class="co">##   unknown or conflicting information in MIAPE field 'email'; using information from first object 'x'</span></span></code></pre>
<div class="sourceCode" id="cb187"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pSet-class.html">dim</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 55  4</span></span></code></pre>
<div class="sourceCode" id="cb189"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pSet-class.html">dim</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 5 4</span></span></code></pre>
<div class="sourceCode" id="cb191"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pSet-class.html">dim</a></span><span class="op">(</span><span class="va">exp12</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 60  4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="sec:comb2">Combine different samples<a class="anchor" aria-label="anchor" href="#sec:comb2"></a>
</h3>
<p>Lets now create two <em>MSnSet</em>s from the same raw data to
simulate two different independent experiments that share some features.
As done previously (see section @ref(sec:feataggregation)), we combine
the spectra based on the proteins they have been identified to belong
to. Features can thus naturally be named using protein accession
numbers. Alternatively, if peptide sequences would have been used as
grouping factor in <code>combineFeatures</code>, then these would be
good feature name candidates.</p>
<div class="sourceCode" id="cb193"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="../reference/pSet-class.html">length</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span>, <span class="fl">35</span><span class="op">)</span></span>
<span><span class="va">j</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="../reference/pSet-class.html">length</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span>, <span class="fl">35</span><span class="op">)</span></span>
<span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, reporters <span class="op">=</span> <span class="va">iTRAQ4</span>,</span>
<span>                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/quantify-methods.html">quantify</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>, reporters <span class="op">=</span> <span class="va">iTRAQ4</span>,</span>
<span>                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##    14    21</span></span></code></pre>
<div class="sourceCode" id="cb195"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineFeatures.html">combineFeatures</a></span><span class="op">(</span><span class="va">exp1</span>,</span>
<span>                        groupBy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span><span class="op">$</span><span class="va">ProteinAccession</span><span class="op">)</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combineFeatures.html">combineFeatures</a></span><span class="op">(</span><span class="va">exp2</span>,</span>
<span>                        groupBy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">$</span><span class="va">ProteinAccession</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Your data contains missing values. Please read the relevant section in</span></span>
<span><span class="co">## the combineFeatures manual page for details on the effects of missing</span></span>
<span><span class="co">## values on data aggregation.</span></span></code></pre>
<div class="sourceCode" id="cb197"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BSA"     "ECA0172" "ECA0469" "ECA0631" "ECA0691" "ECA0871"</span></span></code></pre>
<div class="sourceCode" id="cb199"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BSA"     "ECA0435" "ECA0469" "ECA0621" "ECA0871" "ECA1032"</span></span></code></pre>
<p>The <code>droplevels</code> drops the unused <code>featureData</code>
levels. This is required to avoid passing absent levels as
<code>groupBy</code> in <code>combineFeatures</code>. Alternatively, one
could also use <code>factor(fData(exp1)\$ProteinAccession)</code> as
<code>groupBy</code> argument.</p>
<p>The feature names are updated automatically by
<code>combineFeatures</code>, using the <code>groupBy</code> argument.
Proper feature names, reflecting the nature of the features (spectra,
peptides or proteins) is critical when multiple experiments are to be
combined, as this is done using common features as defined by their
names (see below).</p>
<p>Sample names should also be updated to replace anonymous reporter
names with relevant identifiers; the individual reporter data is stored
in the <code>phenoData</code> and is not lost. A convenience function
<code>updateSampleNames</code> is provided to append the
<em>MSnSet</em>’s variable name to the already defined names, although
in general, biologically relevant identifiers are preferred.</p>
<div class="sourceCode" id="cb201"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">sampleNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "iTRAQ4.114" "iTRAQ4.115" "iTRAQ4.116" "iTRAQ4.117"</span></span></code></pre>
<div class="sourceCode" id="cb203"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSnSet-class.html">updateSampleNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">sampleNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "iTRAQ4.114.exp1" "iTRAQ4.115.exp1" "iTRAQ4.116.exp1" "iTRAQ4.117.exp1"</span></span></code></pre>
<div class="sourceCode" id="cb205"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">sampleNames</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Ctrl1"</span>, <span class="st">"Cond1"</span>, <span class="st">"Ctrl2"</span>, <span class="st">"Cond2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">sampleNames</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Ctrl3"</span>, <span class="st">"Cond3"</span>, <span class="st">"Ctrl4"</span>, <span class="st">"Cond4"</span><span class="op">)</span></span></code></pre></div>
<p>At this stage, it is not yet possible to combine the two experiments,
because their feature data is not compatible yet; they share the same
feature variable labels, i.e. the feature data column names (spectrum,
ProteinAccession, ProteinDescription, …), but the part of the content is
different because the original data was (in particular all the spectrum
centric data: identical peptides in different runs will have different
retention times, precursor intensities, …). Feature data with identical
labels (columns in the data frame) and names (row in the data frame) are
expected to have the same data and produce an error if not conform.</p>
<div class="sourceCode" id="cb206"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fvarLabels</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fvarLabels</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span><span class="op">[</span><span class="st">"BSA"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##     spectrum ProteinAccession   ProteinDescription PeptideSequence</span></span>
<span><span class="co">## BSA        1              BSA bovine serum albumin          NYQEAK</span></span></code></pre>
<div class="sourceCode" id="cb208"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">[</span><span class="st">"BSA"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##     spectrum ProteinAccession   ProteinDescription PeptideSequence</span></span>
<span><span class="co">## BSA        1              BSA bovine serum albumin          NYQEAK</span></span></code></pre>
<p>Instead of removing these identical feature data columns, one can use
a second convenience function, <code>updateFvarLabels</code>, to update
feature labels based on the experiements variable name and maintain all
the metadata.</p>
<div class="sourceCode" id="cb210"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSnSet-class.html">updateFvarLabels</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSnSet-class.html">updateFvarLabels</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fvarLabels</a></span><span class="op">(</span><span class="va">exp1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "spectrum.exp1"           "ProteinAccession.exp1"  </span></span>
<span><span class="co">## [3] "ProteinDescription.exp1" "PeptideSequence.exp1"   </span></span>
<span><span class="co">## [5] "fileIdx.exp1"            "retention.time.exp1"</span></span></code></pre>
<div class="sourceCode" id="cb212"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fvarLabels</a></span><span class="op">(</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "spectrum.exp2"           "ProteinAccession.exp2"  </span></span>
<span><span class="co">## [3] "ProteinDescription.exp2" "PeptideSequence.exp2"   </span></span>
<span><span class="co">## [5] "fileIdx.exp2"            "retention.time.exp2"</span></span></code></pre>
<p>It is now possible to combine <code>exp1</code> and
<code>exp2</code>, including all the meta-data, with the
<code>combine</code> method. The new experiment will contain the union
of the feature names of the individual experiments with missing values
inserted appropriately.</p>
<div class="sourceCode" id="cb214"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/combine.html" class="external-link">combine</a></span><span class="op">(</span><span class="va">exp1</span>, <span class="va">exp2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pSet-class.html">dim</a></span><span class="op">(</span><span class="va">exp12</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 36  8</span></span></code></pre>
<div class="sourceCode" id="cb216"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="va">exp12</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             mz reporters</span></span>
<span><span class="co">## Ctrl1 114.1112    iTRAQ4</span></span>
<span><span class="co">## Cond1 115.1083    iTRAQ4</span></span>
<span><span class="co">## Ctrl2 116.1116    iTRAQ4</span></span>
<span><span class="co">## Cond2 117.1150    iTRAQ4</span></span>
<span><span class="co">## Ctrl3 114.1112    iTRAQ4</span></span>
<span><span class="co">## Cond3 115.1083    iTRAQ4</span></span>
<span><span class="co">## Ctrl4 116.1116    iTRAQ4</span></span>
<span><span class="co">## Cond4 117.1150    iTRAQ4</span></span></code></pre>
<div class="sourceCode" id="cb218"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">exp12</span><span class="op">)</span><span class="op">[</span><span class="fl">25</span><span class="op">:</span><span class="fl">28</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##             Ctrl1     Cond1     Ctrl2     Cond2     Ctrl3     Cond3     Ctrl4</span></span>
<span><span class="co">## ECA4513 10154.953 10486.943 11018.191 11289.552        NA        NA        NA</span></span>
<span><span class="co">## ECA4514  6516.397  6746.337  6658.973  6838.491 12457.173 12695.491 14423.640</span></span>
<span><span class="co">## ENO     77239.040 49352.793 22493.050 11187.588 39965.733 24967.397        NA</span></span>
<span><span class="co">## ECA0435        NA        NA        NA        NA  4923.628  5557.818  5775.203</span></span>
<span><span class="co">##             Cond4</span></span>
<span><span class="co">## ECA4513        NA</span></span>
<span><span class="co">## ECA4514 14556.855</span></span>
<span><span class="co">## ENO      5925.663</span></span>
<span><span class="co">## ECA0435  5079.295</span></span></code></pre>
<div class="sourceCode" id="cb220"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp12</span></span></code></pre></div>
<pre><code><span><span class="co">## MSnSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 36 features, 8 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: Ctrl1 Cond1 ... Cond4 (8 total)</span></span>
<span><span class="co">##   varLabels: mz reporters</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: BSA ECA0172 ... ECA4037 (36 total)</span></span>
<span><span class="co">##   fvarLabels: spectrum.exp1 ProteinAccession.exp1 ...</span></span>
<span><span class="co">##     CV.iTRAQ4.117.exp2 (38 total)</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:  </span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Combined [27,4] and [24,4] MSnSets Tue Apr 30 12:24:04 2024 </span></span>
<span><span class="co">##  MSnbase version: 2.29.5</span></span></code></pre>
<p>In summary, when experiments with different samples need to be
combined (along the columns), one needs to (1) clarify the sample names
using <code>updateSampleNames</code> or better manually, for biological
relevance and (2) update the feature data variable labels with
<code>updateFvarLabels</code>. The individual experiments (there can be
more than 2) can then easily be combined with the <code>combine</code>
method while retaining the meta-data.</p>
<p>If runs for the same sample (different fractions for example) need to
be combines, one needs to (1) differentiate the feature provenance with
<code>updateFeatureNames</code> prior to use <code>combine</code>.</p>
</div>
<div class="section level3">
<h3 id="sec:split">Splitting and unsplitting <em>MSnSet</em> instances<a class="anchor" aria-label="anchor" href="#sec:split"></a>
</h3>
<p>A single <em>MSnSet</em> can also be split along the features/rows or
samples/columns using the <code>split</code> method and a factor
defining the splitting groups, resulting in an instance of class
<em>MSnSetList</em>:</p>
<div class="sourceCode" id="cb222"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dunkley2006</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="va">dunkley2006</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        membrane.prep fraction replicate</span></span>
<span><span class="co">## M1F1A              1        1         A</span></span>
<span><span class="co">## M1F4A              1        4         A</span></span>
<span><span class="co">## M1F7A              1        7         A</span></span>
<span><span class="co">## M1F11A             1       11         A</span></span>
<span><span class="co">## M1F2B              1        2         B</span></span>
<span><span class="co">## M1F5B              1        5         B</span></span></code></pre>
<div class="sourceCode" id="cb224"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">dunkley2006</span>, <span class="va">dunkley2006</span><span class="op">$</span><span class="va">replicate</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Instance of class 'MSnSetList' containig 2 objects.</span></span></code></pre>
<div class="sourceCode" id="cb226"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## or, defining the appropriate annotation variable name</span></span>
<span><span class="va">dun</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">dunkley2006</span>, <span class="st">"replicate"</span><span class="op">)</span></span></code></pre></div>
<p>Above, we split along the columns/samples, but the function would
equally work with a factor of length equal to the number of rows of the
<em>MSnSet</em> (or a feature variable name) to split along the
rows/features.</p>
<p>Finally, the effect of <code>split</code> can be reverted by
<code>unsplit</code>.</p>
<div class="sourceCode" id="cb227"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dun2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unsplit.html" class="external-link">unsplit</a></span><span class="op">(</span><span class="va">dun</span>, <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="va">dunkley2006</span><span class="op">)</span><span class="op">$</span><span class="va">replicate</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/compareMSnSets.html">compareMSnSets</a></span><span class="op">(</span><span class="va">dunkley2006</span>, <span class="va">dun2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>See <code><a href="../reference/MSnSetList-class.html">?MSnSetList</a></code> for more details about the class,
<code>split</code> and <code>unsplit</code> and comments about storing
multiple assays pertaining the same experiment.</p>
</div>
<div class="section level3">
<h3 id="sec:avg">Averaging <em>MSnSet</em> instances<a class="anchor" aria-label="anchor" href="#sec:avg"></a>
</h3>
<p>It is sometimes useful to average a set of replicated experiments to
facilitate their visualisation. This can be easily achieved with the
<code>averageMSnSet</code> function, which takes a list of valid
<em>MSnSet</em> instances as input and creates a new object whose
expression values are an average of the original values. A value of
dispersion (<code>disp</code>) and a count of missing values
(<code>nNA</code>) is recorded in the feature metadata slot. The average
and dispersion are computed by default as the median and
(non-parametric) coefficient of variation (see <code><a href="../reference/npcv.html">?npcv</a></code> for
details), although this can easily be parametrised, as described in
<code><a href="../reference/averageMSnSet.html">?averageMSnSet</a></code>.</p>
<p>The next code chunk illustrates the averaging function using three
replicated experiments from <span class="citation">(Tan et al.
2009)</span> available in the <em><a href="https://bioconductor.org/packages/3.18/pRolocdata" class="external-link">pRolocdata</a></em>
package.</p>
<div class="sourceCode" id="cb229"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/pRolocdata" class="external-link">"pRolocdata"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tan2009r2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tan2009r3</span><span class="op">)</span></span>
<span><span class="va">msnl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MSnSetList-class.html">MSnSetList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">tan2009r1</span>, <span class="va">tan2009r2</span>, <span class="va">tan2009r3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">avgtan</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/averageMSnSet.html">averageMSnSet</a></span><span class="op">(</span><span class="va">msnl</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">avgtan</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             X114      X115      X116      X117</span></span>
<span><span class="co">## P20353 0.3605000 0.3035000 0.2095000 0.1265000</span></span>
<span><span class="co">## P53501 0.4299090 0.1779700 0.2068280 0.1852625</span></span>
<span><span class="co">## Q7KU78 0.1704443 0.1234443 0.1772223 0.5290000</span></span>
<span><span class="co">## P04412 0.2567500 0.2210000 0.3015000 0.2205000</span></span>
<span><span class="co">## Q7KJ73 0.2160000 0.1830000 0.3420000 0.2590000</span></span>
<span><span class="co">## Q7JZN0 0.0965000 0.2509443 0.4771667 0.1750557</span></span></code></pre>
<div class="sourceCode" id="cb231"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">avgtan</span><span class="op">)</span><span class="op">$</span><span class="va">disp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               X114      X115        X116       X117</span></span>
<span><span class="co">## P20353 0.076083495 0.1099127 0.109691169 0.14650198</span></span>
<span><span class="co">## P53501 0.034172542 0.2640556 0.005139653 0.17104568</span></span>
<span><span class="co">## Q7KU78 0.023198743 0.4483795 0.027883087 0.04764499</span></span>
<span><span class="co">## P04412 0.053414021 0.2146751 0.090972139 0.27903810</span></span>
<span><span class="co">## Q7KJ73 0.000000000 0.0000000 0.000000000 0.00000000</span></span>
<span><span class="co">## Q7JZN0 0.007681865 0.1959534 0.097873350 0.06210542</span></span></code></pre>
<div class="sourceCode" id="cb233"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">avgtan</span><span class="op">)</span><span class="op">$</span><span class="va">nNA</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        X114 X115 X116 X117</span></span>
<span><span class="co">## P20353    1    1    1    1</span></span>
<span><span class="co">## P53501    1    1    1    1</span></span>
<span><span class="co">## Q7KU78    0    0    0    0</span></span>
<span><span class="co">## P04412    1    1    1    1</span></span>
<span><span class="co">## Q7KJ73    2    2    2    2</span></span>
<span><span class="co">## Q7JZN0    0    0    0    0</span></span></code></pre>
<p>We are going to visualise the average data on a principle component
(PCA) plot using the <code>plot2D</code> function from the <em><a href="https://bioconductor.org/packages/3.18/pRoloc" class="external-link">pRoloc</a></em>
package <span class="citation">(L. Gatto et al. 2014)</span>. In
addition, we are going to use the measure of dispersion to highlight
averages with high variability by taking, for each protein, the maximum
observed dispersion in the 4 samples. Note that in the default
implementation, dispersions estimated from a single measurement
(i.e. that had 2 missing values in our example) are set to 0; we will
set these to the overal maximum observed dispersion.</p>
<div class="sourceCode" id="cb235"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/rowQ.html" class="external-link">rowMax</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">avgtan</span><span class="op">)</span><span class="op">$</span><span class="va">disp</span><span class="op">)</span></span>
<span><span class="va">disp</span><span class="op">[</span><span class="va">disp</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">disp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">disp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.01152877 1.20888923</span></span></code></pre>
<div class="sourceCode" id="cb237"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/pRoloc" class="external-link">"pRoloc"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pRoloc/man/plot2D.html" class="external-link">plot2D</a></span><span class="op">(</span><span class="va">avgtan</span>, cex <span class="op">=</span> <span class="fl">3</span> <span class="op">*</span> <span class="va">disp</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="v01-MSnbase-demo_files/figure-html/plot2Davg-1.png" alt="PCA plot of the averaged *MSnSet*. The point sizes are proportional to the dispersion of the protein quantitation across the averaged data." width="700"><p class="caption">
PCA plot of the averaged <em>MSnSet</em>. The point sizes are
proportional to the dispersion of the protein quantitation across the
averaged data.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="sec:mse">MS<sup>E</sup> data processing<a class="anchor" aria-label="anchor" href="#sec:mse"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
can also be used for MS<sup>E</sup> data independent acquisition from
Waters instrument. The MS<sup>E</sup> pipeline depends on the
Bioconductor <em><a href="https://bioconductor.org/packages/3.18/synapter" class="external-link">synapter</a></em>
package <span class="citation">(Bond et al. 2013)</span> that produces
<em>MSnSet</em> instances for indvidual acquisitions. The <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
infrastructure can subsequently be used to further combine experiments,
as shown in section @ref(sec:comb2) and apply <em>top3</em> quantitation
using the <code>topN</code> method.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] gplots_3.1.3.1       msdata_0.42.0        pRoloc_1.42.0       </span></span>
<span><span class="co">##  [4] BiocParallel_1.36.0  MLInterfaces_1.82.0  cluster_2.1.6       </span></span>
<span><span class="co">##  [7] annotate_1.80.0      XML_3.99-0.16.1      AnnotationDbi_1.64.1</span></span>
<span><span class="co">## [10] IRanges_2.36.0       pRolocdata_1.40.0    Rdisop_1.62.0       </span></span>
<span><span class="co">## [13] zoo_1.8-12           MSnbase_2.29.5       ProtGenerics_1.34.0 </span></span>
<span><span class="co">## [16] S4Vectors_0.40.2     mzR_2.36.0           Rcpp_1.0.12         </span></span>
<span><span class="co">## [19] Biobase_2.62.0       BiocGenerics_0.48.1  ggplot2_3.5.1       </span></span>
<span><span class="co">## [22] BiocStyle_2.30.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] splines_4.3.3               bitops_1.0-7               </span></span>
<span><span class="co">##   [3] filelock_1.0.3              tibble_3.2.1               </span></span>
<span><span class="co">##   [5] hardhat_1.3.1               preprocessCore_1.61.0      </span></span>
<span><span class="co">##   [7] pROC_1.18.5                 rpart_4.1.23               </span></span>
<span><span class="co">##   [9] lifecycle_1.0.4             doParallel_1.0.17          </span></span>
<span><span class="co">##  [11] globals_0.16.3              lattice_0.22-6             </span></span>
<span><span class="co">##  [13] MASS_7.3-60.0.1             MultiAssayExperiment_1.28.0</span></span>
<span><span class="co">##  [15] dendextend_1.17.1           magrittr_2.0.3             </span></span>
<span><span class="co">##  [17] limma_3.58.1                plotly_4.10.4              </span></span>
<span><span class="co">##  [19] sass_0.4.9                  rmarkdown_2.26             </span></span>
<span><span class="co">##  [21] jquerylib_0.1.4             yaml_2.3.8                 </span></span>
<span><span class="co">##  [23] MsCoreUtils_1.14.1          DBI_1.2.2                  </span></span>
<span><span class="co">##  [25] RColorBrewer_1.1-3          lubridate_1.9.3            </span></span>
<span><span class="co">##  [27] abind_1.4-5                 zlibbioc_1.48.2            </span></span>
<span><span class="co">##  [29] GenomicRanges_1.54.1        purrr_1.0.2                </span></span>
<span><span class="co">##  [31] mixtools_2.0.0              AnnotationFilter_1.26.0    </span></span>
<span><span class="co">##  [33] RCurl_1.98-1.14             nnet_7.3-19                </span></span>
<span><span class="co">##  [35] rappdirs_0.3.3              ipred_0.9-14               </span></span>
<span><span class="co">##  [37] lava_1.8.0                  GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">##  [39] listenv_0.9.1               parallelly_1.37.1          </span></span>
<span><span class="co">##  [41] pkgdown_2.0.9.9000          ncdf4_1.22                 </span></span>
<span><span class="co">##  [43] codetools_0.2-20            DelayedArray_0.28.0        </span></span>
<span><span class="co">##  [45] xml2_1.3.6                  tidyselect_1.2.1           </span></span>
<span><span class="co">##  [47] farver_2.1.1                viridis_0.6.5              </span></span>
<span><span class="co">##  [49] matrixStats_1.3.0           BiocFileCache_2.10.2       </span></span>
<span><span class="co">##  [51] jsonlite_1.8.8              caret_6.0-94               </span></span>
<span><span class="co">##  [53] e1071_1.7-14                survival_3.6-4             </span></span>
<span><span class="co">##  [55] iterators_1.0.14            systemfonts_1.0.6          </span></span>
<span><span class="co">##  [57] foreach_1.5.2               segmented_2.0-4            </span></span>
<span><span class="co">##  [59] tools_4.3.3                 progress_1.2.3             </span></span>
<span><span class="co">##  [61] ragg_1.3.0                  glue_1.7.0                 </span></span>
<span><span class="co">##  [63] prodlim_2023.08.28          gridExtra_2.3              </span></span>
<span><span class="co">##  [65] SparseArray_1.2.4           mgcv_1.9-1                 </span></span>
<span><span class="co">##  [67] xfun_0.43                   MatrixGenerics_1.14.0      </span></span>
<span><span class="co">##  [69] GenomeInfoDb_1.38.8         dplyr_1.1.4                </span></span>
<span><span class="co">##  [71] withr_3.0.0                 BiocManager_1.30.22        </span></span>
<span><span class="co">##  [73] fastmap_1.1.1               fansi_1.0.6                </span></span>
<span><span class="co">##  [75] caTools_1.18.2              digest_0.6.35              </span></span>
<span><span class="co">##  [77] timechange_0.3.0            R6_2.5.1                   </span></span>
<span><span class="co">##  [79] textshaping_0.3.7           colorspace_2.1-0           </span></span>
<span><span class="co">##  [81] gtools_3.9.5                lpSolve_5.6.20             </span></span>
<span><span class="co">##  [83] biomaRt_2.58.2              RSQLite_2.3.6              </span></span>
<span><span class="co">##  [85] utf8_1.2.4                  tidyr_1.3.1                </span></span>
<span><span class="co">##  [87] generics_0.1.3              hexbin_1.28.3              </span></span>
<span><span class="co">##  [89] data.table_1.15.4           recipes_1.0.10             </span></span>
<span><span class="co">##  [91] FNN_1.1.4                   class_7.3-22               </span></span>
<span><span class="co">##  [93] prettyunits_1.2.0           PSMatch_1.6.0              </span></span>
<span><span class="co">##  [95] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">##  [97] S4Arrays_1.2.1              ModelMetrics_1.2.2.2       </span></span>
<span><span class="co">##  [99] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span><span class="co">## [101] timeDate_4032.109           blob_1.2.4                 </span></span>
<span><span class="co">## [103] impute_1.76.0               XVector_0.42.0             </span></span>
<span><span class="co">## [105] htmltools_0.5.8.1           bookdown_0.39              </span></span>
<span><span class="co">## [107] MALDIquant_1.22.2           clue_0.3-65                </span></span>
<span><span class="co">## [109] scales_1.3.0                png_0.1-8                  </span></span>
<span><span class="co">## [111] gower_1.0.1                 knitr_1.46                 </span></span>
<span><span class="co">## [113] reshape2_1.4.4              coda_0.19-4.1              </span></span>
<span><span class="co">## [115] nlme_3.1-164                curl_5.2.1                 </span></span>
<span><span class="co">## [117] proxy_0.4-27                cachem_1.0.8               </span></span>
<span><span class="co">## [119] stringr_1.5.1               KernSmooth_2.23-22         </span></span>
<span><span class="co">## [121] parallel_4.3.3              mzID_1.40.0                </span></span>
<span><span class="co">## [123] vsn_3.70.0                  desc_1.4.3                 </span></span>
<span><span class="co">## [125] pillar_1.9.0                vctrs_0.6.5                </span></span>
<span><span class="co">## [127] pcaMethods_1.94.0           randomForest_4.7-1.1       </span></span>
<span><span class="co">## [129] dbplyr_2.5.0                xtable_1.8-4               </span></span>
<span><span class="co">## [131] evaluate_0.23               mvtnorm_1.2-4              </span></span>
<span><span class="co">## [133] cli_3.6.2                   compiler_4.3.3             </span></span>
<span><span class="co">## [135] rlang_1.1.3                 crayon_1.5.2               </span></span>
<span><span class="co">## [137] future.apply_1.11.2         labeling_0.4.3             </span></span>
<span><span class="co">## [139] LaplacesDemon_16.1.6        mclust_6.1.1               </span></span>
<span><span class="co">## [141] QFeatures_1.12.0            affy_1.80.0                </span></span>
<span><span class="co">## [143] plyr_1.8.9                  fs_1.6.4                   </span></span>
<span><span class="co">## [145] stringi_1.8.3               viridisLite_0.4.2          </span></span>
<span><span class="co">## [147] munsell_0.5.1               Biostrings_2.70.3          </span></span>
<span><span class="co">## [149] lazyeval_0.2.2              Matrix_1.6-5               </span></span>
<span><span class="co">## [151] hms_1.1.3                   future_1.33.2              </span></span>
<span><span class="co">## [153] bit64_4.0.5                 KEGGREST_1.42.0            </span></span>
<span><span class="co">## [155] statmod_1.5.0               highr_0.10                 </span></span>
<span><span class="co">## [157] SummarizedExperiment_1.32.0 kernlab_0.9-32             </span></span>
<span><span class="co">## [159] igraph_2.0.3                memoise_2.0.1              </span></span>
<span><span class="co">## [161] affyio_1.72.0               bslib_0.7.0                </span></span>
<span><span class="co">## [163] sampling_2.10               bit_4.0.5</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bolstad03" class="csl-entry">
Bolstad, B M, R A Irizarry, M Astrand, and T P Speed. 2003. <span>“<a href="https://www.ncbi.nlm.nih.gov/pubmed/12538238" class="external-link">A Comparison of
Normalization Methods for High Density Oligonucleotide Array Data Based
on Variance and Bias.</a>”</span> <em>Bioinformatics</em> 19 (2):
185–93.
</div>
<div id="ref-Bond2013" class="csl-entry">
Bond, N J, P V Shliaha, K S Lilley, and L Gatto. 2013. <span>“Improving
Qualitative and Quantitative Performance for <span>MS</span><span class="math inline">\(^E\)</span>-Based Label-Free Proteomics.”</span>
<em>J Proteome Res</em> 12 (6): 2340–53. <a href="https://doi.org/10.1021/pr300776t" class="external-link">https://doi.org/10.1021/pr300776t</a>.
</div>
<div id="ref-Foster11" class="csl-entry">
Foster, Joseph M, Sven Degroeve, Laurent Gatto, Matthieu Visser, Rui
Wang, Johannes Griss, Rolf Apweiler, and Lennart Martens. 2011. <span>“A
Posteriori Quality Control for the Curation and Reuse of Public
Proteomics Data.”</span> <em>Proteomics</em> 11 (11): 2182–94. <a href="https://doi.org/10.1002/pmic.201000602" class="external-link">https://doi.org/10.1002/pmic.201000602</a>.
</div>
<div id="ref-Gatto2012" class="csl-entry">
Gatto, Laurent, and Kathryn S Lilley. 2012. <span>“<span>MS</span>nbase
– an <span>R</span>/<span>B</span>ioconductor Package for Isobaric
Tagged Mass Spectrometry Data Visualization, Processing and
Quantitation.”</span> <em>Bioinformatics</em> 28 (2): 288–89. <a href="https://doi.org/10.1093/bioinformatics/btr645" class="external-link">https://doi.org/10.1093/bioinformatics/btr645</a>.
</div>
<div id="ref-Gatto2014" class="csl-entry">
Gatto, L, L M Breckels, S Wieczorek, T Burger, and K S Lilley. 2014.
<span>“Mass-Spectrometry-Based Spatial Proteomics Data Analysis Using
<span class="nocase">pRoloc</span> and <span class="nocase">pRolocdata</span>.”</span> <em>Bioinformatics</em> 30
(9): 1322–24. <a href="https://doi.org/10.1093/bioinformatics/btu013" class="external-link">https://doi.org/10.1093/bioinformatics/btu013</a>.
</div>
<div id="ref-Gentleman2004" class="csl-entry">
Gentleman, Robert C., Vincent J. Carey, Douglas M. Bates, Ben Bolstad,
Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004.
<span>“Bioconductor: Open Software Development for Computational Biology
and Bioinformatics.”</span> <em>Genome Biol</em> 5 (10): –80. <a href="https://doi.org/10.1186/gb-2004-5-10-r80" class="external-link">https://doi.org/10.1186/gb-2004-5-10-r80</a>.
</div>
<div id="ref-Gibb2012" class="csl-entry">
Gibb, S, and K Strimmer. 2012. <span>“<span>MALDIquant</span>: A
Versatile <span>R</span> Package for the Analysis of Mass Spectrometry
Data.”</span> <em>Bioinformatics</em> 28 (17): 2270–71. <a href="https://doi.org/10.1093/bioinformatics/bts447" class="external-link">https://doi.org/10.1093/bioinformatics/bts447</a>.
</div>
<div id="ref-Griffin2010" class="csl-entry">
Griffin, N M, J Yu, F Long, P Oh, S Shore, Y Li, J A Koziol, and J E
Schnitzer. 2010. <span>“Label-Free, Normalized Quantification of Complex
Mass Spectrometry Data for Proteomic Analysis.”</span> <em>Nat
Biotechnol</em> 28 (1): 83–89. <a href="https://doi.org/10.1038/nbt.1592" class="external-link">https://doi.org/10.1038/nbt.1592</a>.
</div>
<div id="ref-Huber2002" class="csl-entry">
Huber, Wolfgang, Anja von Heydebreck, Holger Sueltmann, Annemarie
Poustka, and Martin Vingron. 2002. <span>“Variance Stabilization Applied
to Microarray Data Calibration and to the Quantification of Differential
Expression.”</span> <em>Bioinformatics</em> 18 Suppl. 1: S96–104.
</div>
<div id="ref-Jones2012" class="csl-entry">
Jones, Andrew R., Martin Eisenacher, Gerhard Mayer, Oliver Kohlbacher,
Jennifer Siepen, Simon J. Hubbard, Julian N. Selley, et al. 2012.
<span>“The <span class="nocase">mzIdentML</span> Data Standard for Mass
Spectrometry-Based Proteomics Results.”</span> <em>Molecular &amp;
Cellular Proteomics</em> 11 (7). <a href="https://doi.org/10.1074/mcp.M111.014381" class="external-link">https://doi.org/10.1074/mcp.M111.014381</a>.
</div>
<div id="ref-Karp2010" class="csl-entry">
Karp, Natasha A, Wolfgang Huber, Pawel G Sadowski, Philip D Charles,
Svenja V Hester, and Kathryn S Lilley. 2010. <span>“Addressing Accuracy
and Precision Issues in iTRAQ Quantitation.”</span> <em>Mol. Cell
Proteomics</em> 9 (9): 1885–97. <a href="https://doi.org/10.1074/mcp.M900628-MCP200" class="external-link">https://doi.org/10.1074/mcp.M900628-MCP200</a>.
</div>
<div id="ref-Lazar:2016" class="csl-entry">
Lazar, C, L Gatto, M Ferro, C Bruley, and T Burger. 2016.
<span>“Accounting for the Multiple Natures of Missing Values in
Label-Free Quantitative Proteomics Data Sets to Compare Imputation
Strategies.”</span> <em>J Proteome Res</em> 15 (4): 1116–25. <a href="https://doi.org/10.1021/acs.jproteome.5b00981" class="external-link">https://doi.org/10.1021/acs.jproteome.5b00981</a>.
</div>
<div id="ref-Martens2010" class="csl-entry">
Martens, Lennart, Matthew Chambers, Marc Sturm, Darren Kes sner, Fredrik
Levander, Jim Shofstahl, Wilfred H Tang, et al. 2010. <span>“mzML - a
Community Standard for Mass Spectrometry Data.”</span> <em>Molecular
&amp; Cellular Proteomics : MCP</em>. <a href="https://doi.org/10.1074/mcp.R110.000133" class="external-link">https://doi.org/10.1074/mcp.R110.000133</a>.
</div>
<div id="ref-Orchard2007" class="csl-entry">
Orchard, Sandra, Luisa Montechi-Palazzi, Eric W Deutsch, Pierre-Alain
Binz, Andrew R Jones, Norman Paton, Angel Pizarro, David M Creasy,
Jérôme Wojcik, and Henning Hermjakob. 2007. <span>“Five Years of
Progress in the Standardization of Proteomics Data 4th Annual Spring
Workshop of the HUPO-Proteomics Standards Initiative April 23-25, 2007
Ecole Nationale Supérieure (ENS), Lyon, France.”</span>
<em>Proteomics</em> 7 (19): 3436–40. <a href="https://doi.org/10.1002/pmic.200700658" class="external-link">https://doi.org/10.1002/pmic.200700658</a>.
</div>
<div id="ref-Paoletti2006" class="csl-entry">
Paoletti, A C, T J Parmely, C Tomomori-Sato, S Sato, D Zhu, R C Conaway,
J W Conaway, L Florens, and M P Washburn. 2006. <span>“Quantitative
Proteomic Analysis of Distinct Mammalian Mediator Complexes Using
Normalized Spectral Abundance Factors.”</span> <em>Proc Natl Acad Sci U
S A</em> 103 (50): 18928–33. <a href="https://doi.org/10.1073/pnas.0606379103" class="external-link">https://doi.org/10.1073/pnas.0606379103</a>.
</div>
<div id="ref-Pedrioli2004" class="csl-entry">
Pedrioli, Patrick G A, Jimmy K Eng, Robert Hubley, Mathijs Vogelzang,
Eric W Deutsch, Brian Raught, Brian Pratt, et al. 2004. <span>“A Common
Open Representation of Mass Spectrometry Data and Its Application to
Proteomics Research.”</span> <em>Nat. Biotechnol.</em> 22 (11): 1459–66.
<a href="https://doi.org/10.1038/nbt1031" class="external-link">https://doi.org/10.1038/nbt1031</a>.
</div>
<div id="ref-Rstat" class="csl-entry">
R Development Core Team. 2011. <em>R: A Language and Environment for
Statistical Computing</em>. Vienna, Austria: R Foundation for
Statistical Computing. <a href="http://www.R-project.org/" class="external-link">http://www.R-project.org/</a>.
</div>
<div id="ref-Tan2009" class="csl-entry">
Tan, Denise J, Heidi Dvinge, Andy Christoforou, Paul Bertone, Alfonso A
Martinez, and Kathryn S Lilley. 2009. <span>“Mapping Organelle Proteins
and Protein Complexes in Drosophila Melanogaster.”</span> <em>J Proteome
Res</em> 8 (6): 2667–78. <a href="https://doi.org/10.1021/pr800866n" class="external-link">https://doi.org/10.1021/pr800866n</a>.
</div>
<div id="ref-Taylor2008" class="csl-entry">
Taylor, Chris F, Pierre-Alain Binz, Ruedi Aebersold, Michel Affolter,
Robert Barkovich, Eric W Deutsch, David M Horn, et al. 2008.
<span>“Guidelines for Reporting the Use of Mass Spectrometry in
Proteomics.”</span> <em>Nat. Biotechnol.</em> 26 (8): 860–61. <a href="https://doi.org/10.1038/nbt0808-860" class="external-link">https://doi.org/10.1038/nbt0808-860</a>.
</div>
<div id="ref-Taylor2007" class="csl-entry">
Taylor, Chris F., Norman W. Paton, Kathryn S. Lilley, Pierre-Alain Binz,
Randall K. Julian, Andrew R. Jones, Weimin Zhu, et al. 2007. <span>“The
Minimum Information about a Proteomics Experiment (MIAPE).”</span>
<em>Nat Biotechnol</em> 25 (8): 887–93. <a href="https://doi.org/10.1038/nbt1329" class="external-link">https://doi.org/10.1038/nbt1329</a>.
</div>
<div id="ref-ggplot2" class="csl-entry">
Wickham, Hadley. 2009. <em>Ggplot2: Elegant Graphics for Data
Analysis</em>. Springer New York. <a href="http://had.co.nz/ggplot2/book" class="external-link">http://had.co.nz/ggplot2/book</a>.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>in R, open it with
<code>vignette("MSnbase-development")</code> or read it online <a href="https://bioconductor.org/packages/devel/bioc/vignettes/MSnbase/inst/doc/MSnbase-development.html" class="external-link">here</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>in R, open it with <code>vignette("benchmarking")</code>
or read it online <a href="https://bioconductor.org/packages/devel/bioc/vignettes/MSnbase/inst/doc/benchmarking.html" class="external-link">here</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Mascot Generic Format, see <a href="http://www.matrixscience.com/help/data_file_help.html#GEN" class="external-link uri">http://www.matrixscience.com/help/data_file_help.html#GEN</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>This part will be automatically updated when the object
is modified with it’s <em>ad hoc</em> methods, as illustrated later.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>The identification data can also be passed as dedicated
identification objects such as <code>mzRident</code> from the <em><a href="https://bioconductor.org/packages/3.18/mzR" class="external-link">mzR</a></em> package
or <code>mzID</code> from thr <em><a href="https://bioconductor.org/packages/3.18/mzID" class="external-link">mzID</a></em>
package, or as a <code>data.frame</code> - see
<code>?addIdentifionData</code> for details.<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p>The code to generate the histograms has been contributed
by Guangchuang Yu.<a href="#fnref6" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
