<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>MSnbase IO capabilities • MSnbase</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MSnbase IO capabilities">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MSnbase</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.31.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a></li>
    <li><a class="dropdown-item" href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a></li>
    <li><a class="dropdown-item" href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a></li>
    <li><a class="dropdown-item" href="../articles/v04-benchmarking.html">MSnbase benchmarking</a></li>
    <li><a class="dropdown-item" href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/lgatto/MSnbase/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>MSnbase IO capabilities</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
            <address class="author_afil">
      de Duve Institute, UCLouvain, Belgium<br><small class="dont-index">Source: <a href="https://github.com/lgatto/MSnbase/blob/master/vignettes/v02-MSnbase-io.Rmd" class="external-link"><code>vignettes/v02-MSnbase-io.Rmd</code></a></small>
      <div class="d-none name"><code>v02-MSnbase-io.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette describes <em>MSnbase</em>’s input and output
      capabilities.</p>
    </div>
    
<div class="section level2">
<h2 class="unnumbered" id="foreword">Foreword<a class="anchor" aria-label="anchor" href="#foreword"></a>
</h2>
<p>This software is free and open-source software. If you use it, please
support the project by citing it in publications:</p>
<blockquote>
<p>Gatto L, Lilley KS. MSnbase-an R/Bioconductor package for isobaric
tagged mass spectrometry data visualization, processing and
quantitation. Bioinformatics. 2012 Jan 15;28(2):288-9. doi: <a href="https://doi.org/10.1093/bioinformatics/btr645" class="external-link">10.1093/bioinformatics/btr645</a>.
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22113085" class="external-link">22113085</a>.</p>
</blockquote>
<blockquote>
<p><em><code>MSnbase</code>, efficient and elegant R-based processing
and visualisation of raw mass spectrometry data</em>. Laurent Gatto,
Sebastian Gibb, Johannes Rainer. bioRxiv 2020.04.29.067868; doi: <a href="https://doi.org/10.1101/2020.04.29.067868" class="external-link uri">https://doi.org/10.1101/2020.04.29.067868</a></p>
</blockquote>
</div>
<div class="section level2">
<h2 class="unnumbered" id="questions-and-bugs">Questions and bugs<a class="anchor" aria-label="anchor" href="#questions-and-bugs"></a>
</h2>
<p>For bugs, typos, suggestions or other questions, please file an issue
in our tracking system (<a href="https://github.com/lgatto/MSnbase/issues" class="external-link uri">https://github.com/lgatto/MSnbase/issues</a>) providing as
much information as possible, a reproducible example and the output of
<code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code>.</p>
<p>If you don’t have a GitHub account or wish to reach a broader
audience for general questions about proteomics analysis using R, you
may want to use the Bioconductor support site: <a href="https://support.bioconductor.org/" class="external-link uri">https://support.bioconductor.org/</a>.</p>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.19/MSnbase" class="external-link">MSnbase</a></em>’s
aims are to facilitate the reproducible analysis of mass spectrometry
data within the R environment, from raw data import and processing,
feature quantification, quantification and statistical analysis of the
results <span class="citation">(Gatto and Lilley 2012)</span>. Data
import functions for several formats are provided and intermediate or
final results can also be saved or exported. These capabilities are
presented below.</p>
</div>
<div class="section level2">
<h2 id="data-input">Data input<a class="anchor" aria-label="anchor" href="#data-input"></a>
</h2>
<div class="section level5">
<h5 class="unnumbered" id="raw-data">Raw data<a class="anchor" aria-label="anchor" href="#raw-data"></a>
</h5>
<p>Data stored in one of the published <code>XML</code>-based formats.
i.e. <code>mzXML</code> <span class="citation">(Pedrioli et al.
2004)</span>, <code>mzData</code> <span class="citation">(Orchard et al.
2007)</span> or <code>mzML</code> <span class="citation">(Martens et al.
2010)</span>, can be imported with the <code>readMSData</code> method,
which makes use of the <em><a href="https://bioconductor.org/packages/3.19/mzR" class="external-link">mzR</a></em> package
to create <code>MSnExp</code> objects. The files can be in profile or
centroided mode. See <code><a href="../reference/readMSData.html">?readMSData</a></code> for details.</p>
<p>Data from <code>mzML</code> files containing chromatographic data
(e.g. generated in SRM/MRM experiments) can be imported with the
<code>readSRMData</code> function that returns the chromatographic data
as a <code>MChromatograms</code> object. See <code><a href="../reference/readSRMData.html">?readSRMData</a></code>
for more details.</p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="peak-lists">Peak lists<a class="anchor" aria-label="anchor" href="#peak-lists"></a>
</h5>
<p>Peak lists in the <code>mgf</code> format<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="http://www.matrixscience.com/help/data_file_help.html" class="external-link uri"&gt;http://www.matrixscience.com/help/data_file_help.html&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> can be imported using
the <code>readMgfData</code>. In this case, the peak data has generally
been pre-processed by other software. See <code><a href="../reference/readMgfData.html">?readMgfData</a></code> for
details.</p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="quantitation-data">Quantitation data<a class="anchor" aria-label="anchor" href="#quantitation-data"></a>
</h5>
<p>Third party software can be used to generate quantitative data and
exported as a spreadsheet (generally comma or tab separated format).
This data as well as any additional meta-data can be imported with the
<code>readMSnSet</code> function. See <code><a href="../reference/readMSnSet.html">?readMSnSet</a></code> for
details.</p>
<p><em><a href="https://bioconductor.org/packages/3.19/MSnbase" class="external-link">MSnbase</a></em>
also supports the <code>mzTab</code> format<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://github.com/HUPO-PSI/mzTab" class="external-link uri"&gt;https://github.com/HUPO-PSI/mzTab&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>, a light-weight,
tab-delimited file format for proteomics data developed within the
Proteomics Standards Initiative (PSI). <code>mzTab</code> files can be
read into R with <code>readMzTabData</code> to create and
<code>MSnSet</code> instance.</p>
<div class="float">
<img src="Figures%2FMSnbase-io-in.png" alt="MSnbase input capabilities. The white and red boxes represent R functions/methods and objects respectively. The blue boxes represent different disk storage formats."><div class="figcaption">
<em>MSnbase</em> input capabilities. The white
and red boxes represent R functions/methods and objects respectively.
The blue boxes represent different disk storage formats.</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="data-output">Data output<a class="anchor" aria-label="anchor" href="#data-output"></a>
</h2>
<div class="section level5">
<h5 class="unnumbered" id="rdata-files">RData files<a class="anchor" aria-label="anchor" href="#rdata-files"></a>
</h5>
<p>R objects can most easily be stored on disk with the
<code>save</code> function. It creates compressed binary images of the
data representation that can later be read back from the file with the
<code>load</code> function.</p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="mzmlmzxml-files">mzML/mzXML files<a class="anchor" aria-label="anchor" href="#mzmlmzxml-files"></a>
</h5>
<p><code>MSnExp</code> and <code>OnDiskMSnExp</code> files can be
written to MS data files in <code>mzML</code> or <code>mzXML</code>
files with the <code>writeMSData</code> method. See
<code><a href="../reference/writeMSData.html">?writeMSData</a></code> for details.</p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="peak-lists-1">Peak lists<a class="anchor" aria-label="anchor" href="#peak-lists-1"></a>
</h5>
<p><code>MSnExp</code> instances as well as individual spectra can be
written as <code>mgf</code> files with the <code>writeMgfData</code>
method. Note that the meta-data in the original R object can not be
included in the file. See <code><a href="../reference/writeMgfData-methods.html">?writeMgfData</a></code> for details.</p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="quantitation-data-1">Quantitation data<a class="anchor" aria-label="anchor" href="#quantitation-data-1"></a>
</h5>
<p>Quantitation data can be exported to spreadsheet files with the
<code>write.exprs</code> method. Feature meta-data can be appended to
the feature intensity values. See <code><a href="../reference/writeMgfData-methods.html">?writeMgfData</a></code> for
details.</p>
<p><strong>Deprecated</strong> <code>MSnSet</code> instances can also be
exported to <code>mzTab</code> files using the
<code>writeMzTabData</code> function.</p>
<div class="float">
<img src="Figures%2FMSnbase-io-out.png" alt="MSnbase output capabilities. The white and red boxes represent R functions/methods and objects respectively. The blue boxes represent different disk storage formats."><div class="figcaption">
<em>MSnbase</em> output capabilities. The white
and red boxes represent R functions/methods and objects respectively.
The blue boxes represent different disk storage formats.</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="creating-msnset-from-text-spread-sheets">Creating <code>MSnSet</code> from text spread sheets<a class="anchor" aria-label="anchor" href="#creating-msnset-from-text-spread-sheets"></a>
</h2>
<p>This section describes the generation of <code>MSnSet</code> objects
using data available in a text-based spreadsheet. This entry point into
R and <em><a href="https://bioconductor.org/packages/3.19/MSnbase" class="external-link">MSnbase</a></em>
allows to import data processed by any of the third party
mass-spectrometry processing software available and proceed with data
exploration, normalisation and statistical analysis using functions
available in and the numerous Bioconductor packages.</p>
<div class="section level3">
<h3 id="a-complete-work-flow">A complete work flow<a class="anchor" aria-label="anchor" href="#a-complete-work-flow"></a>
</h3>
<p>The following section describes a work flow that uses three input
files to create the <code>MSnSet</code>. These files respectively
describe the quantitative expression data, the sample meta-data and the
feature meta-data. It is taken from the <em><a href="https://bioconductor.org/packages/3.19/pRoloc" class="external-link">pRoloc</a></em>
tutorial and uses example files from the <em><a href="https://bioconductor.org/packages/3.19/pRolocdat" class="external-link">pRolocdat</a></em>
package.</p>
<p>We start by describing the <code>csv</code> to be used as input using
the <code>read.csv</code> function.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The original data for replicate 1, available</span></span>
<span><span class="co">## from the pRolocdata package</span></span>
<span><span class="va">f0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,</span>
<span>          full.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          pattern <span class="op">=</span> <span class="st">"pr800866n_si_004-rep1.csv"</span><span class="op">)</span></span>
<span><span class="va">csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f0</span><span class="op">)</span></span></code></pre></div>
<p>The three first lines of the original spreadsheet, containing the
data for replicate one, are illustrated below (using the function
<code>head</code>). It contains 888 rows (proteins) and 16 columns,
including protein identifiers, database accession numbers, gene symbols,
reporter ion quantitation values, information related to protein
identification, …</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">csv</span>, n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Protein.ID        FBgn Flybase.Symbol No..peptide.IDs Mascot.score</span></span>
<span><span class="co">## 1    CG10060 FBgn0001104    G-ialpha65A               3       179.86</span></span>
<span><span class="co">## 2    CG10067 FBgn0000044         Act57B               5       222.40</span></span>
<span><span class="co">## 3    CG10077 FBgn0035720        CG10077               5       219.65</span></span>
<span><span class="co">##   No..peptides.quantified area.114 area.115 area.116 area.117</span></span>
<span><span class="co">## 1                       1 0.379000 0.281000 0.225000 0.114000</span></span>
<span><span class="co">## 2                       9 0.420000 0.209667 0.206111 0.163889</span></span>
<span><span class="co">## 3                       3 0.187333 0.167333 0.169667 0.476000</span></span>
<span><span class="co">##   PLS.DA.classification Peptide.sequence Precursor.ion.mass</span></span>
<span><span class="co">## 1                    PM                                    </span></span>
<span><span class="co">## 2                    PM                                    </span></span>
<span><span class="co">## 3                                                          </span></span>
<span><span class="co">##   Precursor.ion.charge pd.2013 pd.markers</span></span>
<span><span class="co">## 1                           PM    unknown</span></span>
<span><span class="co">## 2                           PM    unknown</span></span>
<span><span class="co">## 3                      unknown    unknown</span></span></code></pre>
<p>Below read in turn the spread sheets that contain the quantitation
data (<code>exprsFile.csv</code>), feature meta-data
(<code>fdataFile.csv</code>) and sample meta-data
(<code>pdataFile.csv</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The quantitation data, from the original data</span></span>
<span><span class="va">f1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,</span>
<span>          full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"exprsFile.csv"</span><span class="op">)</span></span>
<span><span class="va">exprsCsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f1</span><span class="op">)</span></span>
<span><span class="co">## Feature meta-data, from the original data</span></span>
<span><span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,</span>
<span>          full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"fdataFile.csv"</span><span class="op">)</span></span>
<span><span class="va">fdataCsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f2</span><span class="op">)</span></span>
<span><span class="co">## Sample meta-data, a new file</span></span>
<span><span class="va">f3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,</span>
<span>          full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"pdataFile.csv"</span><span class="op">)</span></span>
<span><span class="va">pdataCsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f3</span><span class="op">)</span></span></code></pre></div>
<p><code>exprsFile.csv</code> contains the quantitation (expression)
data for the 888 proteins and 4 reporter tags.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exprsCsv</span>, n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          FBgn     X114     X115     X116     X117</span></span>
<span><span class="co">## 1 FBgn0001104 0.379000 0.281000 0.225000 0.114000</span></span>
<span><span class="co">## 2 FBgn0000044 0.420000 0.209667 0.206111 0.163889</span></span>
<span><span class="co">## 3 FBgn0035720 0.187333 0.167333 0.169667 0.476000</span></span></code></pre>
<p><code>fdataFile.csv</code> contains meta-data for the 888 features
(here proteins).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fdataCsv</span>, n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          FBgn ProteinID FlybaseSymbol NoPeptideIDs MascotScore</span></span>
<span><span class="co">## 1 FBgn0001104   CG10060   G-ialpha65A            3      179.86</span></span>
<span><span class="co">## 2 FBgn0000044   CG10067        Act57B            5      222.40</span></span>
<span><span class="co">## 3 FBgn0035720   CG10077       CG10077            5      219.65</span></span>
<span><span class="co">##   NoPeptidesQuantified PLSDA</span></span>
<span><span class="co">## 1                    1    PM</span></span>
<span><span class="co">## 2                    9    PM</span></span>
<span><span class="co">## 3                    3</span></span></code></pre>
<p><code>pdataFile.csv</code> contains samples (here fractions)
meta-data. This simple file has been created manually.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pdataCsv</span></span></code></pre></div>
<pre><code><span><span class="co">##   sampleNames Fractions</span></span>
<span><span class="co">## 1        X114       4/5</span></span>
<span><span class="co">## 2        X115     12/13</span></span>
<span><span class="co">## 3        X116        19</span></span>
<span><span class="co">## 4        X117        21</span></span></code></pre>
<p>The self-contained <code>MSnSet</code> can now easily be generated
using the <code>readMSnSet</code> constructor, providing the respective
<code>csv</code> file names shown above and specifying that the data is
comma-separated (with <code>sep = ","</code>). Below, we call that
object <code>res</code> and display its content.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://lgatto.github.io/MSnbase">"MSnbase"</a></span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMSnSet.html">readMSnSet</a></span><span class="op">(</span>exprsFile <span class="op">=</span> <span class="va">f1</span>,</span>
<span>                  featureDataFile <span class="op">=</span> <span class="va">f2</span>,</span>
<span>                  phenoDataFile <span class="op">=</span> <span class="va">f3</span>,</span>
<span>                  sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## MSnSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 888 features, 4 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData</span></span>
<span><span class="co">##   sampleNames: X114 X115 X116 X117</span></span>
<span><span class="co">##   varLabels: Fractions</span></span>
<span><span class="co">##   varMetadata: labelDescription</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: FBgn0001104 FBgn0000044 ... FBgn0001215 (888 total)</span></span>
<span><span class="co">##   fvarLabels: ProteinID FlybaseSymbol ... PLSDA (6 total)</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:  </span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">## Quantitation data loaded: Tue Oct 15 15:28:20 2024  using readMSnSet. </span></span>
<span><span class="co">##  MSnbase version: 2.31.2</span></span></code></pre>
<div class="section level4">
<h4 id="the-msnset-class">The <code>MSnSet</code> class<a class="anchor" aria-label="anchor" href="#the-msnset-class"></a>
</h4>
<p>Although there are additional specific sub-containers for additional
meta-data (for instance to make the object MIAPE compliant), the feature
(the sub-container, or slot <code>featureData</code>) and sample (the
<code>phenoData</code> slot) are the most important ones. They need to
meet the following validity requirements (see figure below):</p>
<ul>
<li><p>the number of row in the expression/quantitation data and feature
data must be equal and the row names must match exactly, and</p></li>
<li><p>the number of columns in the expression/quantitation data and
number of row in the sample meta-data must be equal and the column/row
names must match exactly.</p></li>
</ul>
<p>A detailed description of the <code>MSnSet</code> class is available
by typing <code><a href="../reference/MSnSet-class.html">?MSnSet</a></code> in the R console.</p>
<div class="float">
<img src="Figures%2Fmsnset.png" alt="Dimension requirements for the respective expression, feature and sample meta-data slots."><div class="figcaption">Dimension requirements for the respective
expression, feature and sample meta-data slots.</div>
</div>
<p>The individual parts of this data object can be accessed with their
respective accessor methods:</p>
<ul>
<li>the quantitation data can be retrieved with
<code>exprs(res)</code>,</li>
<li>the feature meta-data with <code>fData(res)</code> and</li>
<li>the sample meta-data with <code>pData(res)</code>.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="a-shorter-work-flow">A shorter work flow<a class="anchor" aria-label="anchor" href="#a-shorter-work-flow"></a>
</h3>
<p>The <code>readMSnSet2</code> function provides a simplified import
workforce. It takes a single spreadsheet as input (default is
<code>csv</code>) and extract the columns identified by
<code>ecol</code> to create the expression data, while the others are
used as feature meta-data. <code>ecol</code> can be a
<code>character</code> with the respective column labels or a numeric
with their indices. In the former case, it is important to make sure
that the names match exactly. Special characters like <code>'-'</code>
or <code>'('</code> will be transformed by R into <code>'.'</code> when
the <code>csv</code> file is read in. Optionally, one can also specify a
column to be used as feature names. Note that these must be unique to
guarantee the final object validity.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ecol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="fl">114</span><span class="op">:</span><span class="fl">117</span>, sep <span class="op">=</span> <span class="st">"."</span><span class="op">)</span></span>
<span><span class="va">fname</span> <span class="op">&lt;-</span> <span class="st">"Protein.ID"</span></span>
<span><span class="va">eset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readMSnSet.html">readMSnSet2</a></span><span class="op">(</span><span class="va">f0</span>, <span class="va">ecol</span>, <span class="va">fname</span><span class="op">)</span></span>
<span><span class="va">eset</span></span></code></pre></div>
<pre><code><span><span class="co">## MSnSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 888 features, 4 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData: none</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: CG10060 CG10067 ... CG9983 (888 total)</span></span>
<span><span class="co">##   fvarLabels: Protein.ID FBgn ... pd.markers (12 total)</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:  </span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">##  MSnbase version: 2.31.2</span></span></code></pre>
<p>The <code>ecol</code> columns can also be queried interactively from
R using the <code>getEcols</code> and <code>grepEcols</code> function.
The former return a character with all column names, given a splitting
character, i.e. the separation value of the spreadsheet (typically
<code>","</code> for <code>csv</code>, <code>"\t"</code> for
<code>tsv</code>, …). The latter can be used to grep a pattern of
interest to obtain the relevant column indices.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/grepEcols.html">getEcols</a></span><span class="op">(</span><span class="va">f0</span>, <span class="st">","</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "\"Protein ID\""              "\"FBgn\""                   </span></span>
<span><span class="co">##  [3] "\"Flybase Symbol\""          "\"No. peptide IDs\""        </span></span>
<span><span class="co">##  [5] "\"Mascot score\""            "\"No. peptides quantified\""</span></span>
<span><span class="co">##  [7] "\"area 114\""                "\"area 115\""               </span></span>
<span><span class="co">##  [9] "\"area 116\""                "\"area 117\""               </span></span>
<span><span class="co">## [11] "\"PLS-DA classification\""   "\"Peptide sequence\""       </span></span>
<span><span class="co">## [13] "\"Precursor ion mass\""      "\"Precursor ion charge\""   </span></span>
<span><span class="co">## [15] "\"pd.2013\""                 "\"pd.markers\""</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/grepEcols.html">grepEcols</a></span><span class="op">(</span><span class="va">f0</span>, <span class="st">"area"</span>, <span class="st">","</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  7  8  9 10</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/grepEcols.html">grepEcols</a></span><span class="op">(</span><span class="va">f0</span>, <span class="st">"area"</span>, <span class="st">","</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/readMSnSet.html">readMSnSet2</a></span><span class="op">(</span><span class="va">f0</span>, <span class="va">e</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSnSet (storageMode: lockedEnvironment)</span></span>
<span><span class="co">## assayData: 888 features, 4 samples </span></span>
<span><span class="co">##   element names: exprs </span></span>
<span><span class="co">## protocolData: none</span></span>
<span><span class="co">## phenoData: none</span></span>
<span><span class="co">## featureData</span></span>
<span><span class="co">##   featureNames: 1 2 ... 888 (888 total)</span></span>
<span><span class="co">##   fvarLabels: Protein.ID FBgn ... pd.markers (12 total)</span></span>
<span><span class="co">##   fvarMetadata: labelDescription</span></span>
<span><span class="co">## experimentData: use 'experimentData(object)'</span></span>
<span><span class="co">## Annotation:  </span></span>
<span><span class="co">## - - - Processing information - - -</span></span>
<span><span class="co">##  MSnbase version: 2.31.2</span></span></code></pre>
<p>The <code>phenoData</code> slot can now be updated accordingly using
the replacement functions <code>phenoData&lt;-</code> or
<code>pData&lt;-</code> (see <code><a href="../reference/MSnSet-class.html">?MSnSet</a></code> for details).</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] pRolocdata_1.43.3   MSnbase_2.31.2      ProtGenerics_1.36.0</span></span>
<span><span class="co">## [4] S4Vectors_0.42.1    mzR_2.38.0          Rcpp_1.0.13        </span></span>
<span><span class="co">## [7] Biobase_2.64.0      BiocGenerics_0.50.0 BiocStyle_2.32.1   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] rlang_1.1.4                 magrittr_2.0.3             </span></span>
<span><span class="co">##  [3] clue_0.3-65                 matrixStats_1.4.1          </span></span>
<span><span class="co">##  [5] compiler_4.4.1              systemfonts_1.1.0          </span></span>
<span><span class="co">##  [7] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">##  [9] stringr_1.5.1               pkgconfig_2.0.3            </span></span>
<span><span class="co">## [11] crayon_1.5.3                fastmap_1.2.0              </span></span>
<span><span class="co">## [13] XVector_0.44.0              utf8_1.2.4                 </span></span>
<span><span class="co">## [15] rmarkdown_2.28              UCSC.utils_1.0.0           </span></span>
<span><span class="co">## [17] preprocessCore_1.67.1       ragg_1.3.3                 </span></span>
<span><span class="co">## [19] purrr_1.0.2                 xfun_0.48                  </span></span>
<span><span class="co">## [21] MultiAssayExperiment_1.30.3 zlibbioc_1.50.0            </span></span>
<span><span class="co">## [23] cachem_1.1.0                GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">## [25] jsonlite_1.8.9              DelayedArray_0.30.1        </span></span>
<span><span class="co">## [27] BiocParallel_1.38.0         parallel_4.4.1             </span></span>
<span><span class="co">## [29] cluster_2.1.6               R6_2.5.1                   </span></span>
<span><span class="co">## [31] bslib_0.8.0                 stringi_1.8.4              </span></span>
<span><span class="co">## [33] limma_3.60.6                GenomicRanges_1.56.2       </span></span>
<span><span class="co">## [35] jquerylib_0.1.4             iterators_1.0.14           </span></span>
<span><span class="co">## [37] bookdown_0.40               SummarizedExperiment_1.34.0</span></span>
<span><span class="co">## [39] knitr_1.48                  IRanges_2.38.1             </span></span>
<span><span class="co">## [41] Matrix_1.7-0                igraph_2.0.3               </span></span>
<span><span class="co">## [43] tidyselect_1.2.1            abind_1.4-8                </span></span>
<span><span class="co">## [45] yaml_2.3.10                 doParallel_1.0.17          </span></span>
<span><span class="co">## [47] codetools_0.2-20            affy_1.82.0                </span></span>
<span><span class="co">## [49] lattice_0.22-6              tibble_3.2.1               </span></span>
<span><span class="co">## [51] plyr_1.8.9                  evaluate_1.0.1             </span></span>
<span><span class="co">## [53] desc_1.4.3                  pillar_1.9.0               </span></span>
<span><span class="co">## [55] affyio_1.74.0               BiocManager_1.30.25        </span></span>
<span><span class="co">## [57] MatrixGenerics_1.16.0       foreach_1.5.2              </span></span>
<span><span class="co">## [59] MALDIquant_1.22.3           ncdf4_1.23                 </span></span>
<span><span class="co">## [61] generics_0.1.3              ggplot2_3.5.1              </span></span>
<span><span class="co">## [63] munsell_0.5.1               scales_1.3.0               </span></span>
<span><span class="co">## [65] glue_1.8.0                  lazyeval_0.2.2             </span></span>
<span><span class="co">## [67] tools_4.4.1                 mzID_1.42.0                </span></span>
<span><span class="co">## [69] QFeatures_1.14.2            vsn_3.72.0                 </span></span>
<span><span class="co">## [71] fs_1.6.4                    XML_3.99-0.17              </span></span>
<span><span class="co">## [73] grid_4.4.1                  impute_1.78.0              </span></span>
<span><span class="co">## [75] tidyr_1.3.1                 MsCoreUtils_1.16.1         </span></span>
<span><span class="co">## [77] colorspace_2.1-1            GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">## [79] PSMatch_1.8.0               cli_3.6.3                  </span></span>
<span><span class="co">## [81] textshaping_0.4.0           fansi_1.0.6                </span></span>
<span><span class="co">## [83] S4Arrays_1.4.1              dplyr_1.1.4                </span></span>
<span><span class="co">## [85] AnnotationFilter_1.28.0     pcaMethods_1.96.0          </span></span>
<span><span class="co">## [87] gtable_0.3.5                sass_0.4.9                 </span></span>
<span><span class="co">## [89] digest_0.6.37               SparseArray_1.4.8          </span></span>
<span><span class="co">## [91] htmlwidgets_1.6.4           htmltools_0.5.8.1          </span></span>
<span><span class="co">## [93] pkgdown_2.1.1.9000          lifecycle_1.0.4            </span></span>
<span><span class="co">## [95] httr_1.4.7                  statmod_1.5.0              </span></span>
<span><span class="co">## [97] MASS_7.3-61</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Gatto2012" class="csl-entry">
Gatto, Laurent, and Kathryn S Lilley. 2012. <span>“<span>MS</span>nbase
– an <span>R</span>/<span>B</span>ioconductor Package for Isobaric
Tagged Mass Spectrometry Data Visualization, Processing and
Quantitation.”</span> <em>Bioinformatics</em> 28 (2): 288–89. <a href="https://doi.org/10.1093/bioinformatics/btr645" class="external-link">https://doi.org/10.1093/bioinformatics/btr645</a>.
</div>
<div id="ref-Martens2010" class="csl-entry">
Martens, Lennart, Matthew Chambers, Marc Sturm, Darren Kes sner, Fredrik
Levander, Jim Shofstahl, Wilfred H Tang, et al. 2010. <span>“mzML - a
Community Standard for Mass Spectrometry Data.”</span> <em>Molecular
&amp; Cellular Proteomics : MCP</em>. <a href="https://doi.org/10.1074/mcp.R110.000133" class="external-link">https://doi.org/10.1074/mcp.R110.000133</a>.
</div>
<div id="ref-Orchard2007" class="csl-entry">
Orchard, Sandra, Luisa Montechi-Palazzi, Eric W Deutsch, Pierre-Alain
Binz, Andrew R Jones, Norman Paton, Angel Pizarro, David M Creasy,
Jérôme Wojcik, and Henning Hermjakob. 2007. <span>“Five Years of
Progress in the Standardization of Proteomics Data 4th Annual Spring
Workshop of the HUPO-Proteomics Standards Initiative April 23-25, 2007
Ecole Nationale Supérieure (ENS), Lyon, France.”</span>
<em>Proteomics</em> 7 (19): 3436–40. <a href="https://doi.org/10.1002/pmic.200700658" class="external-link">https://doi.org/10.1002/pmic.200700658</a>.
</div>
<div id="ref-Pedrioli2004" class="csl-entry">
Pedrioli, Patrick G A, Jimmy K Eng, Robert Hubley, Mathijs Vogelzang,
Eric W Deutsch, Brian Raught, Brian Pratt, et al. 2004. <span>“A Common
Open Representation of Mass Spectrometry Data and Its Application to
Proteomics Research.”</span> <em>Nat. Biotechnol.</em> 22 (11): 1459–66.
<a href="https://doi.org/10.1038/nbt1031" class="external-link">https://doi.org/10.1038/nbt1031</a>.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
