<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A short introduction to *MSnbase* development • MSnbase</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A short introduction to *MSnbase* development">
<meta property="og:description" content="MSnbase">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.29.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/v04-benchmarking.html">MSnbase benchmarking</a>
    </li>
    <li>
      <a href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/MSnbase/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A short introduction to <em>MSnbase</em>
development</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
            <address class="author_afil">
      de Duve Institute, UCLouvain,
Belgium<br><h4 data-toc-skip class="author">Johannes
Rainer</h4>
            <address class="author_afil">
      Institute for Biomedicine, Eurac Research, Bolzano,
Italy.<br><h4 data-toc-skip class="author">Sebastian
Gibb</h4>
            <address class="author_afil">
      Department of Anesthesiology and Intensive Care, University
Medicine Greifswald, Germany.<br><small class="dont-index">Source: <a href="https://github.com/lgatto/MSnbase/blob/HEAD/vignettes/v05-MSnbase-development.Rmd" class="external-link"><code>vignettes/v05-MSnbase-development.Rmd</code></a></small>
      <div class="hidden name"><code>v05-MSnbase-development.Rmd</code></div>

    </address>
</address>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette describes the classes implemented in package. It
      is intended as a starting point for developers or users who would
      like to learn more or further develop/extend mass spectrometry and
      proteomics data structures.</p>
    </div>
    
<div class="section level2">
<h2 class="unnumbered" id="foreword">Foreword<a class="anchor" aria-label="anchor" href="#foreword"></a>
</h2>
<p>This software is free and open-source software. If you use it, please
support the project by citing it in publications:</p>
<blockquote>
<p>Gatto L, Lilley KS. MSnbase-an R/Bioconductor package for isobaric
tagged mass spectrometry data visualization, processing and
quantitation. Bioinformatics. 2012 Jan 15;28(2):288-9. doi: <a href="https://doi.org/10.1093/bioinformatics/btr645" class="external-link">10.1093/bioinformatics/btr645</a>.
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22113085" class="external-link">22113085</a>.</p>
</blockquote>
<blockquote>
<p><em><code>MSnbase</code>, efficient and elegant R-based processing
and visualisation of raw mass spectrometry data</em>. Laurent Gatto,
Sebastian Gibb, Johannes Rainer. bioRxiv 2020.04.29.067868; doi: <a href="https://doi.org/10.1101/2020.04.29.067868" class="external-link uri">https://doi.org/10.1101/2020.04.29.067868</a></p>
</blockquote>
</div>
<div class="section level2">
<h2 class="unnumbered" id="questions-and-bugs">Questions and bugs<a class="anchor" aria-label="anchor" href="#questions-and-bugs"></a>
</h2>
<p>For bugs, typos, suggestions or other questions, please file an issue
in our tracking system (<a href="https://github.com/lgatto/MSnbase/issues" class="external-link uri">https://github.com/lgatto/MSnbase/issues</a>) providing as
much information as possible, a reproducible example and the output of
<code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code>.</p>
<p>If you don’t have a GitHub account or wish to reach a broader
audience for general questions about proteomics analysis using R, you
may want to use the Bioconductor support site: <a href="https://support.bioconductor.org/" class="external-link uri">https://support.bioconductor.org/</a>.</p>
<p><strong>NB</strong> This document is going to be updated based on
current major development plans in <code>MSnbase</code>.</p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document is not a replacement for the individual manual pages,
that document the slots of the <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
classes. It is a centralised high-level description of the package
design.</p>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
aims at being compatible with the <em><a href="https://bioconductor.org/packages/3.18/Biobase" class="external-link">Biobase</a></em>
infrastructure <span class="citation">(Gentleman et al. 2004)</span>.
Many meta data structures that are used in <code>eSet</code> and
associated classes are also used here. As such, knowledge of the
<em>Biobase development and the new eSet</em> vignette would be
beneficial; the vignette can directly be accessed with
<code><a href="https://bioconductor.org/packages/release/bioc/vignettes/Biobase/inst/doc/BiobaseDevelopment.html" class="external-link">vignette("BiobaseDevelopment", package="Biobase")</a></code>.</p>
<p>The initial goal is to use the <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
infrastructure for MS2 labelled (iTRAQ <span class="citation">(Ross et
al. 2004)</span> and TMT <span class="citation">(Thompson et al.
2003)</span>) and label-free (spectral counting, index and abundance)
quantitation - see the documentation for the <code>quantify</code>
function for details. The infrastructure is currently extended to
support a wider range of technologies, including metabolomics.</p>
</div>
<div class="section level2">
<h2 id="coding-style">Coding style<a class="anchor" aria-label="anchor" href="#coding-style"></a>
</h2>
<p><code>MSnbase</code> follows the <a href="https://bioconductor.org/developers/how-to/coding-style/" class="external-link">Bioconductor
style guide</a>. In particular</p>
<ul>
<li>Do not use <code>.</code> when naming symbols.</li>
<li>A leading <code>.</code> can be used for hidden/local functions or
variables.</li>
<li>Snake case should be restricted to internal functions. For
consistency, we favour camel case for public functions.</li>
<li>Class names should start with a capital and each class should posses
a constructor with identical name. Running the constructor without any
input should produce a valid empty object.</li>
<li>Use <code>##</code> to start full-line comments.</li>
<li>For roxygen headers <code>#'</code> is preferred, although
<code>##'</code> is tolerated.</li>
<li>Use spaces between <code>=</code> in function arguments or class
definition: <code>f(a = 1, b = 2)</code>.</li>
<li>Always use a space after a comma: <code>a, b, c</code>.</li>
<li>Always use spaces around binary operators: <code>a + b</code>.</li>
<li>Lines should be kept shorter than 80 characters. For example the
following code isn’t accepted</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># no wrap at 80</span></span>
<span><span class="va">someVeryLongVariableName</span> <span class="op">&lt;-</span> <span class="fu">someVeryLongFunctionName</span><span class="op">(</span>withSomeEvenLongerFunctionArgumentA <span class="op">=</span> <span class="fl">1</span>, withSomeEvenLongerFunctionArgumentB <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>and should be wrapped as shown below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># alternative 1</span></span>
<span><span class="va">someVeryLongVariableName</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">someVeryLongFunctionName</span><span class="op">(</span>withSomeEvenLongerFunctionArgumentA <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                             withSomeEvenLongerFunctionArgumentB <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># alternative 2</span></span>
<span><span class="va">someVeryLongVariableName</span> <span class="op">&lt;-</span> <span class="fu">someVeryLongFunctionName</span><span class="op">(</span></span>
<span>    withSomeEvenLongerFunctionArgumentA <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    withSomeEvenLongerFunctionArgumentB <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="msnbase-classes">
<em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
classes<a class="anchor" aria-label="anchor" href="#msnbase-classes"></a>
</h2>
<p>All classes have a <code>.__classVersion__</code> slot, of class
<code>Versioned</code> from the <em><a href="https://bioconductor.org/packages/3.18/Biobase" class="external-link">Biobase</a></em>
package. This slot documents the class version for any instance to be
used for debugging and object update purposes. Any change in a class
implementation should trigger a version change.</p>
<div class="section level3">
<h3 id="pset-a-virtual-class-for-raw-mass-spectrometry-data-and-meta-data">
<code>pSet</code>: a virtual class for raw mass spectrometry data
and meta data<a class="anchor" aria-label="anchor" href="#pset-a-virtual-class-for-raw-mass-spectrometry-data-and-meta-data"></a>
</h3>
<p>This virtual class is the main container for mass spectrometry data,
i.e spectra, and meta data. It is based on the <code>eSet</code>
implementation for genomic data. The main difference with
<code>eSet</code> is that the <code>assayData</code> slot is an
environment containing any number of <code>Spectrum</code> instances
(see the <a href="#Spectrum"><code>Spectrum</code> section</a>).</p>
<p>One new slot is introduced, namely <code>processingData</code>, that
contains one <code>MSnProcess</code> instance (see the <a href="#MSnProcess"><code>MSnProcess</code> section</a>). and the
<code>experimentData</code> slot is now expected to contain
<code>MIAPE</code> data. The <code>annotation</code> slot has not been
implemented, as no prior feature annotation is known in shotgun
proteomics.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"pSet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Virtual Class "pSet" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:           assayData          phenoData        featureData</span></span>
<span><span class="co">## Class:        environment AnnotatedDataFrame AnnotatedDataFrame</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:      experimentData       protocolData     processingData</span></span>
<span><span class="co">## Class:              MIAxE AnnotatedDataFrame         MSnProcess</span></span>
<span><span class="co">##                                             </span></span>
<span><span class="co">## Name:              .cache  .__classVersion__</span></span>
<span><span class="co">## Class:        environment           Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: "Versioned"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Known Subclasses: </span></span>
<span><span class="co">## Class "MSnExp", directly</span></span>
<span><span class="co">## Class "OnDiskMSnExp", by class "MSnExp", distance 2, with explicit coerce</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="msnexp-a-class-for-ms-experiments">
<code>MSnExp</code>: a class for MS experiments<a class="anchor" aria-label="anchor" href="#msnexp-a-class-for-ms-experiments"></a>
</h3>
<p><code>MSnExp</code> extends <code>pSet</code> to store MS
experiments. It does not add any new slots to <code>pSet</code>.
Accessors and setters are all inherited from <code>pSet</code> and new
ones should be implemented for <code>pSet</code>. Methods that
manipulate actual data in experiments are implemented for
<code>MSnExp</code> objects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"MSnExp"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "MSnExp" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:           assayData          phenoData        featureData</span></span>
<span><span class="co">## Class:        environment AnnotatedDataFrame AnnotatedDataFrame</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:      experimentData       protocolData     processingData</span></span>
<span><span class="co">## Class:              MIAxE AnnotatedDataFrame         MSnProcess</span></span>
<span><span class="co">##                                             </span></span>
<span><span class="co">## Name:              .cache  .__classVersion__</span></span>
<span><span class="co">## Class:        environment           Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: </span></span>
<span><span class="co">## Class "pSet", directly</span></span>
<span><span class="co">## Class "Versioned", by class "pSet", distance 2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Known Subclasses: </span></span>
<span><span class="co">## Class "OnDiskMSnExp", directly, with explicit coerce</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="ondiskmsnexp-a-on-disk-implementation-of-the-msnexp-class">
<code>OnDiskMSnExp</code>: a on-disk implementation of the
<code>MSnExp</code> class<a class="anchor" aria-label="anchor" href="#ondiskmsnexp-a-on-disk-implementation-of-the-msnexp-class"></a>
</h3>
<p>The <code>OnDiskMSnExp</code> class extends <code>MSnExp</code> and
inherits all of its functionality but is aimed to use as little memory
as possible based on a balance between memory demand and performance.
Most of the spectrum-specific data, like retention time, polarity, total
ion current are stored within the object’s <code>featureData</code>
slot. The actual M/Z and intensity values from the individual spectra
are, in contrast to <code>MSnExp</code> objects, not kept in memory (in
the <code>assayData</code> slot), but are fetched from the original
files on-demand. Because mzML files are indexed, using the <em><a href="https://bioconductor.org/packages/3.18/mzR" class="external-link">mzR</a></em> package
to read the relevant spectrum data is fast and only moderately slower
than for in-memory <code>MSnExp</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<p>To keep track of data manipulation steps that are applied to spectrum
data (such as performed by methods <code>removePeaks</code> or
<code>clean</code>) a <em>lazy execution</em> framework was implemented.
Methods that manipulate or subset a spectrum’s M/Z or intensity values
can not be applied directly to a <code>OnDiskMSnExp</code> object, since
the relevant data is not kept in memory. Thus, any call to a processing
method that changes or subset M/Z or intensity values are added as
<code>ProcessingStep</code> items to the object’s
<code>spectraProcessingQueue</code>. When the spectrum data is then
queried from an <code>OnDiskMSnExp</code>, the spectra are read in from
the file and all these processing steps are applied on-the-fly to the
spectrum data before being returned to the user.</p>
<p>The operations involving extracting or manipulating spectrum data are
applied on a per-file basis, which enables parallel processing. Thus,
all corresponding method implementations for <code>OnDiskMSnExp</code>
objects have an argument <code>BPPARAM</code> and users can set a
<code>PARALLEL_THRESH</code> option flag<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> that enables to define
how and when parallel processing should be performed (using the <em><a href="https://bioconductor.org/packages/3.18/BiocParallel" class="external-link">BiocParallel</a></em>
package).</p>
<p>Note that all data manipulations that are not applied to M/Z or
intensity values of a spectrum (e.g. sub-setting by retention time etc)
are very fast as they operate directly to the object’s
<code>featureData</code> slot.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"OnDiskMSnExp"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "OnDiskMSnExp" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                            </span></span>
<span><span class="co">## Name:  spectraProcessingQueue                backend              assayData</span></span>
<span><span class="co">## Class:                   list              character            environment</span></span>
<span><span class="co">##                                                                            </span></span>
<span><span class="co">## Name:               phenoData            featureData         experimentData</span></span>
<span><span class="co">## Class:     AnnotatedDataFrame     AnnotatedDataFrame                  MIAxE</span></span>
<span><span class="co">##                                                                            </span></span>
<span><span class="co">## Name:            protocolData         processingData                 .cache</span></span>
<span><span class="co">## Class:     AnnotatedDataFrame             MSnProcess            environment</span></span>
<span><span class="co">##                              </span></span>
<span><span class="co">## Name:       .__classVersion__</span></span>
<span><span class="co">## Class:               Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: </span></span>
<span><span class="co">## Class "MSnExp", directly</span></span>
<span><span class="co">## Class "pSet", by class "MSnExp", distance 2</span></span>
<span><span class="co">## Class "Versioned", by class "MSnExp", distance 3</span></span></code></pre>
<p>The distinction between <code>MSnExp</code> and
<code>OnDiskMSnExp</code> is often not explicitly stated as it should
not matter, from a user’s perspective, which data structure they are
working with, as both behave in equivalent ways. Often, they are
referred to as <em>in-memory</em> and <em>on-disk</em>
<code>MSnExp</code> implementations.</p>
</div>
<div class="section level3">
<h3 id="msnset-a-class-for-quantitative-proteomics-data">
<code>MSnSet</code>: a class for quantitative proteomics data<a class="anchor" aria-label="anchor" href="#msnset-a-class-for-quantitative-proteomics-data"></a>
</h3>
<p>This class stores quantitation data and meta data after running
<code>quantify</code> on an <code>MSnExp</code> object or by creating an
<code>MSnSet</code> instance from an external file, as described in the
<em>MSnbase-io</em> vignette and in <code><a href="../reference/readMSnSet.html">?readMSnSet</a></code>,
<code>readMzTabData</code>, etc. The quantitative data is in form of a
<em>n</em> by <em>p</em> matrix, where <em>n</em> is the number of
features/spectra originally in the <code>MSnExp</code> used as parameter
in <code>quantify</code> and <em>p</em> is the number of reporter ions.
If read from an external file, <em>n</em> corresponds to the number of
features (protein groups, proteins, peptides, spectra) in the file and
<span class="math inline">\(p\)</span> is the number of columns with
quantitative data (samples) in the file.</p>
<p>This prompted to keep a similar implementation as the
<code>ExpressionSet</code> class, while adding the proteomics-specific
annotation slot introduced in the <code>pSet</code> class, namely
<code>processingData</code> for objects of class
<code>MSnProcess</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"MSnSet"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "MSnSet" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:      experimentData     processingData               qual</span></span>
<span><span class="co">## Class:              MIAPE         MSnProcess         data.frame</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:           assayData          phenoData        featureData</span></span>
<span><span class="co">## Class:          AssayData AnnotatedDataFrame AnnotatedDataFrame</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:          annotation       protocolData  .__classVersion__</span></span>
<span><span class="co">## Class:          character AnnotatedDataFrame           Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: </span></span>
<span><span class="co">## Class "eSet", directly</span></span>
<span><span class="co">## Class "VersionedBiobase", by class "eSet", distance 2</span></span>
<span><span class="co">## Class "Versioned", by class "eSet", distance 3</span></span></code></pre>
<p>The <code>MSnSet</code> class extends the virtual <code>eSet</code>
class to provide compatibility for <code>ExpressionSet</code>-like
behaviour. The experiment meta-data in <code>experimentData</code> is
also of class <code>MIAPE</code> . The <code>annotation</code> slot,
inherited from <code>eSet</code> is not used. As a result, it is easy to
convert <code>ExpressionSet</code> data from/to <code>MSnSet</code>
objects with the coersion method <code>as</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "MSnSet"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "MSnbase"</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="fu"><a href="../reference/mzRident2dfr.html">as</a></span><span class="op">(</span><span class="va">msnset</span>, <span class="st">"ExpressionSet"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ExpressionSet"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Biobase"</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ExpressionSet"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "Biobase"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="fu"><a href="../reference/mzRident2dfr.html">as</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span>, <span class="st">"MSnSet"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "MSnSet"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "MSnbase"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="MSnProcess">
<code>MSnProcess</code>: a class for logging processing meta
data<a class="anchor" aria-label="anchor" href="#MSnProcess"></a>
</h3>
<p>This class aims at recording specific manipulations applied to
<code>MSnExp</code> or <code>MSnSet</code> instances. The
<code>processing</code> slot is a <code>character</code> vector that
describes major processing. Most other slots are of class
<code>logical</code> that indicate whether the data has been centroided,
smoothed, although many of the functionality is not implemented yet. Any
new processing that is implemented should be documented and logged
here.</p>
<p>It also documents the raw data file from which the data originates
(<code>files</code> slot) and the <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
version that was in use when the <code>MSnProcess</code> instance, and
hence the <code>MSnExp</code>/<code>MSnSet</code> objects, were
originally created.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"MSnProcess"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "MSnProcess" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:              files        processing            merged           cleaned</span></span>
<span><span class="co">## Class:         character         character           logical           logical</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:       removedPeaks          smoothed           trimmed        normalised</span></span>
<span><span class="co">## Class:         character           logical           numeric           logical</span></span>
<span><span class="co">##                                           </span></span>
<span><span class="co">## Name:     MSnbaseVersion .__classVersion__</span></span>
<span><span class="co">## Class:         character          Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: "Versioned"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="miape-minimum-information-about-a-proteomics-experiment">
<code>MIAPE</code>: Minimum Information About a Proteomics
Experiment<a class="anchor" aria-label="anchor" href="#miape-minimum-information-about-a-proteomics-experiment"></a>
</h3>
<p>The Minimum Information About a Proteomics Experiment <span class="citation">(Chris F. Taylor et al. 2007; Chris F. Taylor et al.
2008)</span> <code>MIAPE</code> class describes the experiment,
including contact details, information about the mass spectrometer and
control and analysis software.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"MIAPE"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "MIAPE" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                     title                      url</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                  abstract                pubMedIds</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                   samples            preprocessing</span></span>
<span><span class="co">## Class:                     list                     list</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                     other                dateStamp</span></span>
<span><span class="co">## Class:                     list                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                      name                      lab</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                   contact                    email</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:           instrumentModel   instrumentManufacturer</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:  instrumentCustomisations             softwareName</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:           softwareVersion        switchingCriteria</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:            isolationWidth            parameterFile</span></span>
<span><span class="co">## Class:                  numeric                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                 ionSource         ionSourceDetails</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:                  analyser          analyserDetails</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:              collisionGas        collisionPressure</span></span>
<span><span class="co">## Class:                character                  numeric</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:           collisionEnergy             detectorType</span></span>
<span><span class="co">## Class:                character                character</span></span>
<span><span class="co">##                                                         </span></span>
<span><span class="co">## Name:       detectorSensitivity        .__classVersion__</span></span>
<span><span class="co">## Class:                character                 Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: </span></span>
<span><span class="co">## Class "MIAxE", directly</span></span>
<span><span class="co">## Class "Versioned", by class "MIAxE", distance 2</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="Spectum">
<code>Spectrum</code> <em>et al.</em>: classes for MS spectra<a class="anchor" aria-label="anchor" href="#Spectum"></a>
</h3>
<p><code>Spectrum</code> is a virtual class that defines common
attributes to all types of spectra. MS1 and MS2 specific attributes are
defined in the <code>Spectrum1</code> and <code>Spectrum2</code>
classes, that directly extend <code>Spectrum</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"Spectrum"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Virtual Class "Spectrum" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:            msLevel        peaksCount                rt    acquisitionNum</span></span>
<span><span class="co">## Class:           integer           integer           numeric           integer</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:          scanIndex               tic                mz         intensity</span></span>
<span><span class="co">## Class:           integer           numeric           numeric           numeric</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:           fromFile        centroided          smoothed          polarity</span></span>
<span><span class="co">## Class:           integer           logical           logical           integer</span></span>
<span><span class="co">##                         </span></span>
<span><span class="co">## Name:  .__classVersion__</span></span>
<span><span class="co">## Class:          Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: "Versioned"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Known Subclasses: "Spectrum2", "Spectrum1"</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"Spectrum1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "Spectrum1" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:            msLevel        peaksCount                rt    acquisitionNum</span></span>
<span><span class="co">## Class:           integer           integer           numeric           integer</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:          scanIndex               tic                mz         intensity</span></span>
<span><span class="co">## Class:           integer           numeric           numeric           numeric</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:           fromFile        centroided          smoothed          polarity</span></span>
<span><span class="co">## Class:           integer           logical           logical           integer</span></span>
<span><span class="co">##                         </span></span>
<span><span class="co">## Name:  .__classVersion__</span></span>
<span><span class="co">## Class:          Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: </span></span>
<span><span class="co">## Class "Spectrum", directly</span></span>
<span><span class="co">## Class "Versioned", by class "Spectrum", distance 2</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"Spectrum2"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "Spectrum2" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:              merged        precScanNum        precursorMz</span></span>
<span><span class="co">## Class:            numeric            integer            numeric</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:  precursorIntensity    precursorCharge    collisionEnergy</span></span>
<span><span class="co">## Class:            numeric            integer            numeric</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:             msLevel         peaksCount                 rt</span></span>
<span><span class="co">## Class:            integer            integer            numeric</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:      acquisitionNum          scanIndex                tic</span></span>
<span><span class="co">## Class:            integer            integer            numeric</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:                  mz          intensity           fromFile</span></span>
<span><span class="co">## Class:            numeric            numeric            integer</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:          centroided           smoothed           polarity</span></span>
<span><span class="co">## Class:            logical            logical            integer</span></span>
<span><span class="co">##                          </span></span>
<span><span class="co">## Name:   .__classVersion__</span></span>
<span><span class="co">## Class:           Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: </span></span>
<span><span class="co">## Class "Spectrum", directly</span></span>
<span><span class="co">## Class "Versioned", by class "Spectrum", distance 2</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="reporterions-a-class-for-isobaric-tags">
<code>ReporterIons</code>: a class for isobaric tags<a class="anchor" aria-label="anchor" href="#reporterions-a-class-for-isobaric-tags"></a>
</h3>
<p>The iTRAQ and TMT (or any other peak of interest) are implemented
<code>ReporterIons</code> instances, that essentially defines an
expected MZ position for the peak and a width around this value as well
a names for the reporters.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"ReporterIons"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "ReporterIons" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:               name     reporterNames       description                mz</span></span>
<span><span class="co">## Class:         character         character         character           numeric</span></span>
<span><span class="co">##                                                             </span></span>
<span><span class="co">## Name:                col             width .__classVersion__</span></span>
<span><span class="co">## Class:         character           numeric          Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: "Versioned"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="chromatogram-and-mchromatograms-classes-to-handle-chromatographic-data">
<code>Chromatogram</code> and <code>MChromatograms</code>: classes
to handle chromatographic data<a class="anchor" aria-label="anchor" href="#chromatogram-and-mchromatograms-classes-to-handle-chromatographic-data"></a>
</h3>
<p>The <code>Chromatogram</code> class represents chromatographic MS
data, i.e. retention time and intensity duplets for one file/sample. The
<code>MChromatograms</code> class (Matrix of Chromatograms) allows to
arrange multiple <code>Chromatogram</code> instances in a
two-dimensional grid, with columns supposed to represent different
samples and rows two-dimensional areas in the plane spanned by the m/z
and retention time dimensions from which the intensities are extracted
(e.g. an extracted ion chromatogram for a specific ion). The
<code>MChromatograms</code> class extends the base <code>matrix</code>
class. <code>MChromatograms</code> objects can be extracted from an
<code>MSnExp</code> or <code>OnDiskMSnExp</code> object using the
<code>chromatogram</code> method.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"Chromatogram"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "Chromatogram" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:              rtime         intensity                mz          filterMz</span></span>
<span><span class="co">## Class:           numeric           numeric           numeric           numeric</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:        precursorMz         productMz          fromFile    aggregationFun</span></span>
<span><span class="co">## Class:           numeric           numeric           integer         character</span></span>
<span><span class="co">##                                           </span></span>
<span><span class="co">## Name:            msLevel .__classVersion__</span></span>
<span><span class="co">## Class:           integer          Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: "Versioned"</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"MChromatograms"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "MChromatograms" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                </span></span>
<span><span class="co">## Name:               .Data          phenoData        featureData</span></span>
<span><span class="co">## Class:             matrix AnnotatedDataFrame AnnotatedDataFrame</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: </span></span>
<span><span class="co">## Class "matrix", from data part</span></span>
<span><span class="co">## Class "array", by class "matrix", distance 2</span></span>
<span><span class="co">## Class "replValueSp", by class "matrix", distance 2</span></span>
<span><span class="co">## Class "structure", by class "matrix", distance 3</span></span>
<span><span class="co">## Class "matrix_OR_array_OR_table_OR_numeric", by class "matrix", distance 3</span></span>
<span><span class="co">## Class "vector", by class "matrix", distance 4, with explicit coerce</span></span>
<span><span class="co">## Class "vector_OR_factor", by class "matrix", distance 5, with explicit coerce</span></span>
<span><span class="co">## Class "vector_OR_Vector", by class "matrix", distance 5, with explicit coerce</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="other-classes">Other classes<a class="anchor" aria-label="anchor" href="#other-classes"></a>
</h3>
<div class="section level4">
<h4 class="unnumbered" id="lists-of-msnset-instances">Lists of <code>MSnSet</code> instances<a class="anchor" aria-label="anchor" href="#lists-of-msnset-instances"></a>
</h4>
<p>When several <code>MSnSet</code> instances are related to each other
and should be stored together as different objects, they can be grouped
as a list into and <code>MSnSetList</code> object. In addition to the
actual <code>list</code> slot, this class also has basic logging
functionality and enables iteration over the <code>MSnSet</code>
instances using a dedicated <code>lapply</code> methods.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/methods/getClass.html" class="external-link">getClass</a></span><span class="op">(</span><span class="st">"MSnSetList"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Class "MSnSetList" [package "MSnbase"]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Slots:</span></span>
<span><span class="co">##                                                                               </span></span>
<span><span class="co">## Name:                  x               log       featureData .__classVersion__</span></span>
<span><span class="co">## Class:              list              list         DataFrame          Versions</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Extends: "Versioned"</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a>
</h2>
<div class="section level5">
<h5 class="unnumbered" id="unit-tests">Unit tests<a class="anchor" aria-label="anchor" href="#unit-tests"></a>
</h5>
<p><em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
implements unit tests with the <em><a href="https://CRAN.R-project.org/package=testthat" class="external-link">testthat</a></em>
package.</p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="processing-methods">Processing methods<a class="anchor" aria-label="anchor" href="#processing-methods"></a>
</h5>
<p>Methods that process raw data, i.e. spectra should be implemented for
<code>Spectrum</code> objects first and then <code>eapply</code>ed (or
similar) to the <code>assayData</code> slot of an <code>MSnExp</code>
instance in the specific method.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] MSnbase_2.29.3      ProtGenerics_1.34.0 S4Vectors_0.40.2   </span></span>
<span><span class="co">## [4] mzR_2.36.0          Rcpp_1.0.12         Biobase_2.62.0     </span></span>
<span><span class="co">## [7] BiocGenerics_0.48.1 BiocStyle_2.30.0   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] bitops_1.0-7                rlang_1.1.3                </span></span>
<span><span class="co">##  [3] magrittr_2.0.3              clue_0.3-65                </span></span>
<span><span class="co">##  [5] matrixStats_1.2.0           compiler_4.3.2             </span></span>
<span><span class="co">##  [7] systemfonts_1.0.6           vctrs_0.6.5                </span></span>
<span><span class="co">##  [9] pkgconfig_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">## [11] fastmap_1.1.1               XVector_0.42.0             </span></span>
<span><span class="co">## [13] utf8_1.2.4                  rmarkdown_2.26             </span></span>
<span><span class="co">## [15] preprocessCore_1.61.0       ragg_1.2.7                 </span></span>
<span><span class="co">## [17] purrr_1.0.2                 xfun_0.42                  </span></span>
<span><span class="co">## [19] MultiAssayExperiment_1.28.0 zlibbioc_1.48.0            </span></span>
<span><span class="co">## [21] cachem_1.0.8                GenomeInfoDb_1.38.7        </span></span>
<span><span class="co">## [23] jsonlite_1.8.8              DelayedArray_0.28.0        </span></span>
<span><span class="co">## [25] BiocParallel_1.36.0         parallel_4.3.2             </span></span>
<span><span class="co">## [27] cluster_2.1.6               R6_2.5.1                   </span></span>
<span><span class="co">## [29] bslib_0.6.1                 limma_3.58.1               </span></span>
<span><span class="co">## [31] GenomicRanges_1.54.1        jquerylib_0.1.4            </span></span>
<span><span class="co">## [33] bookdown_0.38               SummarizedExperiment_1.32.0</span></span>
<span><span class="co">## [35] iterators_1.0.14            knitr_1.45                 </span></span>
<span><span class="co">## [37] IRanges_2.36.0              Matrix_1.6-5               </span></span>
<span><span class="co">## [39] igraph_2.0.2                tidyselect_1.2.1           </span></span>
<span><span class="co">## [41] abind_1.4-5                 yaml_2.3.8                 </span></span>
<span><span class="co">## [43] doParallel_1.0.17           codetools_0.2-19           </span></span>
<span><span class="co">## [45] affy_1.80.0                 lattice_0.22-5             </span></span>
<span><span class="co">## [47] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">## [49] evaluate_0.23               desc_1.4.3                 </span></span>
<span><span class="co">## [51] pillar_1.9.0                affyio_1.72.0              </span></span>
<span><span class="co">## [53] BiocManager_1.30.22         MatrixGenerics_1.14.0      </span></span>
<span><span class="co">## [55] foreach_1.5.2               MALDIquant_1.22.2          </span></span>
<span><span class="co">## [57] ncdf4_1.22                  generics_0.1.3             </span></span>
<span><span class="co">## [59] RCurl_1.98-1.14             ggplot2_3.5.0              </span></span>
<span><span class="co">## [61] munsell_0.5.0               scales_1.3.0               </span></span>
<span><span class="co">## [63] glue_1.7.0                  lazyeval_0.2.2             </span></span>
<span><span class="co">## [65] tools_4.3.2                 mzID_1.40.0                </span></span>
<span><span class="co">## [67] QFeatures_1.12.0            vsn_3.70.0                 </span></span>
<span><span class="co">## [69] fs_1.6.3                    XML_3.99-0.16.1            </span></span>
<span><span class="co">## [71] grid_4.3.2                  impute_1.76.0              </span></span>
<span><span class="co">## [73] MsCoreUtils_1.14.1          colorspace_2.1-0           </span></span>
<span><span class="co">## [75] GenomeInfoDbData_1.2.11     PSMatch_1.6.0              </span></span>
<span><span class="co">## [77] cli_3.6.2                   textshaping_0.3.7          </span></span>
<span><span class="co">## [79] fansi_1.0.6                 S4Arrays_1.2.1             </span></span>
<span><span class="co">## [81] dplyr_1.1.4                 AnnotationFilter_1.26.0    </span></span>
<span><span class="co">## [83] pcaMethods_1.94.0           gtable_0.3.4               </span></span>
<span><span class="co">## [85] sass_0.4.8                  digest_0.6.35              </span></span>
<span><span class="co">## [87] SparseArray_1.2.4           htmlwidgets_1.6.4          </span></span>
<span><span class="co">## [89] memoise_2.0.1               htmltools_0.5.7            </span></span>
<span><span class="co">## [91] pkgdown_2.0.7.9000          lifecycle_1.0.4            </span></span>
<span><span class="co">## [93] statmod_1.5.0               MASS_7.3-60.0.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Gentleman2004" class="csl-entry">
Gentleman, Robert C., Vincent J. Carey, Douglas M. Bates, Ben Bolstad,
Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004.
<span>“Bioconductor: Open Software Development for Computational Biology
and Bioinformatics.”</span> <em>Genome Biol</em> 5 (10): –80. <a href="https://doi.org/10.1186/gb-2004-5-10-r80" class="external-link">https://doi.org/10.1186/gb-2004-5-10-r80</a>.
</div>
<div id="ref-Ross2004" class="csl-entry">
Ross, Philip L., Yulin N. Huang, Jason N. Marchese, Brian Williamson,
Kenneth Parker, Stephen Hattan, Nikita Khainovski, et al. 2004.
<span>“Multiplexed Protein Quantitation in Saccharomyces Cerevisiae
Using Amine-Reactive Isobaric Tagging Reagents.”</span> <em>Mol Cell
Proteomics</em> 3 (12): 1154–69. <a href="https://doi.org/10.1074/mcp.M400129-MCP200" class="external-link">https://doi.org/10.1074/mcp.M400129-MCP200</a>.
</div>
<div id="ref-Taylor2008" class="csl-entry">
Taylor, Chris F, Pierre-Alain Binz, Ruedi Aebersold, Michel Affolter,
Robert Barkovich, Eric W Deutsch, David M Horn, et al. 2008.
<span>“Guidelines for Reporting the Use of Mass Spectrometry in
Proteomics.”</span> <em>Nat. Biotechnol.</em> 26 (8): 860–61. <a href="https://doi.org/10.1038/nbt0808-860" class="external-link">https://doi.org/10.1038/nbt0808-860</a>.
</div>
<div id="ref-Taylor2007" class="csl-entry">
Taylor, Chris F., Norman W. Paton, Kathryn S. Lilley, Pierre-Alain Binz,
Randall K. Julian, Andrew R. Jones, Weimin Zhu, et al. 2007. <span>“The
Minimum Information about a Proteomics Experiment (MIAPE).”</span>
<em>Nat Biotechnol</em> 25 (8): 887–93. <a href="https://doi.org/10.1038/nbt1329" class="external-link">https://doi.org/10.1038/nbt1329</a>.
</div>
<div id="ref-Thompson2003" class="csl-entry">
Thompson, Andrew, Jürgen Schäfer, Karsten Kuhn, Stefan Kienle, Josef
Schwarz, Günter Schmidt, Thomas Neumann, R Johnstone, A Karim A
Mohammed, and Christian Hamon. 2003. <span>“<a href="https://www.ncbi.nlm.nih.gov/pubmed/12713048" class="external-link">Tandem Mass Tags: A
Novel Quantification Strategy for Comparative Analysis of Complex
Protein Mixtures by <span>MS/MS</span>.</a>”</span> <em>Anal. Chem.</em>
75 (8): 1895–1904.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>The <em>benchmarking</em> vignette compares data size
and operation speed of the two implementations.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>see <code><a href="../reference/MSnbaseOptions.html">?MSnbaseOptions</a></code> for details.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
